---
title: 'Linking Spi-B and PU.1 binding sites to genes'
author: "Holly Giles"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
---

This is an rmd to identify SPIB and PU.1 peaks that are differentially accesible between tri12 and WT CLL, and to link these to downstream genes. 

# Set up 
```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, message = TRUE, warning = TRUE, cache = TRUE)
```

Load libraries
```{r Load libraries}

################# load libraries ################# 

library(readr)
library(dplyr)
library(tidyverse)
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
library("org.Hs.eg.db")
library(clusterProfiler)
library(ggpubr)
library(magrittr)
library(patchwork)
library(ggplotify)
library(msigdbr)

```




Set plot directory
```{r plotDir4, message=FALSE}
plotDir = ifelse(exists(".standalone"), "", "../../inst/figs/")
if(plotDir!="") if(!file.exists(plotDir)) dir.create(plotDir)
```


## Load data
Load raw files
```{r loadData4, message=FALSE}

#df: tibble containing all screening viability data
peaklfc <- read_tsv("../data/BIGCLL52.SPIB.output.tsv.gz")

peaklfcbackup <- peaklfc

peaklfc %>% data.table::as.data.table()

```

## Define Aesthetics
```{r defineAesthetics4}

source("../R/themes_colors.R")

```

# Spi-B binding sites
Find top peaks 
```{r}
#filter for top peaks (all or up)
TFBS <- peaklfc %>% dplyr::filter(l2FC >1)
TFBS_pval <- peaklfc %>% dplyr::filter(pval_adj <0.1)
TFBS_neg <- peaklfc %>% dplyr::filter(l2FC < -1)


#remove Chr
TFBS <- TFBS %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))


#remove Chr
TFBS_pval <- TFBS_pval %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))

#remove Chr
TFBS_neg <- TFBS_neg %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))

```

Convert to a genomic ranges object
```{r}

peak<- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

peak_pval <- 
  TFBS_pval %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

peak_neg <- 
  TFBS_neg %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

```

Visualise where the bindng sites are and the size of the log2FC
```{r}

covplot(peak, weightCol="l2FC")

covplot(peak_pval, weightCol="l2FC")

covplot(peak_neg)


```

Find overlap between TSS regions and SPIB binding sites
```{r}
#get annotation database
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

#define TSS regions
promoter <- getPromoters(TxDb=txdb, upstream=3000, downstream=3000)

#Align peaks that map to tehse regions and generate TAG matrix
tagMatrix <- getTagMatrix(peak, windows=promoter)


```

Heatmap of SPIB binding sites in TSS regions
```{r}
tagHeatmap(tagMatrix, xlim=c(-3000, 3000), color="red")

```

Visualise distance from promoter regions
```{R}
plotAvgProf(tagMatrix, xlim=c(-3000, 3000),
            xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency")

```

Annotate peaks    
    
Here I use the annotatePeak function to find nearest gene to each of my SPIB binding sites, looking in the region Â±3kb around the tss sites defined in the TxDB object. The resulting `peakAnno` object contains all the original information on my SPIB bding sites, along with information on the annotation of these sites (eg promoter/Exon/intron/intergenic) and the nearest gene and its function. 
```{r}

peakAnno <- annotatePeak(peak, 
                         tssRegion=c(-3000, 3000), #region in whcih to look for genes
                         TxDb=txdb, #define where to get transcript info
                         annoDb="org.Hs.eg.db") #set annotation package to add ensembl, smbol and genames

#peakAnno is a csAnno onject - can convert:
as.data.frame(peakAnno) %>%
  DT::datatable()


#try with : 
#annotatePeakInBatch
```
Visualise annotations of SPIB binding sties
```{r}
plotAnnoPie(peakAnno)

```
Visualise distance of SPIB binding sites relative to TSS 
```{r}

plotDistToTSS(peakAnno,
              title="Distribution of SPIB-binding loci\nrelative to TSS")
```

Enrichment of annotated genes - look for biological terms that are enriched in nearby genes    

```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test
kegg <-
  dplyr::filter(kegg, 
         #All KEGG immune signaling pathways
         gs_name %in% c("KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION", 
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY" ,
                        "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY"  , 
                        "KEGG_NOTCH_SIGNALING_PATHWAY", 
                        "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY", 
                        "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        #Control genesets
                        "KEGG_OXIDATIVE_PHOSPHORYLATION",                  
                        "KEGG_DNA_REPLICATION"
         ))

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )
kegg$term <- gsub("B CELL RECEPTOR SIGNALING PATHWAY", 
                  "BCR Signaling Pathway",kegg$term )
kegg$term <- gsub("CHEMOKINE SIGNALING PATHWAY", 
                  "Chemokine Signaling Pathway",kegg$term )
kegg$term <- gsub("CYTOKINE CYTOKINE RECEPTOR INTERACTION", 
                  "Cytokine Cytokine Receptor Interaction",kegg$term )
kegg$term <- gsub("JAK STAT SIGNALING PATHWAY", 
                  "JAK STAT signaling pathway",kegg$term )
kegg$term <- gsub("NOD LIKE RECEPTOR SIGNALING PATHWAY", 
                  "NLR Signaling Pathway",kegg$term )
kegg$term <- gsub("NOTCH SIGNALING PATHWAY", 
                  "Notch Signaling Pathway",kegg$term )
kegg$term <- gsub("RIG I LIKE RECEPTOR SIGNALING PATHWAY", 
                  "RLR Signaling Pathway",kegg$term )
kegg$term <- gsub("TOLL LIKE RECEPTOR SIGNALING PATHWAY", 
                  "TLR Signaling Pathway",kegg$term )

kegg$term <- gsub("T CELL RECEPTOR SIGNALING PATHWAY", 
                  "TCR Signaling Pathway",kegg$term )
kegg$term <- gsub("OXIDATIVE PHOSPHORYLATION", 
                  "Oxidative Phosphorylation",kegg$term )
kegg$term <- gsub("P53 SIGNALING PATHWAY", 
                  "P53 Signaling Pathway",kegg$term )
kegg$term <- gsub("DNA REPLICATION", 
                  "DNA Replication",kegg$term )

kegg$genesetDatabase <- "KEGG"


#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test
reactome <-
  dplyr::filter(reactome, 
         gs_name %in% c(#selected immune pathways
                        "REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX",
                        "REACTOME_TNF_SIGNALING",
                        "REACTOME_INTERLEUKIN_1_SIGNALING",
                        "REACTOME_INTERLEUKIN_10_SIGNALING",
                        "REACTOME_INTERLEUKIN_15_SIGNALING",
                        "REACTOME_INTERLEUKIN_2_SIGNALING",
                        "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
                        "REACTOME_INTERLEUKIN_6_SIGNALING" ,
                        "REACTOME_OTHER_INTERLEUKIN_SIGNALING",
                        "REACTOME_SIGNALING_BY_INTERLEUKINS",
                        "REACTOME_INTERFERON_ALPHA_BETA_SIGNALING" ,
                        "REACTOME_INTERFERON_GAMMA_SIGNALING", 
                        #Control genesets
                        "REACTOME_CELL_CYCLE",                  
                        "REACTOME_REGULATION_OF_TP53_ACTIVITY" 
                        ))

colnames(reactome) <- c("term", "gene")

#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )


#tidy up and put in lower case
reactome$term <- gsub("CELL CYCLE", 
                      "Cell Cycle",reactome$term )
reactome$term <- gsub("INTERLEUKIN 1 SIGNALING", 
                      "Interleukin 1 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 10 SIGNALING", 
                      "Interleukin 10 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 15 SIGNALING", 
                      "Interleukin 15 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 2 SIGNALING", 
                      "Interleukin 2 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 4 AND INTERLEUKIN 13 SIGNALING", 
                      "Interleukin 4 and Interleukin 13 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 6 SIGNALING",
                      "Interleukin 6 Signaling",reactome$term )
reactome$term <- gsub("OTHER INTERLEUKIN SIGNALING", 
                      "Other Interleukin Signaling",reactome$term )
reactome$term <- gsub("SIGNALING BY INTERLEUKINS", 
                      "Signaling by Interleukins",reactome$term )
reactome$term <- gsub("REGULATION OF TP53 ACTIVITY", 
                      "Regulation of TP53 Pathway",reactome$term )
reactome$term <- gsub("SIGNALING BY TGF BETA RECEPTOR COMPLEX",
                      # "Signaling by TGFbeta \nReceptor Complex", reactome$term )
                      "Signaling by TGFbeta Receptor Complex", reactome$term )
reactome$term <- gsub("TNF SIGNALING", 
                      "TNF Signaling",reactome$term )
reactome$term <- gsub("INTERFERON ALPHA BETA SIGNALING", 
                      "Interferon Alpha Beta Signaling",reactome$term )

reactome$term <- gsub("INTERFERON GAMMA SIGNALING", 
                      "Interferon Gamma Signaling", reactome$term )

#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"



#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.01)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

selectedPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


selectedPathways

```

```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )


kegg$genesetDatabase <- "KEGG"


#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test


colnames(reactome) <- c("term", "gene")

#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )





#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"



#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.01)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

allPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


allPathways

```
Just KEGG
```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )


kegg$genesetDatabase <- "KEGG"


#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- kegg %>% dplyr::select(term, gene)

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
              
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

keggPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


keggPathways

```


