---
title: 'Linking Spi-B and PU.1 binding sites to genes'
author: "Holly Giles"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
---
Make the figures and analysis solid in here, then move into supplement and figure 4, and tell Peter, Sascha and Wolfgang
Aim to do as much of this today as I can, so that tomorrow is for text and other htigns 


1. get out correct peaks [x] 
2. Optimise coverage plot <- almost there, just make sure x scale is appropriate and final check [x] 

3. Try adding Ibet peaks on top of that (adapt .id part of function)


4. use function to get three genes. 
4. Optimise GSEa
5. optimse GSEA plot
6. Add back in  

= good progress now

Carry on with this anlaysis tomorrow morning
Then use Ivan's analyses in afternoon 
THen Wednesday put things into text and rebuttal 
Then carry on with other ranomd things after esp on Firday 

This is an rmd to identify SPIB and PU.1 peaks that are differentially accessible between tri12 and WT CLL, and to link these to downstream genes. 

# Set up 
```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, message = TRUE, warning = TRUE, cache = TRUE)
```

Load libraries
```{r Load libraries}

################# load libraries ################# 

library(readr)
library(dplyr)
library(tidyverse)
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(GenomicRanges)
library("org.Hs.eg.db")
library(clusterProfiler)
library(ggpubr)
library(magrittr)
library(patchwork)
library(ggplotify)
library(msigdbr)
library(GenomicRanges)

```

Set plot directory
```{r plotDir4, message=FALSE}
plotDir = ifelse(exists(".standalone"), "", "../../inst/figs/")
if(plotDir!="") if(!file.exists(plotDir)) dir.create(plotDir)
```


## Load data
Load raw files
```{r loadData4, message=FALSE}

#df: tibble containing all screening viability data
peaklfc <- read_tsv("../data/BIGCLL52.SPIB.output.tsv.gz")
peaklfc_ibet <- read_tsv("../data/IBETvsCTRL.SPIB.0.A.output.tsv.gz")

peaklfcbackup <- peaklfc

peaklfc %>% data.table::as.data.table()

```

## Define Aesthetics
```{r defineAesthetics4}

source("../R/themes_colors.R")

```


## Define Functions

```{r}

#sort chr by number 
sortChrName <- function(chr.name) {
    ## X, Y and M will cause warnings when change to number.
    noChr <- suppressWarnings(as.numeric(sub("chr", "", chr.name)))
    ## index of chromosome name are character, such as X, Y
    ch.idx <- which(is.na(noChr))
    
    n.idx <- which(!is.na(noChr))
    chr.n <- noChr[n.idx]

    chr.sorted <- chr.name[n.idx][order(chr.n)]
    if (length(ch.idx) != 0) {
        chr.sorted <- c(chr.sorted, sort(chr.name[ch.idx]))
    }
         
    return(chr.sorted)
}

```

# Spi-B binding sites
Find top peaks 
```{r}
#filter for top peaks (all or up)
TFBS <- peaklfc %>% dplyr::filter(abs(l2FC) >1, pval_adj <0.1)


#remove Chr
TFBS <- TFBS %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))

```

Convert to a genomic ranges object
```{r}

peak<- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

```

```{r fig.height = 10, fig.width = 10}

#get a dataframe of all SPIB binding sites that show sig Log2FC
peak_df <- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj)

#make factors / numeric
chr.sorted <- unique(sortChrName(peak_df$chr))
peak_df$chr <- factor(peak_df$chr, levels = chr.sorted)
peak_df$start <- as.numeric(peak_df$start)
peak_df$end <- as.numeric(peak_df$end)


p <- ggplot(peak_df, aes(start, l2FC))+
  geom_rect(aes(xmin = start, ymin = 0, xmax = end, ymax = l2FC,
                color = ifelse(l2FC<0, palreds[5], palblues[1]), 
                fill= ifelse(l2FC<0, palreds[5], palblues[1]))) +
  scale_fill_manual(values = c(palreds[5], palblues[1]), guide = "none") +
  scale_color_manual(values = c(palreds[5], palblues[1]), guide = "none") +
  facet_grid(chr ~ ., scales = "fixed", space = "fixed")+ 
  xlab("Chromosome Size (bp)") + 
  ylab("L2FC") + 
  ggtitle("L2FCs for Spi-B binding sites")+
  t2 + theme(strip.text.y = element_text(angle = 360)) + 
  ylim(-3, 3) + 
  geom_hline(yintercept=0)

  

p


```



# Spi-B binding sites and IBET binding sties

Find top peaks 
```{r}
#filter for all SPIB sites that are downregulated by IBET treatment
TFBS_IBET <- peaklfc_ibet %>% dplyr::filter(l2FC <0)

#filter for all SPIB sites that are upregulated in tri12
TFBS_up <- peaklfc %>% dplyr::filter(l2FC>0)


#remove Chr
TFBS_IBET <- TFBS_IBET %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))


TFBS_up <- TFBS_up %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))

```

Convert to a genomic ranges object
```{r}

peak_IBET <- 
  TFBS_IBET %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

peak_up <- 
  TFBS_up %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

```

```{r fig.height = 10, fig.width = 10}

#get a dataframe of all SPIB binding sites that show sig Log2FC
peak_IBET_df <- 
  TFBS_IBET %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj)

peak_up_df <- 
  TFBS_up %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj)

#make factors / numeric
chr.sorted <- unique(sortChrName(peak_IBET_df$chr))
peak_IBET_df$chr <- factor(peak_IBET_df$chr, levels = chr.sorted)
peak_IBET_df$start <- as.numeric(peak_IBET_df$start)
peak_IBET_df$end <- as.numeric(peak_IBET_df$end)

chr.sorted <- unique(sortChrName(peak_up_df$chr))
peak_up_df$chr <- factor(peak_up_df$chr, levels = chr.sorted)
peak_up_df$start <- as.numeric(peak_up_df$start)
peak_up_df$end <- as.numeric(peak_up_df$end)

#check how many peaks overlap
#cehck how many Spi-B binding sites change with IBET treatment
#3326 SPIb sites are upregulated in tri 12
#SPIB sites are downregulated by IBET
#36 of these overlap = this isnt a very specific effect

peak_up
subsetByOverlaps(peak_up, peak_IBET)

#add column to denote origin
peak_up_df$condition <- "tri12"
peak_IBET_df$condition <- "ibet"

#merge dataframes
all_peaks <- rbind(peak_up_df, peak_IBET_df)


#plot both sets of peaks
q <- ggplot(all_peaks, aes(start, l2FC))+
  geom_rect(aes(xmin = start, ymin = 0, xmax = end, ymax = l2FC,
                color = condition)) +
  scale_color_manual(values = c( palblues[1], palreds[5]), guide = "none") +
  facet_grid(chr ~ ., scales = "fixed", space = "fixed")+ 
  xlab("Chromosome Size (bp)") + 
  ylab("L2FC") + 
  ggtitle("L2FCs for Spi-B binding sites for Trisomy 12 versus WT (Red), IBET-762 versus Control (Blue)")+
  t2 + theme(strip.text.y = element_text(angle = 360)) + 
  ylim(-3, 3) + 
  geom_hline(yintercept=0)

  

q


```
# Peak annotation


Annotate peaks    
    
Here I use the annotatePeak function to find nearest gene to each of my SPIB binding sites, looking in the region Â±3kb around the tss sites defined in the TxDB object. The resulting `peakAnno` object contains all the original information on my SPIB bding sites, along with information on the annotation of these sites (eg promoter/Exon/intron/intergenic) and the nearest gene and its function. 
```{r}

#get annotation database
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

#annoate all genes within 5kb
peakAnno <- annotatePeak(peak, 
                         tssRegion=c(-3000, 3000), #region in whcih to look for genes
                         TxDb=txdb, #define where to get transcript info
                         annoDb="org.Hs.eg.db", addFlankGeneInfo = TRUE) #set annotation package to add ensembl, smbol and genames

#peakAnno is a csAnno onject - can convert:
as.data.frame(peakAnno) %>%
  DT::datatable()



```


## Enrichment of selected pathways  

```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test
kegg <-
  dplyr::filter(kegg, 
         #All KEGG immune signaling pathways
         gs_name %in% c("KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION", 
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY" ,
                        "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY"  , 
                        "KEGG_NOTCH_SIGNALING_PATHWAY", 
                        "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY", 
                        "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        #Control genesets
                        "KEGG_OXIDATIVE_PHOSPHORYLATION",                  
                        "KEGG_DNA_REPLICATION"
         ))

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )
kegg$term <- gsub("B CELL RECEPTOR SIGNALING PATHWAY", 
                  "BCR Signaling Pathway",kegg$term )
kegg$term <- gsub("CHEMOKINE SIGNALING PATHWAY", 
                  "Chemokine Signaling Pathway",kegg$term )
kegg$term <- gsub("CYTOKINE CYTOKINE RECEPTOR INTERACTION", 
                  "Cytokine Cytokine Receptor Interaction",kegg$term )
kegg$term <- gsub("JAK STAT SIGNALING PATHWAY", 
                  "JAK STAT signaling pathway",kegg$term )
kegg$term <- gsub("NOD LIKE RECEPTOR SIGNALING PATHWAY", 
                  "NLR Signaling Pathway",kegg$term )
kegg$term <- gsub("NOTCH SIGNALING PATHWAY", 
                  "Notch Signaling Pathway",kegg$term )
kegg$term <- gsub("RIG I LIKE RECEPTOR SIGNALING PATHWAY", 
                  "RLR Signaling Pathway",kegg$term )
kegg$term <- gsub("TOLL LIKE RECEPTOR SIGNALING PATHWAY", 
                  "TLR Signaling Pathway",kegg$term )

kegg$term <- gsub("T CELL RECEPTOR SIGNALING PATHWAY", 
                  "TCR Signaling Pathway",kegg$term )
kegg$term <- gsub("OXIDATIVE PHOSPHORYLATION", 
                  "Oxidative Phosphorylation",kegg$term )
kegg$term <- gsub("P53 SIGNALING PATHWAY", 
                  "P53 Signaling Pathway",kegg$term )
kegg$term <- gsub("DNA REPLICATION", 
                  "DNA Replication",kegg$term )

kegg$genesetDatabase <- "KEGG"


#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test
reactome <-
  dplyr::filter(reactome, 
         gs_name %in% c(#selected immune pathways
                        "REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX",
                        "REACTOME_TNF_SIGNALING",
                        "REACTOME_INTERLEUKIN_1_SIGNALING",
                        "REACTOME_INTERLEUKIN_10_SIGNALING",
                        "REACTOME_INTERLEUKIN_15_SIGNALING",
                        "REACTOME_INTERLEUKIN_2_SIGNALING",
                        "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
                        "REACTOME_INTERLEUKIN_6_SIGNALING" ,
                        "REACTOME_OTHER_INTERLEUKIN_SIGNALING",
                        "REACTOME_SIGNALING_BY_INTERLEUKINS",
                        "REACTOME_INTERFERON_ALPHA_BETA_SIGNALING" ,
                        "REACTOME_INTERFERON_GAMMA_SIGNALING", 
                        #Control genesets
                        "REACTOME_CELL_CYCLE",                  
                        "REACTOME_REGULATION_OF_TP53_ACTIVITY" 
                        ))

colnames(reactome) <- c("term", "gene")

#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )


#tidy up and put in lower case
reactome$term <- gsub("CELL CYCLE", 
                      "Cell Cycle",reactome$term )
reactome$term <- gsub("INTERLEUKIN 1 SIGNALING", 
                      "Interleukin 1 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 10 SIGNALING", 
                      "Interleukin 10 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 15 SIGNALING", 
                      "Interleukin 15 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 2 SIGNALING", 
                      "Interleukin 2 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 4 AND INTERLEUKIN 13 SIGNALING", 
                      "Interleukin 4 and Interleukin 13 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 6 SIGNALING",
                      "Interleukin 6 Signaling",reactome$term )
reactome$term <- gsub("OTHER INTERLEUKIN SIGNALING", 
                      "Other Interleukin Signaling",reactome$term )
reactome$term <- gsub("SIGNALING BY INTERLEUKINS", 
                      "Signaling by Interleukins",reactome$term )
reactome$term <- gsub("REGULATION OF TP53 ACTIVITY", 
                      "Regulation of TP53 Pathway",reactome$term )
reactome$term <- gsub("SIGNALING BY TGF BETA RECEPTOR COMPLEX",
                      # "Signaling by TGFbeta \nReceptor Complex", reactome$term )
                      "Signaling by TGFbeta Receptor Complex", reactome$term )
reactome$term <- gsub("TNF SIGNALING", 
                      "TNF Signaling",reactome$term )
reactome$term <- gsub("INTERFERON ALPHA BETA SIGNALING", 
                      "Interferon Alpha Beta Signaling",reactome$term )

reactome$term <- gsub("INTERFERON GAMMA SIGNALING", 
                      "Interferon Gamma Signaling", reactome$term )

#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"

```

### Single gene annotation
```{r fig.width =10}

#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

selectedPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


selectedPathways


```

### All flanking genes 
```{r fig.width = 7.5}

#get all genes including flanking genes 
#get flanking genes as a list
newpeakAnno <- 
  as.data.frame(peakAnno) %>%
  separate(flank_geneIds, c("A", "B", "C", "D", "E", "F", "G", "H", "I"), ";")
#get column as a vector
geneList<- c(newpeakAnno$geneId, newpeakAnno$A, newpeakAnno$B, newpeakAnno$C, newpeakAnno$D, newpeakAnno$E, newpeakAnno$F, newpeakAnno$G, newpeakAnno$H, newpeakAnno$I)

#remove NAs
geneList <- geneList[!is.na(geneList)]

#remove duplicates
geneList <- unique(geneList)

spibRes <- enricher(geneList, 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(geneList),")", sep= ""), 
                      "SPIB\np-value")

selectedPathways_allGenes <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


selectedPathways_allGenes


```

## Enrichment of all KEGG and Reactome
```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )


kegg$genesetDatabase <- "KEGG"


#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test
colnames(reactome) <- c("term", "gene")

#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )


#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"

```

### Single gene annotation
```{r fig.width = 10, fig.height=10}

#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

allPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


allPathways

```

### All flanking genes
```{r fig.width = 10, fig.height =10}

#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

#get all genes including flanking genes 
#get flanking genes as a list
newpeakAnno <- 
  as.data.frame(peakAnno) %>%
  separate(flank_geneIds, c("A", "B", "C", "D", "E", "F", "G", "H", "I"), ";")
#get column as a vector
geneList<- c(newpeakAnno$geneId, newpeakAnno$A, newpeakAnno$B, newpeakAnno$C, newpeakAnno$D, newpeakAnno$E, newpeakAnno$F, newpeakAnno$G, newpeakAnno$H, newpeakAnno$I)

#remove NAs
geneList <- geneList[!is.na(geneList)]

#remove duplicates
geneList <- unique(geneList)

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(geneList, 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(geneList)) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(geneList),")", sep= ""), 
                      "SPIB\np-value")

allPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


allPathways

```

## Only KEGG pathways
```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )


kegg$genesetDatabase <- "KEGG"

```

### Single gene annotation
```{r fig.width=10}
#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- kegg %>% dplyr::select(term, gene)

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
              
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

keggPathways <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


keggPathways

```


### All flanking genes 
```{r fig.width=10}
#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- kegg %>% dplyr::select(term, gene)


##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene


#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

#get all genes including flanking genes 
#get flanking genes as a list
newpeakAnno <- 
  as.data.frame(peakAnno) %>%
  separate(flank_geneIds, c("A", "B", "C", "D", "E", "F", "G", "H", "I"), ";")
#get column as a vector
geneList<- c(newpeakAnno$geneId, newpeakAnno$A, newpeakAnno$B, newpeakAnno$C, newpeakAnno$D, newpeakAnno$E, newpeakAnno$F, newpeakAnno$G, newpeakAnno$H, newpeakAnno$I)

#remove NAs
geneList <- geneList[!is.na(geneList)]

#remove duplicates
geneList <- unique(geneList)

spibRes <- enricher(geneList, 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.05)

colnames(resTab) <- c("Pathway",
              
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(geneList),")", sep= ""), 
                      "SPIB\np-value")

keggPathways_allgenes <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


keggPathways_allgenes

```
