---
title: '"Trisomy12_Annotation"
author: "Ivan Berest and Holly Giles"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
      code_folding: "hide" 
---

This is an rmd to annotate Trisomy 12 status of ATACseq samples taken from Rendeiro et al., 2016, EGA: EGAD00001002110, and used in Figure 4B of paper:


<br />

**Mapping drug-microenvironment-genetic interplay in CLL reveals trisomy 12 as a modulator of microenvironmental signals**

<br>


# Set up 
```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE, message = TRUE, warning = TRUE, cache = TRUE)
set.seed(1996)
```

Load libraries
```{r loadLibraries}

library(tidyverse)

```


# Load data
Note: Peaks file was generated by running processing pipeline outlined in Berest et al., 2019  (PMID: 31801079 DOI: 10.1016/j.celrep.2019.10.106), on  bam files downloaded from EGA: EGAD00001002110, Rendeiro et al., 2016 (PMID: 27346425  DOI: 10.1038/ncomms11938)

```{r loadData}


#peaks file contains genomic locations of peaks, and number of reads for each sample, this can be generated as aboce
peaks.CLLall = read.table("../../data/CLL.allBams.peaks.overlaps.bed", header = T)

#sample table contains meta data with SampleID, location of bam files, location of Peak files, Treatment, IGHV status, Sex, this can be acquired along with the bam files 
meta.CLLall = read_tsv("../../data/sampleTable_TF.tsv")

```

# Annotate Trisomy 12 status
```{r annotateTri12}

# make 

colnames(peaks.CLLall) = c(colnames(peaks.CLLall[,1:6]), meta.CLLall$SampleID)
vector = c()
for (sample in 7:ncol(peaks.CLLall)) {
  meanchr12 = mean(peaks.CLLall[which(peaks.CLLall$Chr == "chr12"),sample])
  meanall = mean(peaks.CLLall[which(peaks.CLLall$Chr != "chr12"),sample])
  finalratio = meanchr12 / meanall
  names(finalratio) = colnames(peaks.CLLall)[sample]
  if (finalratio > 1.4) {
    print(finalratio)
    vector <- append(vector, finalratio)
  }
}

```

# Add to table 
```{r}
tri12 <- data.frame(vector) %>% rownames_to_column("SampleID")
tri12$vector <- 1
colnames(tri12) <- c("SampleID", "trisomy12")

sampleTable <- left_join(meta.CLLall, tri12, by = "SampleID" )
sampleTable$trisomy12[is.na(sampleTable$trisomy12)] <- 0

sampleTable$bamReads <- gsub("/scratch/carnold/CLL/TF_act_noGCBias/input/bam/", "", sampleTable$bamReads)

sampleTable <- select(sampleTable, SampleID, bamReads, Treatment, Condition, Gender, trisomy12)
colnames(sampleTable) <- c("SampleID", "bamReads", "Batch", "IGHV.status", "Gender", "trisomy12")

write_tsv(sampleTable, "sampleTable_withtrisomy12.tsv")


```