---
title: 'Drug-microenvironment mapping reveals resistance mechanisms and prognostic patient subgroups in CLL'
subtitle: 'Appendix'
author: 
- Peter-Martin Bruch\*,
- Holly A. R. Giles\*,
- Carolin Kolb ,
- Sophie A. Herbst,
- Tina Becirovic,
- Tobias Roider,
- Junyan Lu,
- Sebastian Scheinost,
- Lena Wagner,
- Jennifer Huellein,
- Ivan Berest,
- Mark Kriegsmann,
- Katharina Kriegsmann,
- Christiane Zgorzelski,
- Peter Dreger,
- Judith B. Zaugg,
- Carsten MÃ¼ller-Tidow,
- Thorsten Zenz,
- Wolfgang Huber\*,
- Sascha Dietrich\*
- \* These authors contributed equally to this work.
date: "14th April 2022"
header-includes:
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \usepackage{float}
- \usepackage{fontspec}
- \setmainfont{Arial}

output:
  pdf_document:
      toc: true
      extra_dependencies: ["float"]
      dev: cairo_pdf
      latex_engine: xelatex
      includes:  
          in_header: preamble.tex
  BiocStyle::html_document:
      code_folding: "hide"
      toc: yes
      toc_depth: 1
      toc_float: yes
      self_contained: yes
---

\newpage
<!-- Set up  -->
```{R setUp_supp, echo = FALSE}

knitr::opts_chunk$set(echo = FALSE, message=FALSE, out.extra = "")

```

<!-- Set plot directory-->
```{r plotDirApp, message=FALSE, warning=FALSE}

plotDir = ifelse(exists(".standalone"), "", "../../inst/figs/")
if(plotDir!="") if(!file.exists(plotDir)) dir.create(plotDir)

```

<!-- Load Libraries -->
```{r loadLibraries_supp,  warning=FALSE}
#Libraries
library(survival)
library(maxstat)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(ChIPseeker)
library(org.Hs.eg.db)
library(clusterProfiler)
library(msigdbr)
library(rstatix)
library(kableExtra)
library(readxl)
library(xlsx)
library(fgsea)
library(DESeq2)
library(biobroom)
library(survminer)
library(patchwork)
library(data.table)
library(ggbeeswarm)
library(ggpubr)
library(ggrepel)
library(pheatmap)
library(Rtsne)
library(scales)
library(survival)
library(tidyverse)
library(corrplot)
library(gtools)
library(magrittr)
library(gtable)
library(glmnet)
library(RColorBrewer)
library(grid)
library(gridExtra)
library(cowplot)
library(ggh4x)
library(magick)
library(cicero)

```

<!-- # Define Functions -->
```{r factor2ind_supp}

factor2ind <- function(x, baseline)
{
  # Given a factor variable x, create an indicator matrix of dimension
  # length(x) x (nlevels(x)-1) dropping the column corresponding to the
  # baseline level (by default the first level is used as baseline).

  xname <- deparse(substitute(x))
  n <- length(x)
  x <- as.factor(x)
  if(!missing(baseline)) x <- relevel(x, baseline)
  X <- matrix(0, n, length(levels(x)))
  X[(1:n) + n*(unclass(x)-1)] <- 1
  X[is.na(x),] <- NA
  dimnames(X) <- list(names(x), paste(xname, levels(x), sep = ":"))
  return(X[,-1,drop=FALSE])
}
```

```{r enrichmentPlot_supp}

enrichmentPlot <- 
  function (pathway, stats, gseaParam = 1, ticksSize = 0.2)
{   rnk <- rank(-stats)
    ord <- order(rnk)
    statsAdj <- stats[ord]
    statsAdj <- sign(statsAdj) * (abs(statsAdj)^gseaParam)
    statsAdj <- statsAdj/max(abs(statsAdj))
    pathway <- unname(as.vector(na.omit(match(pathway, names(statsAdj)))))
    pathway <- sort(pathway)
    gseaRes <- calcGseaStat(statsAdj, selectedStats = pathway, returnAllExtremes = TRUE)
    bottoms <- gseaRes$bottoms
    tops <- gseaRes$tops
    n <- length(statsAdj)
    xs <- as.vector(rbind(pathway - 1, pathway))
    ys <- as.vector(rbind(bottoms, tops))
    toPlot <- data.frame(x = c(0, xs, n + 1), y = c(0, ys, 0))
    diff <- (max(tops) - min(bottoms))/8
    x = y = NULL
    g <- ggplot(toPlot, aes(x = x, y = y)) +
         geom_point(color = colors[1], size = 0.1) +
         geom_hline(yintercept = max(tops), colour = palreds[7], linetype = "dashed") +
         geom_hline(yintercept = min(bottoms), colour = palreds[7], linetype = "dashed") +
         geom_hline(yintercept = 0, colour = "black") +
         geom_line(color = colors[1]) +
         geom_segment(data = data.frame(x = pathway), mapping = aes(x = x, y = -diff/2, xend = x, yend = diff/2), 
                      size = ticksSize) +
         t2 +
         labs(x = "Rank", y = "Enrich. score")
    g}


```


```{r runGlm_supp}

#Function for multi-variant regression
runGlm <- function(X, y, method = "lasso", repeats = 20, folds = 3) {
  #set up objects to store results
  modelList <- list()
  lambdaList <- c()
  varExplain <- c()
  coefMat <- matrix(NA, ncol(X), repeats)
  rownames(coefMat) <- colnames(X)

  #set alpha
  if (method == "lasso"){
    alpha = 1
  } else if (method == "ridge") {
    alpha = 0
  }
  
  #for the set number of repeats, run the regression
  for (i in seq(repeats)) {
    if (ncol(X) > 2) {
      #run cross validated generalised linear model with given parameters
      res <- cv.glmnet(X,y, type.measure = "mse", family="gaussian", 
                       nfolds = folds, alpha = alpha, standardize = FALSE)
      
      #store lamdas and min lambda value
      lambdaList <- c(lambdaList, res$lambda.min)
      
      #store result of cv.glmnet
      modelList[[i]] <- res
      
      #get coefficents with min lambda value 
      coefModel <- coef(res, s = "lambda.min")[-1] #remove intercept row
      
      #store coefficients for this repeat
      coefMat[,i] <- coefModel
      
      #calculate variance explained
      if(sum(coefModel !=0)){
      y.pred <- predict(res, s = "lambda.min", newx = X)
      #if there are no predictors, all y.pred will be the same so its not possible to calculate variance explained because the SD is 0
      varExp <- cor(as.vector(y),as.vector(y.pred))^2
      }else{ varExp <- NA}
      varExplain[i] <- ifelse(is.na(varExp), 0, varExp) 
      
     
      
    } else {
      fitlm<-lm(y~., data.frame(X))
      varExp <- summary(fitlm)$r.squared
      varExplain <- c(varExplain, varExp)
      
    }

  }
  #store all results 
  list(modelList = modelList, lambdaList = lambdaList, varExplain = varExplain, coefMat = coefMat)
}
```


```{r lassoPlot_supp}

lassoPlot <- function(lassoOut, geneMatrix, betaMatrix, freqCut = 1, coefCut = 0.01) {
  #object to hold all plots
  plotList <- list()
  
  #for each drug - stimuli combination, run the following:
  for (seaName in names(lassoOut)) {
    ###FOR THE BAR PLOT
    #extract mean coefficients for each drug - stimuli combination
    barValue <- rowMeans(lassoOut[[seaName]]$coefMat)
    #extract proportion of repeats for which each coefficient is significant
    freqValue <- rowMeans(abs(sign(lassoOut[[seaName]]$coefMat)))
    #filter out coefficients that don't meet freqCut and coefCut thresholds
    barValue <- barValue[abs(barValue) >= coefCut & freqValue >= freqCut] 
    #arrange the bar values in numerical order
    barValue <- barValue[order(barValue)]
    #if there are no sig coefficients, don't plot
    if(length(barValue) == 0) {
      plotList[[seaName]] <- NA
      next
    }
    
   
    ###FOR THE HEATMAP AND SCATTER PLOT
    #get feature matrix and response matrix to plot
    allData <- geneMatrix
    betaValue <- unlist(betaMatrix[seaName,])
    
    #get feature matrix for features with significant coefficients only
    tabValue <- allData[, names(barValue),drop=FALSE]
    ord <- order(betaValue)
    betaValue <- betaValue[ord]
    tabValue <- tabValue[ord, ,drop=FALSE]
    sampleIDs <- rownames(tabValue)
    tabValue <- as_tibble(tabValue)
    tabValue$Sample <- sampleIDs
    
    #annotate features as mutations, methylation cluster or IGHV, and apply different scaling     so that different colours can be used in plotting 
    
    #for mutations:
    matValue <- gather(tabValue, key = "Var",value = "Value", -Sample)
    matValue$Type <- "mut"
    
    #for methylation cluster
    matValue$Type[grep("Methylation",matValue$Var)] <- "meth"
    
    #for IGHV status
    matValue$Type[grep("IGHV.status",matValue$Var)] <- "ighv"
    
    #change the scale of the value so that IGHV, Methylation and Mutation do not overlap
    matValue[matValue$Type == "mut",]$Value = matValue[matValue$Type == "mut",]$Value + 10
    matValue[matValue$Type == "meth",]$Value = matValue[matValue$Type == "meth",]$Value + 20
    matValue[matValue$Type == "ighv",]$Value = matValue[matValue$Type == "ighv",]$Value + 30
    
    #change continous to catagorical
    matValue$Value <- factor(matValue$Value,levels = sort(unique(matValue$Value)))
    
    #arrange order of feature rows and columns in heatmap
    #heatmap rows should align with order of genetic coefficients
    matValue$Var <- factor(matValue$Var, levels = names(barValue))
    
    #sample columns should be ascending order according to value of coefficient of each patient, so that heatmap aligns with scatter plot below
    matValue$Sample <- factor(matValue$Sample, levels = names(betaValue))
   
    
    
    #MAKE PLOTS
    #plot the heatmap of genetic feature values
    p1 <- ggplot(matValue, aes(x=Sample, y=Var)) + 
      geom_tile(aes(fill=Value), color = "white") + #ghost white
      theme_bw()+
      scale_y_discrete(expand=c(0,0)) + 
      theme(axis.title.x = element_text( size=fontsize+4),
            axis.text.y=element_text(hjust=0, size=18, face="bold"), 
            axis.ticks=element_blank(),
            panel.border=element_rect(colour="gainsboro"),  
            plot.title=element_text(face="bold", size = 18, margin = margin(t = -5, b = 1)), 
            panel.background=element_blank(),
            panel.grid.major=element_blank(), 
            panel.grid.minor=element_blank()) + 
      xlab("Mutation status for each patient") + 
      ylab("") + 
      scale_fill_manual(name="Mutated", 
                        values=c(`10`= offwhite,  #WT
                                 `11`=palreds[7], #Mutant
                                 `20`= offwhite, #LP
                                 `20.5`=palblues[6], #IP
                                 `21` = palblues[3], #HP
                                 `30` = offwhite, #IGHV-U
                                 `31` = palblues[6]), #IGHV-M
                        guide=FALSE) + 
            ggtitle(seaName)
    
    
    #Plot the bar plot on the left of the heatmap 
    barDF = data.frame(barValue, nm=factor(names(barValue),levels=names(barValue)))
    
    p2 <- ggplot(data=barDF, aes(x=nm, y=barValue)) + 
      geom_bar(stat="identity", 
               fill=ifelse(barDF$nm =="Methylation_Cluster"|barDF$nm =="IGHV.status",palblues[6],palreds[8]), 
               colour="black", 
               # intercept=0, 
               size=0.3)+
      scale_x_discrete(expand=c(0,0.5))+ 
      scale_y_continuous(expand=c(0,0), n.breaks = 4)+ 
      coord_flip(ylim=c(-0.3,0.35))+ 
      theme(panel.grid.major=element_blank(), 
            panel.background=element_blank(), 
            axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), 
            axis.text=element_text(size=fontsize, angle = 0, hjust = 0),
                axis.title = element_text(size=fontsize+4), 
            panel.border=element_blank()) +
      ylab("Size of coefficient") + 
      geom_vline(xintercept=c(0.5), 
                 color="black", 
                 size=0.6)
    
    #Plot the scatter plot of patient coefficient values under the heatmap
    scatterDF = data.frame(X=factor(names(betaValue), 
                                    levels=names(betaValue)), 
                           Y=unlist(betaValue))
    
    p3 <- ggplot(scatterDF, aes(x=X, y=Y)) + 
          geom_point(shape=21, 
                     fill="dimgrey", 
                     colour="#707372", #dark grey
                     size=1.2) + 
          theme_bw() +
          theme(panel.grid.minor=element_blank(), 
                panel.grid.major.x=element_blank(), 
                axis.ticks.x=element_blank(), 
                axis.text.y=element_text(size=fontsize),
                axis.title = element_text(size=fontsize+4),
                panel.border=element_rect(colour="dimgrey", size=0.1),
                panel.background=element_rect(fill="white")) +
  xlab(expression(paste("Patient-specific ", beta["int"])))
    
    
    #Assemble all the plots together

    # construct the gtable
    wdths = c(0.2, 1.5, 0.4, 1.3*ncol(matValue), 1.7, 0.2)
    hghts = c(0.3, 0.3, 0.0020*nrow(matValue), 0.2, 0.8, 0.3)*1.5
    gt = gtable(widths=unit(wdths, "in"), heights=unit(hghts, "in"))
    
    ## make grobs
    gg1 = ggplotGrob(p1)
    gg2 = ggplotGrob(p2)
    gg3 = ggplotGrob(p3)

    ## fill in the gtable
   
    #HEATMAP
    #5:1 = "PREDICTORS"
    gt = gtable_add_grob(gt, gtable_filter(gg1, "panel"), 3, 4) # add heatmap
    gt = gtable_add_grob(gt, gtable_filter(gg1, "panel"), 3, 4) #add legend
    gt = gtable_add_grob(gt, gtable_filter(gg1, "title"), 1, 4) #add title to plot
    gt = gtable_add_grob(gt, gtable_filter(gg1, "axis-l"), 3, 5) # variable names
    gt = gtable_add_grob(gt, gtable_filter(gg1, "xlab-b"), 2, 4) # axis title
    
    #BARPLOT
    gt = gtable_add_grob(gt, gtable_filter(gg2, "panel"), 3, 2) # add barplot
    gt = gtable_add_grob(gt, gtable_filter(gg2, "axis-b"), 4, 2) # y axis for barplot
    gt = gtable_add_grob(gt, gtable_filter(gg2, "xlab-b"), 2, 2) # y lab for barplot

    
    #SCATTER PLOT
    gt = gtable_add_grob(gt, gtable_filter(gg3, "panel"), 5, 4) # add scatterplot
    gt = gtable_add_grob(gt, gtable_filter(gg3, "xlab-b"), 6, 4) # x label for scatter plot
    gt = gtable_add_grob(gt, gtable_filter(gg3, "axis-l"), 5, 3) #  axis for scatter plot
    
   

    
    #plot
    plotList[[seaName]] <- gt
  }
  return(plotList)
}

```


```{r lassoPlot3Appendix}


lassoPlot3Appendix <- function(lassoOut, geneMatrix, viabMatrix, freqCut = 1, coefCut = 0.01) {
  
  plotList <- list()
  
  for (seaName in names(lassoOut)) { 
    ###FOR THE BAR PLOT
    #extract coefMat for each stimulus
    barValue <- rowMeans(lassoOut[[seaName]]$coefMat)
    #extract proportion of repeats for which each coefficient is significant
    freqValue <- rowMeans(abs(sign(lassoOut[[seaName]]$coefMat)))
    #filter out coefficients that don't meet freqCut and coefCut thresholds
    barValue <- barValue[abs(barValue) >= coefCut & freqValue >= freqCut] 
    #arrange the bar values in numerical order
    barValue <- barValue[order(barValue)]
    #if there are no sig coefficients, don't plot
    if(length(barValue) == 0) {
      plotList[[seaName]] <- NA
      next
    }

    ###FOR THE HEATMAP AND SCATTER PLOT
    #get feature matrix and response matrix to plot
    allData <- geneMatrix
    viabValue <- unlist(viabMatrix[seaName,])
    
    #get feature matrix for sig coefficients only
    tabValue <- allData[, names(barValue),drop=FALSE]
    ord <- order(viabValue)
    viabValue <- viabValue[ord]
    tabValue <- tabValue[ord, ,drop=FALSE]
    sampleIDs <- rownames(tabValue)
    tabValue <- as_tibble(tabValue)
    tabValue$Sample <- sampleIDs
    
    
    #annotate mutations by mutation, methylation or IGHV
    matValue <- gather(tabValue, key = "Var",value = "Value", -Sample)
    matValue$Type <- "mut"
    
    #for Methylation Cluster
    matValue$Type[grep("Methylation",matValue$Var)] <- "meth"
    
    #for IGHV status
    matValue$Type[grep("IGHV",matValue$Var)] <- "ighv"
    
    #change the scale of the value so that IGHV, Methylation and Mutation do not overlap
    matValue[matValue$Type == "mut",]$Value = matValue[matValue$Type == "mut",]$Value + 10
    matValue[matValue$Type == "meth",]$Value = matValue[matValue$Type == "meth",]$Value + 20
    matValue[matValue$Type == "ighv",]$Value = matValue[matValue$Type == "ighv",]$Value + 30
    
    #change continuous to categorical
    matValue$Value <- factor(matValue$Value,levels = sort(unique(matValue$Value)))
    
    #arrange order of heatmap
    matValue$Var <- factor(matValue$Var, levels = names(barValue))
    matValue$Sample <- factor(matValue$Sample, levels = names(viabValue))
    
    #change labels if mutation is a Doehner mutation 
    matValue$Var <- plyr::revalue(matValue$Var, c("del11q" = "del(11q)", "del13q" = "del(13q)", "del17p" = "del(17p)", "trisomy12" = "trisomy 12")) 
    

   
     #plot the heatmap 
      #title
    if(seaName=="TGF-b1"){
      thetitle <- "TGF-\u03B21"     
      } else {
        if(seaName=='IL-4'){ 
          thetitle <- "IL4"  
          } else {
            thetitle <- seaName
          }
        }
    
      p1 <- ggplot(matValue, aes(x=Sample, y=Var)) + 
            geom_tile(aes(fill=Value), color = "white") + #ghost white
            theme_bw()+
            scale_y_discrete(expand=c(0,0)) + 
            theme(axis.title.y = element_text( size=14),
                  axis.title.x = element_text( size=14),
                  axis.text.x=element_text(hjust=0, size=11),
                  axis.text.y=element_text(hjust=0.1, size=11),
                  axis.ticks=element_blank(),
                  panel.border=element_rect(colour="gainsboro"),  
                  plot.title=element_text(face="bold", size = 18, margin = margin(t = -5, b = 1)), 
                  panel.background=element_blank(),
                  panel.grid.major=element_blank(), 
                  panel.grid.minor=element_blank()) + 
            xlab("Mutation status for each patient") + 
            ylab("") + 
            scale_fill_manual(name="Mutated", 
                              values=c(`10`= offwhite,  #WT
                                       `11`="#373A36", #Mutant
                                       `20`= offwhite, #LP
                                       `20.5`= "#707372", #IP
                                       `21` = "#A8A99E", #HP
                                       `30` = offwhite, #IGHV-U
                                       `31` = "#707372"), #IGHV-M
                                       guide="none") + 
            ggtitle(thetitle)
        

         
    #Plot the bar plot on the left of the heatmap 
    barDF = data.frame(barValue, nm=factor(names(barValue),levels=names(barValue)))
    
    p2 <- ggplot(data=barDF, aes(x=nm, y=barValue)) + 
      geom_bar(stat="identity", 
               fill=ifelse(barValue<0,
                           palblues[6],palreds[8]), 
               colour="black", 
               size=0.3) + 
      scale_x_discrete(expand=c(0,0.5)) + 
      scale_y_continuous(expand=c(0,0)) + 
      coord_flip(ylim=c(-0.3,0.35)) + #changed from min(barValue) and max(barValue)
      theme(panel.grid.major=element_blank(), 
            panel.background=element_blank(), 
            axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), 
            axis.text=element_text(size=11, angle = 45, hjust = 1, vjust = 1),
                axis.title = element_text(size=14), 
            panel.border=element_blank()) +
      ylab("Size of predictor") + 
      geom_vline(xintercept=c(0.5), 
                 color="black", 
                 size=0.6)
    
    #Plot the scatter plot under the heatmap
    scatterDF = data.frame(X=factor(names(viabValue), 
                                    levels=names(viabValue)), 
                           Y=unlist(viabValue))
    
    p3 <- 
      ggplot(scatterDF, aes(x=X, y=Y)) + 
      geom_point(shape=21, 
                     fill="dimgrey", 
                     colour="#707372", #dark grey
                     size=1.2) + 
      theme_bw() +
      theme(panel.grid.minor=element_blank(), 
                panel.grid.major.x=element_blank(), 
                #axis.title.x=element_blank(), 
                axis.ticks.x=element_blank(), 
                axis.text.y=element_text(size=11), 
                axis.title.x=element_text(size=14), 
                panel.border=element_rect(colour="dimgrey", size=0.1),
                panel.background=element_rect(fill="white")) +
      xlab("Log(Viability) for each sample")
    
    
    #Assemble all the plots togehter

    # construct the gtable
    wdths = c(0.2, 1.5, 0.2, 1.3*ncol(matValue), 1.5, 0.2)
    hghts = c(0.2, 0.2, 0.0020*nrow(matValue), 0.3, 1, 0.2)
    gt = gtable(widths=unit(wdths, "in"), heights=unit(hghts, "in"))
    
    ## make grobs
    gg1 = ggplotGrob(p1)
    gg2 = ggplotGrob(p2)
    gg3 = ggplotGrob(p3)

    ## fill in the gtable
   
    #HEATMAP
    #5:1 = "PREDICTORS"
    gt = gtable_add_grob(gt, gtable_filter(gg1, "panel"), 3, 4) # add heatmap
    gt = gtable_add_grob(gt, gtable_filter(gg1, "panel"), 3, 4) #add legend
    gt = gtable_add_grob(gt, gtable_filter(gg1, "title"), 1, 4) #add title to plot
    gt = gtable_add_grob(gt, gtable_filter(gg1, "axis-l"), 3, 5) # variable names
    gt = gtable_add_grob(gt, gtable_filter(gg1, "xlab-b"), 2, 4) # axis title
    
    #BARPLOT
    gt = gtable_add_grob(gt, gtable_filter(gg2, "panel"), 3, 2) # add barplot
    gt = gtable_add_grob(gt, gtable_filter(gg2, "axis-b"), 4, 2) # y axis for barplot
    gt = gtable_add_grob(gt, gtable_filter(gg2, "xlab-b"), 2, 2) # y lab for barplot

    
    #SCATTER PLOT
    gt = gtable_add_grob(gt, gtable_filter(gg3, "panel"), 5, 4) # add scatterplot
    gt = gtable_add_grob(gt, gtable_filter(gg3, "xlab-b"), 6, 4) # x label for scatter plot
    gt = gtable_add_grob(gt, gtable_filter(gg3, "axis-l"), 5, 3) #  axis for scatter plot
    
   

    
    #plot
    #grid.draw(gt)
    plotList[[seaName]] <- gt
  }
  return(plotList)
}


```

```{r lassoPlot_CLLPD}


 lassoPlot_CLLPD<- function(lassoOut, viabMatrix, PDMatrix, freqCut = 1, coefCut = 0.01) {

    
    ###FOR THE BAR PLOT
    #extract coefMat for each stimulus
    barValue <- rowMeans(lassoOut$coefMat)
    #extract proportion of repeats for which each coefficient is significant
    freqValue <- rowMeans(abs(sign(lassoOut$coefMat)))
    #filter out coefficients that don't meet freqCut and coefCut thresholds
    barValue <- barValue[abs(barValue) >= coefCut & freqValue >= freqCut] 
    #arrange the bar values in numerical order
    barValue <- barValue[order(barValue)]
    #if there are no sig coefficients, don't plot
    if(length(barValue) == 0) {
      
      next
    }

    ###FOR THE HEATMAP AND SCATTER PLOT
    #get feature matrix and response matrix to plot
    allData <- viabMatrix
    PDValue <- PDMatrix
    
    #get feature matrix for sig coefficients only
    tabValue <- allData[, names(barValue),drop=FALSE]
    ord <- order(PDValue)
    PDValue <- PDValue[ord]
    tabValue <- tabValue[ord, ,drop=FALSE]
    sampleIDs <- rownames(tabValue)
    tabValue <- as_tibble(tabValue)
    tabValue$Sample <- sampleIDs
    
    
    
    matValue <- gather(tabValue, key = "Var",value = "Value", -Sample)
    
    
    #arrange order of heatmap
    matValue$Var <- factor(matValue$Var, levels = names(barValue))
    matValue$Sample <- factor(matValue$Sample, levels = names(PDValue))

   
     #plot the heatmap 
    
      p1 <- ggplot(matValue, aes(x=Sample, y=Var)) + 
            geom_tile(aes(fill=Value), color = "white") + #ghost white
            theme_bw()+
            scale_y_discrete(expand=c(0,0)) + 
            theme(axis.title.y = element_text( size=14),
                  axis.title.x = element_text( size=14),
                  axis.text.x=element_text(hjust=0, size=11),
                  axis.text.y=element_text(hjust=0.1, size=11),
                  axis.ticks=element_blank(),
                  panel.border=element_rect(colour="gainsboro"),  
                  plot.title=element_text(face="bold", size = 18, margin = margin(t = -5, b = 1)), 
                  panel.background=element_blank(),
                  panel.grid.major=element_blank(), 
                  panel.grid.minor=element_blank()) + 
            xlab("Viability z score") + 
            ylab("") +
            scale_fill_gradient2(low = palblues[1],high = palreds[8], mid = "white",guide="none")
            
        

         
    #Plot the bar plot on the left of the heatmap 
    barDF = data.frame(barValue, nm = factor(names(barValue),levels=names(barValue)))
    
    p2 <- ggplot(data=barDF, aes(x = nm, y = barValue)) + 
      geom_bar(stat = "identity", 
               fill = ifelse(barValue<0,
                           palblues[6], palreds[8]), 
               colour = "black", 
               size=0.3) + 
      scale_x_discrete(expand = c(0, 0.5)) + 
      scale_y_continuous(expand = c(0, 0)) + 
      coord_flip(ylim = c(-0.3, 0.35)) + #changed from min(barValue) and max(barValue)
      theme(panel.grid.major = element_blank(), 
            panel.background = element_blank(), 
            axis.ticks.y = element_blank(),
            panel.grid.minor = element_blank(), 
            axis.text=element_text(size = 11, angle = 45, hjust = 1, vjust = 1),
                axis.title = element_text(size = 14), 
            panel.border = element_blank()) +
      ylab("Size of predictor") + 
      geom_vline(xintercept = c(0.5), 
                 color = "black", 
                 size = 0.6)
    
    #Plot the scatter plot under the heatmap
    
    scatterDF = data.frame(X=factor(names(PDValue), 
                                    levels=names(PDValue)), 
                           Y=unlist(PDValue))
    
    p3 <- 
      ggplot(scatterDF, aes(x=X, y=Y)) + 
      geom_point(shape=21, 
                     fill="dimgrey", 
                     colour="#707372", #dark grey
                     size=1.2) + 
      theme_bw() +
      theme(panel.grid.minor=element_blank(), 
                panel.grid.major.x=element_blank(), 
                #axis.title.x=element_blank(), 
                axis.ticks.x=element_blank(), 
                axis.text.y=element_text(size=11), 
                axis.title.x=element_text(size=14), 
                panel.border=element_rect(colour="dimgrey", size=0.1),
                panel.background=element_rect(fill="white")) +
      xlab("CLL PD for each sample")
    
    
    #Assemble all the plots togehter

    # construct the gtable
    wdths = c(0.2, 1.5, 0.2, 1.3*ncol(matValue), 1.5, 0.2)
    hghts = c(0.2, 0.2, 0.0020*nrow(matValue), 0.3, 1, 0.2)
    gt = gtable(widths=unit(wdths, "in"), heights=unit(hghts, "in"))
    
    ## make grobs
    gg1 = ggplotGrob(p1)
    gg2 = ggplotGrob(p2)
    gg3 = ggplotGrob(p3)

    ## fill in the gtable
   
    #HEATMAP
    #5:1 = "PREDICTORS"
    gt = gtable_add_grob(gt, gtable_filter(gg1, "panel"), 3, 4) # add heatmap
    gt = gtable_add_grob(gt, gtable_filter(gg1, "panel"), 3, 4) #add legend
    gt = gtable_add_grob(gt, gtable_filter(gg1, "title"), 1, 4) #add title to plot
    gt = gtable_add_grob(gt, gtable_filter(gg1, "axis-l"), 3, 5) # variable names
    gt = gtable_add_grob(gt, gtable_filter(gg1, "xlab-b"), 2, 4) # axis title
    
    #BARPLOT
    gt = gtable_add_grob(gt, gtable_filter(gg2, "panel"), 3, 2) # add barplot
    gt = gtable_add_grob(gt, gtable_filter(gg2, "axis-b"), 4, 2) # y axis for barplot
    gt = gtable_add_grob(gt, gtable_filter(gg2, "xlab-b"), 2, 2) # y lab for barplot

    
    #SCATTER PLOT
    gt = gtable_add_grob(gt, gtable_filter(gg3, "panel"), 5, 4) # add scatterplot
    gt = gtable_add_grob(gt, gtable_filter(gg3, "xlab-b"), 6, 4) # x label for scatter plot
    gt = gtable_add_grob(gt, gtable_filter(gg3, "axis-l"), 5, 3) #  axis for scatter plot
    
   
  # grid.draw(gt)
    return(gt)
 
}


```

```{r makelegends_supp}

makelegends <- function (legendFor, colors) 
{
    x = NULL
    y = NULL
    colors = colors[names(colors) %in% legendFor]
    nleg = length(colors)
    
    #edit these widths to change alignment with lasso plots in patchwork code
    wdths = c(0.4,2,2,2,1.5)
    hghts = c(2)
    
    gtl = gtable(widths=unit(wdths, "in"), heights=unit(hghts, "in"))
    n = 2
    if ("M" %in% names(colors)) {
        Mgg = ggplot(data = data.frame(x = 1, 
                                       y = factor(c("LP", "IP", "HP"), 
                                                  levels = c("LP", "IP", "HP"))), 
                     aes(x = x, y = y, fill = y)) + 
              geom_tile() + 
              scale_fill_manual(name = "Methylation cluster", 
                                values = setNames(colors[["M"]], 
                                                  nm = c("LP", "IP","HP"))) + 
              theme(legend.title = element_text(size = 12), 
                    legend.text = element_text(size = 12))
        
        gtl = gtable_add_grob(gtl, gtable_filter(ggplotGrob(Mgg), "guide-box"), 1, n)
        n = n + 1
    }
    
    if ("I" %in% names(colors)) {
        Igg = ggplot(data = data.frame(x = 1, y = factor(c("Unmutated", 
            "Mutated"), levels = c("Unmutated", "Mutated"))), 
            aes(x = x, y = y, fill = y)) + geom_tile() + scale_fill_manual(name = "IGHV", 
            values = setNames(colors[["I"]], nm = c("Unmutated", "Mutated"))) + 
            theme(legend.title = element_text(size = 12), 
                  legend.text = element_text(size = 12))
        gtl = gtable_add_grob(gtl, gtable_filter(ggplotGrob(Igg), 
            "guide-box"), 1, n)
        n = n + 1
    }
    
    if ("G" %in% names(colors)) {
        Ggg = ggplot(data = data.frame(x = 1, y = factor(c("Wild Type", 
            "Mutated"), levels = c("Wild Type", "Mutated"))), 
            aes(x = x, y = y, fill = y)) + geom_tile() + scale_fill_manual(name = "Gene", 
            values = setNames(colors[["G"]], nm = c("Wild Type", "Mutated"))) + 
            theme(legend.title = element_text(size = 12), 
                  legend.text = element_text(size = 12))
        gtl = gtable_add_grob(gtl, gtable_filter(ggplotGrob(Ggg), 
            "guide-box"), 1, n)
        n = n + 1
    }
    
    return(list(plot = gtl, width = sum(wdths), height = sum(hghts)))
}


```

```{r Scale viability matrix}
########### Scaling and Centering of Viability Matrix ################

# medianCenter_MadScale: 
#function to  scale with MAD, center to merdian
medianCenter_MadScale <- function(x) {
  s <- median(x)
  #s=0
  (x - s) / mad(x, center = s)
}

# scaleCytResp function:  to apply medianCenter_MadScale row wise to viability matrix
scaleCytResp  <- function(x) t(apply(x, 1, medianCenter_MadScale)) 

```

```{r sortchrName}

#sort chromosomes by number 
sortChrName <- function(chr.name) {
    ## X, Y and M will cause warnings when change to number.
    noChr <- suppressWarnings(as.numeric(sub("chr", "", chr.name)))
    ## index of chromosome name are character, such as X, Y
    ch.idx <- which(is.na(noChr))
    
    n.idx <- which(!is.na(noChr))
    chr.n <- noChr[n.idx]

    chr.sorted <- chr.name[n.idx][order(chr.n)]
    if (length(ch.idx) != 0) {
        chr.sorted <- c(chr.sorted, sort(chr.name[ch.idx]))
    }
         
    return(chr.sorted)
}

```


<!-- Define Aesthetics  -->
```{r defineAesthetics_supp}

source("../../R/themes_colors.R")

```

<!-- Load data  -->
```{r loadData_supp, message=FALSE, warning=FALSE}

#drugs: dataframe containing meta data on all drugs used in screen
load( "../../data/drugs.RData")

#cytokines: dataframe containing meta data on all stimuli used in screen
load( "../../data/cytokines.RData")

#df: tibble containing all screening viability data
load( "../../data/df.RData")

#Heatmap_clusters: tibble containing patient cluster assignments
load( "../../data/Heatmap_clusters.RData")

#IbrStat6iComb.RData: tibble containing  Ibrutinib, IL4 and AS1517499 treated CLL PBMCs
load( "../../data/IbrStat6iComb.RData")

#cytReceptors: tibble containing meta data on cytokine receptors
load( "../../data/cytReceptors.RData")

#cyt_and_receptors: tibble containing meta data on cytokines and receptors
load( "../../data/cyt_and_receptors.RData")

#dds_smp: DeseqDataset containing patient gene expression data for matched samples to those in df
load( "../../data/dds_smp.RData")

#diffTF_small: tibble containing results from running diffTF package
load( "../../data/diffTF_small.RData")

#diffTF_wochr12: tibble containing results from running diffTF package without reads from chromosome 12
load( "../../data/diffTF_wochr12.RData")

#diffTF_wochr12_big: tibble containing results from running diffTF package without reads from chromosome 12 for the Rendeiro et al dataset. 
load( "../../data/diffTF_wochr12_big.RData")

#BIGCLL52.SPIB.output: tibble containing Lof2FCs and p values for each Spi-B binding site
load("../../data/BIGCLL52.SPIB.output.RData")

#ChIPseq data for SPIB and PU1 binding in OCILY3 cell line (downloaded from GEO)

#get info on dataset of interest
chipSeq <- getGEOInfo(genome="hg19", simplify=FALSE) %>% dplyr::filter(series_id=="GSE56857")

#Define gsm id for SPIB and PU1 ChIPseq dataset in OCILY3 cell line (DLBCL)
gsm_spib = "GSM1370276"
gsm_pu1 = "GSM1370275"                                                         

#download bed file if not already there                                          
downloadGSMbedFiles(gsm_spib, destDir="../../inst/extdata/")
downloadGSMbedFiles(gsm_pu1, destDir="../../inst/extdata/")

#read the data
#ChIPseq data downloaded from GEO Accession Number: GSE56857
peak.spib <-  genomation ::readBed("../../inst/extdata/GSM1370276_OCILY3_SPIB_GEM_events.bed.gz", track.line = "auto")
peak.pu1 <-  genomation ::readBed( "../../inst/extdata/GSM1370275_OCILY3_PU1_GEM_events.bed.gz", track.line = "auto")

#patMeta: tibble containing patient genetic data
load( "../../data/patMeta.RData")
patMeta$treatment <- as.factor(patMeta$treatment)

#survT: tibble containing patient clinical outcomes
load( "../../data/survT.RData")

#CLL PD 
load( "../../data/CLL_PD.RData")

```

<!-- Load data from tsv  -->
```{r loadData_fromtsv_supp, eval = FALSE}

#drugs: dataframe containing meta data on all drugs used in screen
drugs <- read.delim("../../inst/extdata/Drugs.txt")

#cytokines: dataframe containing meta data on all stimuli used in screen
cytokines <- read.delim("../../inst/extdata/Cytokines.txt")

#df: tibblecontaining all screening viability data
df <- read.table(file= "../../inst/extdata/df.txt",header = TRUE) %>% as_tibble()
df$Cytokine <- factor(df$Cytokine, levels= c("No Cytokine",
                                              "Resiquimod",
                                              "IL-4",
                                              "TGF-b1",
                                              "IL-1b",
                                              "Interferon gamma",
                                              "SDF-1a",
                                              "sCD40L" ,
                                              "sCD40L+IL-4",
                                              "soluble anti-IgM",
                                              "CpG ODN",
                                              "IL-6",
                                              "IL-10",
                                              "IL-21",
                                              "HS-5 CM",
                                              "IL-15",
                                              "BAFF",
                                              "IL-2" ))

df$Drug <- factor(df$Drug, levels =c("DMSO",
                                     "BAY-11-7085",
                                     "Everolimus",
                                     "Fludarabine",
                                     "I-BET 762",
                                     "Ibrutinib","Idelalisib",
                                     "Luminespib",
                                     "Nutlin-3a",
                                     "PRT062607",
                                     "Pyridone-6",
                                     "Ralimetinib",
                                     "Selumetinib"))

#Heatmap_clusters: tibble containing patient cluster assignments
Heatmap_clusters <- read.table(file= "../../inst/extdata/Heatmap_clusters.txt", header = TRUE) %>% as.tibble()

#cytReceptors: tibble containing meta data on cytokine receptors
cytReceptors <- read.table(file= "../../inst/extdata/cytReceptors.txt", header = TRUE) %>% as.tibble()

#cyt_and_receptors: tibble containing meta data on cytokines and receptors
cyt_and_receptors <- read.table(file= "../../inst/extdata/cyt_and_receptors.txt", header = TRUE) %>% as.tibble()

#dds_smp: DeseqDataset containing patient gene expression data for matched samples to those in df
rna_countsMatrix <- read.table(file=  "../../inst/extdata/rna_countsMatrix.txt", header = TRUE)
coldata_rna <- read.table(file="../../inst/extdata/coldata_rna.txt", header= TRUE)
rowdata_rna <- read.table(file=  "../../inst/extdata/rowdata_rna.txt", header= TRUE)

#assemble deseq object
rownames(coldata_rna) <- coldata_rna$PatientID
dds_smp <- DESeqDataSetFromMatrix(rna_countsMatrix, coldata_rna, design =~1)
rowData(dds_smp) <- rowdata_rna


#diffTF_small: tibble containing results from running diffTF package
diffTF_small <- read_tsv("../../inst/extdata/Tri12vsWT.Cons.summary.tsv.gz")


#diffTF_wochr12: tibble containing results from running diffTF package without reads from chromosome 12
diffTF_wochr12 <- read_tsv("../../inst/extdata/Tri12vsWT.wochr12.summary.tsv.gz")

#patMeta: tibble containing patient genetic data
patMeta <- read.table(file= "../../inst/extdata/patMeta.txt",header = TRUE) %>% as_tibble()
patMeta[sapply(patMeta, is.character)] <- lapply(patMeta[sapply(patMeta, is.character)], as.factor)
patMeta[sapply(patMeta, is.integer)] <- lapply(patMeta[sapply(patMeta, is.integer)], as.factor)
patMeta$PatientID <- as.character(patMeta$PatientID)
patMeta$treatment <- as.factor(patMeta$treatment)

#survT: tibble containing patient clinical outcomes
survT <- read.table(file= "../../inst/extdata/survT.txt",header = TRUE) %>% as_tibble()


```

<!-- Process Data  -->
```{r}

## Create dataframe with patMeta and Cluster Information
patMeta_Cluster <- left_join(Heatmap_clusters, patMeta, by = "PatientID")

#add label column to df
df <- df %>%
  dplyr::mutate(Cytokine = as.character(Cytokine)) %>%
  dplyr::mutate(Cytlabel = ifelse(Cytokine == "IL-2", "IL2",
                            ifelse(Cytokine == "IL-4", "IL4",
                                   ifelse(Cytokine == "IL-6", "IL6",
                                          ifelse(Cytokine == "IL-10", "IL10",

                                                        ifelse(Cytokine == "IL-21", "IL21",
                                                               ifelse(Cytokine == "sCD40L+IL-4","sCD40L + IL4",
                                                                      ifelse(Cytokine == "IL-15", "IL15",
                                                                             ifelse(Cytokine == "Interferon gamma","Interferon\u03B3",
                                                                                    ifelse(Cytokine == "SDF-1a","SDF-1\u03B1",
                                                                                           ifelse(Cytokine == "IL-1b","IL1\u03B2",
                                                                                                  ifelse(Cytokine == "TGF-b1","TGF\u03B21", Cytokine)))))))))))) %>%
  
  

  dplyr::mutate(Cytokine = factor(Cytokine))

#generate table of Cytokines and assosciated labels
Cytokine_labels <- df %>% filter(PatientID == "Pat_001", Drug == "DMSO") %>% select( Cytokine, Cytlabel)
colnames(Cytokine_labels) <- c("name", "Cytlabel")



```

# Appendix Tables

```{r, fig.pos="H"}

patMeta %>%
  dplyr::select(PatientID, gender, treatment, IGHV.status, `Methylation_Cluster`, del13q, del11q, trisomy12, del17p) %>%
  dplyr::rename(`Patient ID`=PatientID, `Sex`=gender, `Treated before`=treatment, `IGHV status`=IGHV.status, `Methylation Cluster`=`Methylation_Cluster`, `Del13q`=del13q, `Del11q`=del11q, `Trisomy 12`=trisomy12, `Del17p`=del17p) %>%
   kable( longtable = TRUE) %>%
  row_spec(0,bold=TRUE) %>%
    kable_styling(latex_options = c("hold_position", "repeat_header", "striped"), font_size=8)

```
## Appendix Table 1. Patient samples included in the study.
List of patient samples and selected characteristics. For a full list of characteristics see [online vignette](https://github.com/Huber-group-EMBL/CLLCytokineScreen2021).

\newpage


```{r, fig.pos="H"}

CNVs<-  patMeta %>% 
    select(PatientID, del10p:trisomy12) %>% 
    pivot_longer(cols=del10p:trisomy12, names_to = "Chr_Aberration", values_to = "Value") %>% 
    mutate(Value=as.numeric(levels(Value))[Value]) %>% 
    group_by(PatientID) %>% 
    summarise(CNVs=sum(Value)) %>% 
  mutate(Complex_Karyotype=as.factor(case_when(CNVs>2~1, CNVs<=2~0)),
         Complex_Karyotype5=as.factor(case_when(CNVs>4~1, CNVs<=4~0)),
         CNVs=NULL)  %>% 
  select(PatientID, Complex_Karyotype)

selected_mutations <- 
  patMeta %>% 
  mutate(Ras_Raf=as.factor(ifelse(BRAF==1|KRAS==1|NRAS==1, 1,0))) %>% 
  dplyr::select(-BRAF, -KRAS, -NRAS) %>% 
  
      ## Add Complex Caryotype column, remove single columns
  left_join(CNVs, by = "PatientID") %>% 
  
  dplyr::select( -gender, 
                 -diagnosis, 
                 ) %>%
  pivot_longer(-PatientID, 
               names_to="Genetic_Alteration", 
               values_to = "alteration_value") %>% 
  dplyr::filter(alteration_value %in% c(1, "U")) %>% 
  dplyr::group_by(Genetic_Alteration) %>% 
  dplyr::count() %>% 
  dplyr::filter(n>2) %>% 
  dplyr::select(Genetic_Alteration) %>% 
  unlist()

 Table_of_alterations<-patMeta %>%
  
  ## Add Ras/Raf column, remove single columns
  mutate(Ras_Raf=as.factor(ifelse(BRAF==1|KRAS==1|NRAS==1, 1,0))) %>% 
  dplyr::select(-BRAF, -KRAS, -NRAS) %>% 
  mutate(IGHV.status=as.factor(case_when(IGHV.status=="U"~0, 
                               IGHV.status=="M"~1))) %>% 
   
       ## Add Complex Caryotype column, remove single columns
  left_join(CNVs, by = "PatientID") %>% 

## remove unused columns from metadata
  dplyr::select( -gender, 
                 -diagnosis, 
                 -treatment) %>% 
  
  ## transform data to long format
  pivot_longer(cols=c(-PatientID), 
               names_to = "Genetic_alt", 
               values_to="alt_value") %>% 
  
  group_by(Genetic_alt, alt_value) %>% 
  
  count() %>% 
  pivot_wider(names_from = alt_value, values_from = n) %>% 
   
   filter(Genetic_alt%in%selected_mutations) %>% 
   rename(Genetic_alteration=Genetic_alt, 'Wildtype/IGHV unmutated'="0", 'Mutated/ IGHV Mutated'="1", 'Not available'='NA')%>% #, "IGHV Mutated"="M", "IGHV Unmutated"="U"
   select_if(~sum(!is.na(.)) > 0)
   
  Table_of_alterations %>% 
      kable( longtable = TRUE) %>%
  row_spec(0,bold=TRUE) %>%
    kable_styling(latex_options = c("HOLD_position", "repeat_header", "striped"), font_size=8)

  
```
## Appendix Table 2. List of genetic alterations tested for their association with stimuli response in the univariate model of Figure 3A.
Only genetic alterations with at least 3 positive cases were considered. 
\newpage

```{r, fig.pos="H" }

clusters.surv<- left_join(Heatmap_clusters, survT, by = "PatientID")

#Add meta data to heatmap clusters
clusters.surv <- left_join(clusters.surv, patMeta, by="PatientID")

### TTT
# Multivariate Coxph Analysis
# Cluster 3 as reference group

#run coxphmodel 
c <- 
  clusters.surv %>%
  mutate(Cluster=factor(Cluster)) %>%
  coxph(Surv( TTT, treatedAfter) ~  factor2ind(Cluster, 3)+ IGHV.status + trisomy12 + TP53+Methylation_Cluster,  data=.)

#tidy results, exponentiate coefficients and estimate 95% confidence intervals 
cox_output<- tidy(c, conf.int = TRUE,  exponentiate = TRUE) %>%
  dplyr::select(term, estimate, p.value, conf.low, conf.high)

#tidy names
cox_output$term<-
  cox_output$term %>% 
  gsub(pattern="factor2ind", "", .)%>% 
  gsub(pattern="[()]", "", .)%>% 
  gsub(pattern=":", " ", .)%>% 
  gsub(pattern="Cluster, 3", "Cluster 3 vs ", .)%>% 
  gsub(pattern="IGHV.statusU", "IGHV.status", .)%>% 
  gsub(pattern="trisomy121", "trisomy 12", .)%>% 
  gsub(pattern="TP531", "TP53", .)

colnames(cox_output) <- c("Factor", "HR", "p.value", "CI.low", "CI.High")

cox_output$HR <- round(cox_output$HR, 2)
cox_output$p.value <- round(cox_output$p.value, 2)
cox_output$CI.low <- round(cox_output$CI.low, 2)
cox_output$CI.High <- round(cox_output$CI.High, 2)

cox_output <- cox_output %>% 
  dplyr::mutate(p.value=ifelse(p.value<0.0001,"<0.0001", p.value )) 


colnames(cox_output) <- c("Factor", "HR", "p value", "CI Low", "CI High")

cox_output %>%  
  kable() %>%
  kable_styling(latex_options = c("striped", "HOLD_position")) 

```

## Appendix Table 3. Multivariate survival analysis of response clusters.
Multivariate Cox proportional hazards model of stimuli response clusters and genetic subgroups of disease progression using TTT and C3 as reference.

```{r, fig.pos="H"}
######Arrange data######
clusters.surv <- left_join(Heatmap_clusters, survT, by = "PatientID")

#Add meta data to heatmap clusters
clusters.surv <- left_join(clusters.surv, patMeta, by="PatientID")


coxph <- clusters.surv%>%
  mutate(Cluster=factor(Cluster))%>%
    coxph(Surv( TTT, treatedAfter) ~  factor2ind(Cluster, 3), data=.) 

# cox_outputcoxph_summ$coefficients

#tidy results, exponentiate coefficients and estimate 95% confidence intervals 
cox_output<- tidy(coxph, conf.int = TRUE,  exponentiate = TRUE) %>%
  dplyr::select(term, estimate, p.value, conf.low, conf.high)

#tidy names
cox_output$term<-
  cox_output$term %>% 
  gsub(pattern="factor2ind", "", .)%>% 
  gsub(pattern="[()]", "", .)%>% 
  gsub(pattern=":", " ", .)%>% 
  gsub(pattern="Cluster, 3", "Cluster 3 vs ", .)%>% 
  gsub(pattern="Methylation_ClusterIP", "Methylation Cluster IP", .)%>% 
  gsub(pattern="Methylation_ClusterLP", "Methylation Cluster LP", .)

colnames(cox_output) <- c("Factor", "HR", "p.value", "CI.low", "CI.High")

cox_output$HR <- round(cox_output$HR, 2)
cox_output$p.value <- round(cox_output$p.value, 2)
cox_output$CI.low <- round(cox_output$CI.low, 2)
cox_output$CI.High <- round(cox_output$CI.High, 2)

cox_output <- cox_output %>% 
  dplyr::mutate(p.value=ifelse(p.value<0.0001,"<0.0001", p.value )) 


colnames(cox_output) <- c("Factor", "HR", "p value", "CI Low", "CI High")

cox_output %>%  
  kable() %>%
  kable_styling(latex_options = c("striped", "HOLD_position")) 

```

## Appendix Table 4. Survival analysis of microenvironmental response clusters comparing Cluster 3 to the other Clusters.



```{r, fig.pos="H"}
######Arrange data######
clusters.surv <- left_join(Heatmap_clusters, survT, by = "PatientID")

#Add meta data to heatmap clusters
clusters.surv <- left_join(clusters.surv, patMeta, by="PatientID")


coxph <- clusters.surv%>%
  filter(Cluster%in%c(1,2,3)) %>% 
  mutate(Cluster3vs12=ifelse(Cluster%in%c(1,2), "1_2", "3")) %>% 
  mutate(Cluster3vs12=factor(Cluster3vs12))%>%
    coxph(Surv( TTT, treatedAfter) ~  factor2ind(Cluster3vs12, "3"), data=.) 

# cox_outputcoxph_summ$coefficients

#tidy results, exponentiate coefficients and estimate 95% confidence intervals 
cox_output<- tidy(coxph, conf.int = TRUE,  exponentiate = TRUE) %>%
  dplyr::select(term, estimate, p.value, conf.low, conf.high)

#tidy names
cox_output$term<-
  cox_output$term %>% 
  gsub(pattern="factor2ind", "", .)%>% 
  gsub(pattern="[()]", "", .)%>% 
  gsub(pattern=":", " ", .)%>% 
  gsub(pattern="Cluster3vs12", "Cluster 3 vs combined Clusters 1 and 2 ", .)

colnames(cox_output) <- c("Factor", "HR", "p.value", "CI.low", "CI.High")

cox_output$HR <- round(cox_output$HR, 2)
cox_output$p.value <- round(cox_output$p.value, 2)
cox_output$CI.low <- round(cox_output$CI.low, 2)
cox_output$CI.High <- round(cox_output$CI.High, 2)

cox_output <- cox_output %>% 
  dplyr::mutate(p.value=ifelse(p.value<0.0001,"<0.0001", p.value )) 


colnames(cox_output) <- c("Factor", "HR", "p value", "CI Low", "CI High")

cox_output %>%  
  kable() %>%
  kable_styling(latex_options = c("striped", "HOLD_position")) 

```

## Appendix Table 5. Survival analysis of microenvironmental response clusters comparing Cluster 3 to the combined Cluster 1 and 2.
\blandscape
\newpage

```{r message=FALSE}

#filter for top peaks 

TFBS <- BIGCLL52_SPIB %>% dplyr::filter(abs(l2FC) >1, pval_adj <0.1)


#remove Chr
TFBS <- TFBS %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))

#Convert to a genomic ranges object
peak<- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)



#Annotate peaks    
#Here I use the annotatePeak function to find nearest gene to each of my SPIB binding sites, looking in the region Â±3kb around the tss sites defined in the TxDB object. The resulting `peakAnno` object contains all the original information on my SPIB binding sites, along with information on the annotation of these sites (eg promoter/Exon/intron/intergenic) and the nearest gene and its function. 


#get annotation database
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

#annoate all genes within 5kb
peakAnno <- annotatePeak(peak, 
                         tssRegion=c(-3000, 3000), #region in whcih to look for genes
                         TxDb=txdb, #define where to get transcript info
                         annoDb="org.Hs.eg.db", addFlankGeneInfo = TRUE, verbose = FALSE) #set annotation package to add ensembl, smbol and genames


```

```{r fig.width = 10}

#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test
kegg <-
  dplyr::filter(kegg, 
         #All KEGG immune signaling pathways
         gs_name %in% c("KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION", 
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY" ,
                        "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY"  , 
                        "KEGG_NOTCH_SIGNALING_PATHWAY", 
                        "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY", 
                        "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        #Control genesets
                        "KEGG_OXIDATIVE_PHOSPHORYLATION",                  
                        "KEGG_DNA_REPLICATION"
         ))

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )
kegg$term <- gsub("B CELL RECEPTOR SIGNALING PATHWAY", 
                  "BCR Signaling Pathway",kegg$term )
kegg$term <- gsub("CHEMOKINE SIGNALING PATHWAY", 
                  "Chemokine Signaling Pathway",kegg$term )
kegg$term <- gsub("CYTOKINE CYTOKINE RECEPTOR INTERACTION", 
                  "Cytokine Cytokine Receptor Interaction",kegg$term )
kegg$term <- gsub("JAK STAT SIGNALING PATHWAY", 
                  "JAK STAT signaling pathway",kegg$term )
kegg$term <- gsub("NOD LIKE RECEPTOR SIGNALING PATHWAY", 
                  "NLR Signaling Pathway",kegg$term )
kegg$term <- gsub("NOTCH SIGNALING PATHWAY", 
                  "Notch Signaling Pathway",kegg$term )
kegg$term <- gsub("RIG I LIKE RECEPTOR SIGNALING PATHWAY", 
                  "RLR Signaling Pathway",kegg$term )
kegg$term <- gsub("TOLL LIKE RECEPTOR SIGNALING PATHWAY", 
                  "TLR Signaling Pathway",kegg$term )

kegg$term <- gsub("T CELL RECEPTOR SIGNALING PATHWAY", 
                  "TCR Signaling Pathway",kegg$term )
kegg$term <- gsub("OXIDATIVE PHOSPHORYLATION", 
                  "Oxidative Phosphorylation",kegg$term )
kegg$term <- gsub("P53 SIGNALING PATHWAY", 
                  "P53 Signaling Pathway",kegg$term )
kegg$term <- gsub("DNA REPLICATION", 
                  "DNA Replication",kegg$term )

kegg$genesetDatabase <- "KEGG"


#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test
reactome <-
  dplyr::filter(reactome, 
         gs_name %in% c(#selected immune pathways
                        "REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX",
                        "REACTOME_TNF_SIGNALING",
                        "REACTOME_INTERLEUKIN_1_SIGNALING",
                        "REACTOME_INTERLEUKIN_10_SIGNALING",
                        "REACTOME_INTERLEUKIN_15_SIGNALING",
                        "REACTOME_INTERLEUKIN_2_SIGNALING",
                        "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
                        "REACTOME_INTERLEUKIN_6_SIGNALING" ,
                        "REACTOME_OTHER_INTERLEUKIN_SIGNALING",
                        "REACTOME_SIGNALING_BY_INTERLEUKINS",
                        "REACTOME_INTERFERON_ALPHA_BETA_SIGNALING" ,
                        "REACTOME_INTERFERON_GAMMA_SIGNALING", 
                        #Control genesets
                        "REACTOME_CELL_CYCLE",                  
                        "REACTOME_REGULATION_OF_TP53_ACTIVITY" 
                        ))

colnames(reactome) <- c("term", "gene")

#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )


#tidy up and put in lower case
reactome$term <- gsub("CELL CYCLE", 
                      "Cell Cycle",reactome$term )
reactome$term <- gsub("INTERLEUKIN 1 SIGNALING", 
                      "Interleukin 1 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 10 SIGNALING", 
                      "Interleukin 10 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 15 SIGNALING", 
                      "Interleukin 15 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 2 SIGNALING", 
                      "Interleukin 2 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 4 AND INTERLEUKIN 13 SIGNALING", 
                      "Interleukin 4 and Interleukin 13 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 6 SIGNALING",
                      "Interleukin 6 Signaling",reactome$term )
reactome$term <- gsub("OTHER INTERLEUKIN SIGNALING", 
                      "Other Interleukin Signaling",reactome$term )
reactome$term <- gsub("SIGNALING BY INTERLEUKINS", 
                      "Signaling by Interleukins",reactome$term )
reactome$term <- gsub("REGULATION OF TP53 ACTIVITY", 
                      "Regulation of TP53 Pathway",reactome$term )
reactome$term <- gsub("SIGNALING BY TGF BETA RECEPTOR COMPLEX",
                      # "Signaling by TGFbeta \nReceptor Complex", reactome$term )
                      "Signaling by TGFbeta Receptor Complex", reactome$term )
reactome$term <- gsub("TNF SIGNALING", 
                      "TNF Signaling",reactome$term )
reactome$term <- gsub("INTERFERON ALPHA BETA SIGNALING", 
                      "Interferon Alpha Beta Signaling",reactome$term )

reactome$term <- gsub("INTERFERON GAMMA SIGNALING", 
                      "Interferon Gamma Signaling", reactome$term )

#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"

```


```{r T6}

#generate TERM2GENE dataframe to use in over-representation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, p.adjust, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB", "adjpvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0
resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1
resTab$adjpvalue_SPIB[is.na(resTab$adjpvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB, adjpvalue_SPIB)

#adjust decimal places of p values
resTab$adjpvalue_SPIB <- round(resTab$adjpvalue_SPIB, digits = 3)
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, adjpvalue_SPIB<0.1)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno)$geneId)),")", sep= ""),
                      "SPIB\n p-value",
                      "SPIB\nadj. p-value")


resTab %>%  
  kable() %>%
  kable_styling(latex_options = c("striped", "HOLD_position", "scale_down")) 



```   

## Appendix Table 6. Over-representation test of immune signalling pathways (and control pathways) amongst Spi-B gene targets, based on primary CLL ATACseq data.

Spi-B gene targets defined as nearest gene to each Spi-B binding site. Spi-B binding sites subset from all Spi-B sites defined in the HOCOMOCO database, and subset for those that   that show an absolute Log2 fold change > 1 and a p value < 0.1 in trisomy 12 versus non-trisomy 12 CLL. Log2FC values and p values calculated using the diffTF package based on the CLL ATACseq data from Rendeiro et al. 2016^1^.

```{R, T7, fig.width=15, fig.height=2, message=FALSE }

# Get significant peaks
#filter ChipSeq data for peaks with significant score -log10(Qscore)>1.3
#if q < 0.05, then Qlog >1.3
spib.filtered <- peak.spib[(elementMetadata(peak.spib)[,1] > 1.3)]

# Annotate ChIP peaks
# Annotate peaks with meta data
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

peakAnno.spib <- annotatePeak(spib.filtered, 
                              tssRegion = c(-3000, 3000),
                              TxDb = txdb, 
                              annoDb = "org.Hs.eg.db",
                              verbose=FALSE)

peakAnno.spib.1kb <- as.data.frame(peakAnno.spib) #%>% dplyr::filter(abs(distanceToTSS) <1000)


#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)
#subset for pathways that we want to test
kegg <-
  dplyr::filter(kegg, 
         #All KEGG immune signaling pathways
         gs_name %in% c("KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION", 
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY" ,
                        "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY"  , 
                        "KEGG_NOTCH_SIGNALING_PATHWAY", 
                        "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY", 
                        "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        #Control genesets
                        "KEGG_OXIDATIVE_PHOSPHORYLATION",                  
                        "KEGG_DNA_REPLICATION"
         ))
colnames(kegg) <- c("term", "gene")
#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )
kegg$term <- gsub("B CELL RECEPTOR SIGNALING PATHWAY", 
                  "BCR Signaling Pathway",kegg$term )
kegg$term <- gsub("CHEMOKINE SIGNALING PATHWAY", 
                  "Chemokine Signaling Pathway",kegg$term )
kegg$term <- gsub("CYTOKINE CYTOKINE RECEPTOR INTERACTION", 
                  "Cytokine Cytokine Receptor Interaction",kegg$term )
kegg$term <- gsub("JAK STAT SIGNALING PATHWAY", 
                  "JAK STAT signaling pathway",kegg$term )
kegg$term <- gsub("NOD LIKE RECEPTOR SIGNALING PATHWAY", 
                  "NLR Signaling Pathway",kegg$term )
kegg$term <- gsub("NOTCH SIGNALING PATHWAY", 
                  "Notch Signaling Pathway",kegg$term )
kegg$term <- gsub("RIG I LIKE RECEPTOR SIGNALING PATHWAY", 
                  "RLR Signaling Pathway",kegg$term )
kegg$term <- gsub("TOLL LIKE RECEPTOR SIGNALING PATHWAY", 
                  "TLR Signaling Pathway",kegg$term )
kegg$term <- gsub("T CELL RECEPTOR SIGNALING PATHWAY", 
                  "TCR Signaling Pathway",kegg$term )
kegg$term <- gsub("OXIDATIVE PHOSPHORYLATION", 
                  "Oxidative Phosphorylation",kegg$term )
kegg$term <- gsub("P53 SIGNALING PATHWAY", 
                  "P53 Signaling Pathway",kegg$term )
kegg$term <- gsub("DNA REPLICATION", 
                  "DNA Replication",kegg$term )
kegg$genesetDatabase <- "KEGG"
#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)
#subset for pathways that we want to test
reactome <-
  dplyr::filter(reactome, 
         gs_name %in% c(#selected immune pathways
                        "REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX",
                        "REACTOME_TNF_SIGNALING",
                        "REACTOME_INTERLEUKIN_1_SIGNALING",
                        "REACTOME_INTERLEUKIN_10_SIGNALING",
                        "REACTOME_INTERLEUKIN_15_SIGNALING",
                        "REACTOME_INTERLEUKIN_2_SIGNALING",
                        "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
                        "REACTOME_INTERLEUKIN_6_SIGNALING" ,
                        "REACTOME_OTHER_INTERLEUKIN_SIGNALING",
                        "REACTOME_SIGNALING_BY_INTERLEUKINS",
                        "REACTOME_INTERFERON_ALPHA_BETA_SIGNALING" ,
                        "REACTOME_INTERFERON_GAMMA_SIGNALING", 
                        #Control genesets
                        "REACTOME_CELL_CYCLE",                  
                        "REACTOME_REGULATION_OF_TP53_ACTIVITY" 
                        ))
colnames(reactome) <- c("term", "gene")
#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )
#tidy up and put in lower case
reactome$term <- gsub("CELL CYCLE", 
                      "Cell Cycle",reactome$term )
reactome$term <- gsub("INTERLEUKIN 1 SIGNALING", 
                      "Interleukin 1 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 10 SIGNALING", 
                      "Interleukin 10 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 15 SIGNALING", 
                      "Interleukin 15 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 2 SIGNALING", 
                      "Interleukin 2 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 4 AND INTERLEUKIN 13 SIGNALING", 
                      "Interleukin 4 and Interleukin 13 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 6 SIGNALING",
                      "Interleukin 6 Signaling",reactome$term )
reactome$term <- gsub("OTHER INTERLEUKIN SIGNALING", 
                      "Other Interleukin Signaling",reactome$term )
reactome$term <- gsub("SIGNALING BY INTERLEUKINS", 
                      "Signaling by Interleukins",reactome$term )
reactome$term <- gsub("REGULATION OF TP53 ACTIVITY", 
                      "Regulation of TP53 Pathway",reactome$term )
reactome$term <- gsub("SIGNALING BY TGF BETA RECEPTOR COMPLEX",
                      "Signaling by TGFbeta \nReceptor Complex", reactome$term )
reactome$term <- gsub("TNF SIGNALING", 
                      "TNF Signaling",reactome$term )
reactome$term <- gsub("INTERFERON ALPHA BETA SIGNALING", 
                      "Interferon Alpha Beta Signaling",reactome$term )
reactome$term <- gsub("INTERFERON GAMMA SIGNALING", 
                      "Interferon Gamma Signaling", reactome$term )
#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"
#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)
#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")
##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene
spibRes <- enricher(unique(as.data.frame(peakAnno.spib.1kb)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)
#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)
resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, p.adjust, Count, genesetSize)
colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB", "adjpvalue_SPIB","count_SPIB", "genesetSize")
resTab <- resTab.spib
#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0
resTab$adjpvalue_SPIB[is.na(resTab$adjpvalue_SPIB)] <- 1
resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB, adjpvalue_SPIB)
#adjust decimal places of p values
resTab$adjpvalue_SPIB <- round(resTab$adjpvalue_SPIB, digits = 3)
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, adjpvalue_SPIB<0.1)
colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno.spib.1kb)$geneId)),")", sep= ""),
                      "SPIB\n p-value",
                      "SPIB\n adj. p-value")

resTab %>%  
  kable() %>%
  kable_styling(latex_options = c("striped", "HOLD_position", "scale_down")) 

```
## Appendix Table 7. Over-representation test of immune signalling pathways (and control pathways) amongst Spi-B gene targets, based on DLBCL ChIPseq data. 
Spi-B gene targets defined as nearest gene to each Spi-B binding site. Spi-B binding sites determined using ChIPseq data from Care et al 2014. 


\elandscape
\newpage

# Appendix Figures 

```{r S1, fig.path=plotDir, dev=c("cairo_pdf"), fig.width=10, fig.height=5, fig.pos="H"}

theme_temp<-theme(
  plot.background = element_blank(),
  panel.grid.major = element_blank(),
  panel.grid.major.x = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_blank(),
  panel.background = element_blank(),
  axis.line = element_line(size=.4),
  axis.line.x = element_line(),
  axis.line.y = element_line(),
  axis.text.x  = element_blank(),
  axis.text.y = element_text(size = 12),
  axis.ticks.x = element_blank(),
  axis.ticks.length = unit(0.3,"cm"),
  axis.title.x = element_text(face="bold", size=16),
  axis.title.y = element_text(face="bold", size=16),
  plot.title = element_text(face="bold", size=16, hjust = 0.5)
)

selected_alteration<-c("del13q" ,"IGHV.status","TP53"   , "NOTCH1","trisomy12" ,"SF3B1"   ,"del17p", "ATM","gain8q" ,"del8p"  ,"MED12",  "del11p" ) %>%  rev()

order <- patMeta %>%
  arrange(desc(del13q) ,IGHV.status,desc(TP53)   , desc(NOTCH1),desc(trisomy12) ,desc(SF3B1)   ,desc(del17p), desc(ATM),desc(gain8q) ,desc(del8p)  ,desc(MED12),  desc(del11p)) %>%
  select(PatientID) %>%
  unlist()


patMeta_tmp <- patMeta %>%
  select( -gender, -diagnosis) %>%
  pivot_longer(-PatientID, names_to="Genetic_Alteration", values_to = "alteration_value") %>%

  filter(Genetic_Alteration%in%selected_alteration) %>%
  mutate(alteration_value=as.character(alteration_value),
         Genetic_Alteration=factor(Genetic_Alteration, levels = selected_alteration)) %>%
  mutate(PatientID=factor(PatientID, levels=order))

 p1<- patMeta_tmp %>%
  ggplot(aes(x=PatientID, y=Genetic_Alteration))+
  geom_tile(aes(fill = alteration_value), colour = NA, width=1)+
  theme_temp+
  scale_fill_manual(values = c("0"=Mutant[1], "1"=Mutant[2], "M"=IGHV[1], "U"=IGHV[2],  "unknown"=na_color), na.value=na_color, labels = c("0"="wildtype", "1"="mutated", "M"="IGHV mutated", "U"="IGHV unmutated", "HP"="High programmed", "IP"="Intermediate programmed", "LP"="Low programmmed", "unknown"="N/A"))+
  xlab("")+ ylab("Genetic Alteration")+ labs(fill = "Genetic Alteration")




p2<-
  patMeta_tmp %>%

  filter(!is.na(alteration_value)) %>%
  group_by(Genetic_Alteration, alteration_value) %>%
  tally() %>%
  pivot_wider(names_from = "alteration_value", values_from = "n") %>%
  mutate(perc=case_when(Genetic_Alteration=="IGHV.status"~`M`/(`M`+`U`),

           TRUE~`1`/(`1`+`0`))) %>%

  ggplot(aes(y=Genetic_Alteration, x = perc))+
  geom_bar(stat = "identity", fill=Mutant[2])+
  theme_temp+
  theme(
    axis.text.y = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.x = element_text())+
  guides(fill="none")+
  scale_x_continuous(labels=scales::percent)+
  labs(x="Frequency (%)")


p1+
  p2+
  plot_layout(nrow=1, guides="collect", widths = c(0.7,0.3))
```


```{r, fig.width=6, fig.height=6, eval=FALSE}
p1+
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y= element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank())+
guides(fill="none")
```


## Appendix Figure 1. Genetic profiles of screened patient samples.
Selected genetic alterations on y-axis and screened patient samples (n=192) on x-axis.
\newpage


```{r}
order_Cytokines<-df%>%
  dplyr::filter(Drug=="DMSO") %>%
  group_by(Cytlabel)%>%
      dplyr::summarise(Log=median(Log), .groups="keep")%>%
        arrange(dplyr::desc(Log))

order_Cytokines<-order_Cytokines$Cytlabel%>%
  as.character()

df$Cytlabel<-factor(df$Cytlabel, levels=order_Cytokines)
```


```{r S2,fig.path=plotDir, dev=c("cairo_pdf"),  fig.pos="H", fig.height=8, fig.width=16}

t_tests = df %>%
  dplyr::filter(Drug=="DMSO") %>%
    group_by(Cytlabel) %>%
    summarise(P = t.test(Log, mu = 0)$p.value,
              Sig = case_when(P < 0.00001~ "*****",
                              P < 0.0001~ "****",
                              P < 0.001~ "***",
                              P < 0.01~ "**",
                              P < 0.05~ "*",
                              TRUE~ "ns"),
              .groups = "keep")

df_patMeta <- left_join(df, patMeta, by ="PatientID")

p1<-df_patMeta%>%
  filter(Cytokine%in%c("Resiquimod","IL-4", "sCD40L+IL-4","CpG ODN"), Drug=="DMSO", !is.na(IGHV.status))%>%
  ggplot(aes(x=Cytlabel,
             y=Log))+
  geom_hline(yintercept = 0)+
  geom_violin(cex=0.4, alpha=0.7,  aes(x=Cytlabel,y=Log, color = IGHV.status, fill = IGHV.status))+
   geom_text(aes(label = Sig, y = 2), size = 6, data = filter(t_tests, Cytlabel%in%c("Resiquimod","IL4", "sCD40L + IL4","CpG ODN")))+
  scale_colour_manual(name = 'IGHV status', values = setNames(c(palreds[8],palblues[1]), c("M", "U")))+
 scale_fill_manual(name = 'IGHV status', values = setNames(c(palreds[8],palblues[1]), c("M", "U"))) +
   theme_minimal()+
  # guides(size="none", color = "none")+
  theme(axis.text.x = element_text(angle = 90, hjust = 1), text = element_text(size=20), legend.position = "none")+
  ylab("Log(Viability)")+
  xlab("")


p2<-df_patMeta%>%
  filter(Cytokine%in%c("TGF-b1","IL-1b","Interferon gamma","SDF-1a","sCD40L","soluble anti-IgM","IL-6","IL-10","IL-21","HS-5 CM","IL-15","BAFF","IL-2"), Drug=="DMSO", !is.na(IGHV.status))%>%
  ggplot(aes(x=Cytlabel,
             y=Log))+
  geom_hline(yintercept = 0)+
  geom_violin(cex=0.4, alpha=0.7,aes(x=Cytlabel, y=Log, colour = IGHV.status, fill = IGHV.status))+
   geom_text(aes(label = Sig, y = 0.625), size = 6, data = filter(t_tests, Cytlabel%in%c("IL1\u03B2","Interferon \u03B3","InterferonÎ³","SDF-1\u03B1","sCD40L","soluble anti-IgM","IL6","IL10","IL21","HS-5 CM","IL15","BAFF","IL2", "TGF\u03B21")))+
  scale_colour_manual(name = 'IGHV status', values = setNames(c(palreds[8],palblues[1]), c("M", "U"))) +
  scale_fill_manual(name = 'IGHV status', values = setNames(c(palreds[8],palblues[1]), c("M", "U"))) +
  theme_minimal()+
  theme(axis.text.x = element_text(angle = 90, hjust = 1), text = element_text(size=20), legend.position = "bottom")+
  ylab("") + xlab("Cytokine")


  p1+
  p2+
  plot_layout(nrow = 1, widths = c(1,3.5))
  
  # df_patMeta%>%
  # filter(!is.na(IGHV.status))%>%
  #   group_by(PatientID, IGHV.status) %>% 
  #   summarise() %>% 
  #   ungroup() %>% 
  #   group_by(IGHV.status) %>%
  #   count()
```

## Appendix Figure 2. Response to stimuli by IGHV status.
Natural logarithm of the relative viability after 48h incubation with microenvironmental stimuli, stratified by IGHV status (IGHV mutated: n=99; IGHV unmutated: n=83). BH-transformed p-values are shown from one-sample t-tests of all patient samples. (p<0.00001=&#42;&#42;&#42;&#42;&#42;, p<0.0001=&#42;&#42;&#42;&#42;, p>0.05=ns) 



```{r S3, fig.path=plotDir, dev=c( "cairo_pdf"), fig.pos="H", fig.width=11, fig.height=11}

example_cyt<-c("IL1\u03B2","Resiquimod", "TGF\u03B21","soluble anti-IgM",  "sCD40L",  "BAFF", "CpG ODN","sCD40L + IL4","IL6","SDF-1\u03B1","IL15", "IL4","IL21", "HS-5 CM", "Interferon \u03B3","IL2","IL10")

stat.test <- df %>%
  left_join(Heatmap_clusters, by = "PatientID") %>%
  dplyr::filter(Drug=="DMSO",
                Cytlabel%in%example_cyt,
                Cluster%in%c(1,2,3,4)) %>%
  dplyr::mutate(Cluster=factor(Cluster)
                ,Cytlabel=factor(Cytlabel, levels=example_cyt)
                )  %>%
  group_by(Cytlabel) %>%
  rstatix::t_test(Log ~ Cluster, p.adjust.method="BH")%>%
  filter((group1==1&group2==2)|(group1==3&group2==4)) %>%
  rstatix::add_y_position(step.increase=0.001)

df %>%
  left_join(Heatmap_clusters, by = "PatientID") %>%
  dplyr::filter(Drug=="DMSO",
                Cytlabel%in%example_cyt,
                Cluster%in%c(1,2,3,4)) %>%
  dplyr::mutate(Cluster=factor(Cluster)
                ,Cytlabel=factor(Cytlabel, levels=example_cyt)
                ) %>%

  ggplot(aes(x=Cluster, y= Log))+
  stat_pvalue_manual(stat.test, label = "p.adj", tip.length = 0.01)+
  geom_hline(yintercept = 0)+
  geom_boxplot()+
  geom_beeswarm(aes(color=Cluster))+
  scale_color_manual(values=colors[1:4])+
  t2+ theme(strip.background =element_blank())+
  guides(color="none")+
  facet_wrap(~Cytlabel, ncol=4, scales="free_y")+
  ylab("Log(Viability)")+
  coord_cartesian(clip="off")

# Heatmap_clusters %>% 
#   group_by(Cluster) %>% 
#   count()

```

## Appendix Figure 3. Stimuli response by clusters.
Log transformed viabilities after treatment with stimuli, faceted by cluster. BH-transformed p-values from Student's t-test are shown (C1: n=77; C2:n=27; C3: n=38; C4: n=50).

```{r S4, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.pos="H", fig.width=9, fig.height=10, warning=F, message=F}

# Run model with OS
coxph_summ <- clusters.surv%>%
  dplyr::mutate(Cluster=factor(Cluster), Cluster_12vs34=ifelse(Cluster%in%c(1,2),1,0))%>%
      coxph(Surv( OS, died) ~  Cluster_12vs34,  data=.) %>%
    summary()

coxph_coeff<-coxph_summ$coefficients

C12vsC34_p <- coxph_coeff[1,5] %>% round(3)

##Run survfit on OS
fit <- survfit(Surv(OS, died)~Cluster, clusters.surv)


t_plot<-  t2+theme(plot.title=element_blank(), axis.title.x=element_blank(), legend.key = element_blank())
t_table<- t2+theme(plot.title=element_blank(), panel.grid.major.x = element_blank())

OS_Clusters_KM_surv <- ggsurvplot(fit,
                 surv.median.line = "hv", # Add medians survival
                 legend.title = "Cluster",# Change legends: title & labels
                  legend.labs = c("1", "2","3","4"),
                  pval = TRUE,
                 risk.table = TRUE,
                 palette = c(colors[1], colors[2], colors[3], colors[4]),
                 ggtheme = t_plot,
                 tables.theme = t_table,
                 legend = "bottom",
                 xlab="Time in years",
                 ylab="Overall survival",
                 break.time.by=1,
                 xlim=c(0,6.8),
                 ylim=c(0.6,1))

upper_end<-0.94
lower_end<-0.72
middle<-(upper_end+lower_end)/2
left<-6.2
right<-left+0.1

OS_Clusters_KM_surv_plot<-OS_Clusters_KM_surv$plot+
  annotate(geom="text", x=right+0.5, y=middle,    label=C12vsC34_p, color="black", size=6)+
  geom_segment(x = right, xend = right, y = upper_end, yend = lower_end, color="black")+
  geom_segment(x = left, xend = right, y = upper_end, yend = upper_end, color="black")+
  geom_segment(x = left, xend = right, y = 0.73+0.025, yend = 0.73+0.025, color="black")+
  geom_segment(x = left, xend = right, y = lower_end+0.002, yend = lower_end+0.002, color="black")


OS_Clusters_KM <- wrap_elements(OS_Clusters_KM_surv_plot+ OS_Clusters_KM_surv$table + plot_layout(ncol=1, heights = c(0.75,0.25)))


# Univariate Cox Proportional Hazards on TFT
# Cluster 3 as reference cluster

coxph_summ <- clusters.surv%>%
  mutate(Cluster=factor(Cluster))%>%
    coxph(Surv( TFT, treatment.x) ~  factor2ind(Cluster, 3),  data=.) %>%
  summary()

coxph_coeff<-coxph_summ$coefficients
rownames(coxph_coeff)<-rownames(coxph_coeff) %>%
  gsub(pattern="factor2ind", "", .)%>%
  gsub(pattern="[()]", "", .)%>%
  gsub(pattern=":", " ", .)%>%
  gsub(pattern="Cluster, 3", "", .)



C3vsC4_p<-coxph_coeff[3,5] %>% round(3)

##Run survfit on TFT
fit <- survfit(Surv(TFT, treatment.x)~Cluster, clusters.surv)


t_plot<-  t2+theme(plot.title=element_blank(), axis.title.x=element_blank(), legend.key = element_blank())
t_table<- t2+theme(plot.title=element_blank(), panel.grid.major.x = element_blank())

TFT_Clusters_KM_surv <- ggsurvplot(fit,
                 surv.median.line = "hv", # Add medians survival
                 legend.title = "Cluster",# Change legends: title & labels
                  legend.labs = c("1", "2","3","4"),
                  pval = FALSE,
                 risk.table = TRUE,
                 palette = c(colors[1], colors[2], colors[3], colors[4]),
                 ggtheme = t_plot,
                 tables.theme = t_table,
                 legend = "bottom",
                 xlab="Time in years",
                 ylab="Time to first treatment",
                 break.time.by=5,
                 xlim=c(0,33))

TFT_Clusters_KM_surv_plot<-TFT_Clusters_KM_surv$plot+
  annotate(geom="text", x=33, y=0.3,    label=C3vsC4_p, color="black", size=6)+
  geom_segment(x = 30.7, xend = 30.7, y = 0.535, yend = 0.08, color="black")+
  geom_segment(x = 30.4, xend = 30.7, y = 0.535, yend = 0.535, color="black")+
  geom_segment(x = 30.4, xend = 30.7, y = 0.083, yend = 0.083, color="black")


TFT_Clusters_KM <- wrap_elements(TFT_Clusters_KM_surv_plot+ TFT_Clusters_KM_surv$table + plot_layout(ncol=1, heights = c(0.75,0.25)))

TFT_Clusters_KM+theme(plot.tag=element_text(size = 20))+
OS_Clusters_KM+theme(plot.tag=element_text(size = 20))+
  plot_layout(nrow=2)+
  plot_annotation(tag_levels = "A")

```

## Appendix Figure 4. Time to first treatment and overall survival by clusters.
Kaplan-Meier Curves of time to first treatment with p-value from univariate Cox proportional hazards model between C3 and C4 **(A)** and overall survival with p-value from univariate Cox proportional hazards  model between C1&2 and C3&4 **(B)**. Median survival not reached.



\newpage

```{r S5, fig.path=plotDir, dev=c("png", "cairo_pdf"),fig.height=5}
plotTab<-patMeta_Cluster %>% 
  group_by(Cluster, Methylation_Cluster) %>% 
  count() %>% 
  group_by(Cluster) %>% 
  mutate(freq = n / sum(n)) %>% 
   plyr::ddply("Cluster", transform, label_ypos=cumsum(freq))


  ggplot(plotTab, aes(x=Cluster, y=freq, fill=Methylation_Cluster))+
  geom_bar(stat="identity")+
  ylab("Relative incidence of\n methylation profiles")+
  scale_fill_manual(name = 'Methylation Profile', values = Methylation_cluster)+
  t2
```

## Appendix Figure 5. Methylation profile of Clusters defined by response to microenvironmental stimuli
Barplot of methylation profiles in the four clusters of microenvironmental response. Predominantly IGHV unmutated clusters 1 and 2 show higher abundance of Low programmed Methylation profiles, while predominantly IGHV mutated clusters 3 and 4 show more highly programmed methylation clusters (C1: n=77; C2:n=27; C3: n=38; C4: n=50).

\newpage

<!-- ### Additional Figure of CLL_PD and Clusters of Microenvironmental response -->
```{r fig.height=5}

plotTab<-CLL_PD %>% 
  left_join( patMeta_Cluster, by="PatientID") %>% 
  mutate(Cluster=as.factor(Cluster)) 
  
 CLL_PD_Boxplot<- 
  ggplot(plotTab, aes(x=Cluster, y=CLL_PD, group=Cluster, fill=Cluster ))+
  geom_boxplot(outlier.shape = NA)+
  geom_beeswarm(aes( ))+
  t2+
  scale_fill_manual(values=colors[1:4]) +
  guides(fill="none") +
  ylab("CLL Proliferative Drive") +
  stat_compare_means(method="t.test", comparison=list(c(1,2),c(3,4)), size=4, label.x = 1.3, label.y=3.1)+
  coord_cartesian(clip="off")+
  theme(axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=10), 
        strip.text = element_text(size=12))
 
 # plotTab %>% 
 #   group_by(Cluster) %>% 
 #   count()
```


```{r, message=FALSE, warning=FALSE}

########### Lasso Model of CLL_PD and Cytokine response in all patients ################
#make viability matrix for cytokine treatments for patients
cyt_viab <- dplyr::select(dplyr::filter(df, Drug == "DMSO", Cytokine != "No Cytokine"), 
                PatientID, 
                Log, 
                Cytokine) %>% spread(Cytokine, Log) %>% as.data.frame()

#make PatID the row names
rownames(cyt_viab) <- unlist(cyt_viab[,1]) # the first row will be the header
cyt_viab <- dplyr::select(cyt_viab,-PatientID) 

cyt_viab <- t(cyt_viab)

#run scaleCytResp on viability matrix
x <- scaleCytResp(cyt_viab)

viabMat <-t(x)


## Response matrix 
PDmat <- 
  dplyr::select(CLL_PD, PatientID, CLL_PD) %>% 
  #reshape data
  spread(key = PatientID, value = CLL_PD) %>% 
  data.frame() 
  


#get only viability values for aptients that have a PD
viabMat <- viabMat[colnames(PDmat),]


# Run logistic regression
## Run model
PDmat <- t(PDmat)

#fit the model
cvglmfit <- runGlm(viabMat, PDmat, method = "lasso", repeats = 30, folds = 3)
    
#collect the results for each stimulus in one object
dataResult <- cvglmfit


# 
# Get predictor profiles for stimuli of interest

PDmat <- 
  dplyr::select(CLL_PD, PatientID, CLL_PD) %>% 
  #reshape data
  spread(key = PatientID, value = CLL_PD) %>% 
  data.frame() 
  

#make sample order same as in geneMatrix
PDmat <- PDmat[,rownames(viabMat)]

#get only viability values for aptients that have a PD
viabMat <- viabMat[colnames(PDmat),]


CLLPD_Lasso_all_patients<-lassoPlot_CLLPD(dataResult, 
          viabMat, #use viab matrix for heatmap 
          PDmat, #use CLL_PD matrix for scatter plot
          freqCut = 0.75, #coefficients should be selected in <75% of bootstrapped model files
          coefCut = 0.00) #no minimum value for coefficients 

# str(PDmat)
# str(viabMat)
```



```{r, message=FALSE, warning=FALSE}
########### Lasso Model of CLL_PD and Cytokine response in IGHV M patients ################
#make viability matrix for cytokine treatments for patients
cyt_viab <- dplyr::select(dplyr::filter(df, Drug == "DMSO", Cytokine != "No Cytokine"), 
                PatientID, 
                Log, 
                Cytokine) %>% spread(Cytokine, Log) %>% as.data.frame()

#make PatID the row names
rownames(cyt_viab) <- unlist(cyt_viab[,1]) # the first row will be the header
cyt_viab <- dplyr::select(cyt_viab,-PatientID) 

cyt_viab <- t(cyt_viab)

#run scaleCytResp on viability matrix
x <- scaleCytResp(cyt_viab)

viabMat <-t(x)

Mpat <- patMeta %>% filter(IGHV.status == "M") %>% select(PatientID)

PDmat <- 
  filter(CLL_PD, PatientID %in% Mpat$PatientID) %>%
  dplyr::select(PatientID, CLL_PD) %>% 
  #reshape data
  spread(key = PatientID, value = CLL_PD) %>% 
  data.frame() 
  
#get only viability values for aptients that have a PD
viabMat <- viabMat[colnames(PDmat),]

PDmat <- t(PDmat)

#fit the model
cvglmfit <- runGlm(viabMat, PDmat, method = "lasso", repeats = 30, folds = 3)
    
#collect the results for each stimulus in one object
dataResult <- cvglmfit


PDmat <- 
  dplyr::select(CLL_PD, PatientID, CLL_PD) %>% 
  #reshape data
  spread(key = PatientID, value = CLL_PD) %>% 
  data.frame() 
  

#make sample order same as in geneMatrix
PDmat <- PDmat[,rownames(viabMat)]

#get only viability values for aptients that have a PD
viabMat <- viabMat[colnames(PDmat),]


CLLPD_Lasso_M_patients<-lassoPlot_CLLPD(dataResult, 
          viabMat, #use viab matrix for heatmap 
          PDmat, #use CLL_PD matrix for scatter plot
          freqCut = 0.75, #coefficients should be selected in <75% of bootstrapped model files
          coefCut = 0.00) #no minimum value for coefficients 


# str(PDmat)
# str(viabMat)
```


```{r, message=FALSE, warning=FALSE}
########### Lasso Model of CLL_PD and Cytokine response in IGHV U patients ################
#make viability matrix for cytokine treatments for patients
cyt_viab <- dplyr::select(dplyr::filter(df, Drug == "DMSO", Cytokine != "No Cytokine"), 
                PatientID, 
                Log, 
                Cytokine) %>% spread(Cytokine, Log) %>% as.data.frame()

#make PatID the row names
rownames(cyt_viab) <- unlist(cyt_viab[,1]) # the first row will be the header
cyt_viab <- dplyr::select(cyt_viab,-PatientID) 

cyt_viab <- t(cyt_viab)

#run scaleCytResp on viability matrix
x <- scaleCytResp(cyt_viab)

viabMat <-t(x)

Upat <- patMeta %>% filter(IGHV.status == "U") %>% select(PatientID)

PDmat <- 
  filter(CLL_PD, PatientID %in% Upat$PatientID) %>%
  dplyr::select(PatientID, CLL_PD) %>% 
  #reshape data
  spread(key = PatientID, value = CLL_PD) %>% 
  data.frame() 
  
#get only viability values for aptients that have a PD
viabMat <- viabMat[colnames(PDmat),]

PDmat <- t(PDmat)

#fit the model
cvglmfit <- runGlm(viabMat, PDmat, method = "lasso", repeats = 30, folds = 3)
    
#collect the results for each stimulus in one object
dataResult <- cvglmfit


PDmat <- 
  dplyr::select(CLL_PD, PatientID, CLL_PD) %>% 
  #reshape data
  spread(key = PatientID, value = CLL_PD) %>% 
  data.frame() 
  

#make sample order same as in geneMatrix
PDmat <- PDmat[,rownames(viabMat)]

#get only viability values for aptients that have a PD
viabMat <- viabMat[colnames(PDmat),]


CLLPD_Lasso_U_patients<-lassoPlot_CLLPD(dataResult, 
          viabMat, #use viab matrix for heatmap 
          PDmat, #use CLL_PD matrix for scatter plot
          freqCut = 0.75, #coefficients should be selected in <75% of bootstrapped model files
          coefCut = 0.00) #no minimum value for coefficients 


# str(PDmat)
# str(viabMat)
```

```{r, fig.width=4, fig.height=8}

IL1b_Scatterplot<-left_join(CLL_PD, filter(df, Drug=="DMSO"), by = "PatientID") %>% 
  left_join(patMeta, by = "PatientID") %>% 
  filter(Cytokine%in%c("IL-1b"), !is.na(IGHV.status)) %>% 
  
  ggplot(aes(x=CLL_PD, y=Log)) +
  geom_smooth(method="lm", formula= 'y ~ x')+
  geom_point(aes(color=IGHV.status))+
  facet_wrap(.~IGHV.status,  nrow = 2, strip.position="right", labeller = labeller(IGHV.status = c("U" = "IGHV-U", "M" =  "IGHV-M")))+
  stat_cor()+
  guides(color="none")+
  t2+
  theme(axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=10), 
        strip.text = element_text(size=12),
        plot.title = element_text(size=14))+
  ggtitle("IL1Ã")+
  ylab("Log(Viability)")

Res_Scatterplot<-left_join(CLL_PD, filter(df, Drug=="DMSO"), by = "PatientID") %>% 
  left_join(patMeta, by = "PatientID") %>% 
  filter(Cytokine%in%c("Resiquimod"), !is.na(IGHV.status)) %>% 
  
  ggplot(aes(x=CLL_PD, y=Log)) +
  geom_smooth(method="lm", formula= 'y ~ x')+
  geom_point(aes(color=IGHV.status))+
  facet_wrap(.~IGHV.status,  nrow = 2, strip.position="right", labeller = labeller(IGHV.status = c("U" = "IGHV-U", "M" =  "IGHV-M")))+
  stat_cor()+
  guides(color="none")+
  t2+
  theme(axis.title.x = element_text(size=12),
        axis.title.y = element_text(size=12),
        axis.text.x = element_text(size=10),
        axis.text.y = element_text(size=10), 
        strip.text = element_text(size=12),
        plot.title = element_text(size=14))+
  ggtitle("Resiquimod")+
  ylab("Log(Viability)")

# left_join(CLL_PD, filter(df, Drug=="DMSO"), by = "PatientID") %>% 
#   left_join(patMeta, by = "PatientID") %>% 
#   filter(Cytokine%in%c("IL-1b"), !is.na(IGHV.status)) %>% 
#   group_by(IGHV.status) %>% 
#   count()
  
  
```


```{r S6, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height=13, fig.width=10.5}

design1<-"
  15
  25
  26
  36
  46
  "

wrap_elements(CLL_PD_Boxplot)+theme(plot.tag=element_text(size = 20))+
wrap_elements(CLLPD_Lasso_all_patients)+theme(plot.tag=element_text(size = 20))+
wrap_elements(CLLPD_Lasso_M_patients)+theme(plot.tag=element_text(size = 20))+
wrap_elements(CLLPD_Lasso_U_patients)+theme(plot.tag=element_text(size = 20))+
  Res_Scatterplot+theme(plot.tag=element_text(size = 20))+
  IL1b_Scatterplot+theme(plot.tag=element_text(size = 20))+
  
  
  
  plot_layout(design=design1, heights = c(.9,.8,.1,.7,.85), widths = c(2,.7))+
  plot_annotation(tag_levels = "A")
  
```

## Appendix Figure 6. CLL Proliferative Drive as defined by Lu et al., 2021^2^.
Beeswarm plot of Proliferative Drive in the clusters defined by microenvironmental resonse. P-values are shown from Student's t-test (C1: n=57; C2:n=20; C3: n=29; C4: n=40). **(A)** Predictor profiles for stimuli response to represent CLL Proliferative Drive identified through multivariate analysis using Gaussian linear modelling with L1-penalty for all patient samples which were analysed (n=146) **(B)**, for only IGHV mutated  (n=80) **(C)** and for only IGHV unmutated patient samples (n=65) **(D)** Within the plots, the bar plots on left indicate size and sign of coefficients assigned to the named predictors. Positive coefficients indicate higher viability after stimulation if the feature is present. Scatter plots indicate log(viability) values, in order of magnitude, for each individual sample. Heatmaps show status for each of the genetic predictors for corresponding sample in scatter plot.  Exemplary scatter plots devided by IGHV mutation status shown for Resiquimod **(E)** and IL1Ã **(F)** (IGHV-M: n=80, IGHV-U: n=65). CLL-PD on x-axis and log normalised viability after microenvironmental stimulation on y-axis. Person correlation coefficients are shown with p-values. 

\newpage

<!-- ###  Drug Response heatmap -->
<!-- Heatmap of all scaled log(viability) values, for all patient samples and drugs. -->
```{r }

########### Viability matrix ################
#make viability matrix for cytokine treatments for patients
viab.mat <- filter(df, Drug!="DMSO", Cytokine=="No Cytokine", Drug_Concentration=="High") %>% 
            dplyr::select(PatientID, Log, Drug) %>% 
            tidyr::spread(Drug, Log) %>%
  as.data.frame()

#make PatientID the row names
rownames(viab.mat) <- unlist(viab.mat[,1]) # the first row will be the header
viab.mat <- viab.mat[,-1]

#Transform matrix
viab.mat <- t(viab.mat)

#keep unscaled matrix 
viab.mat.unscaled <- viab.mat

#run scaleCytResp on viability matrix
# viab.mat <- scaleCytResp(viab.mat)

#apply deckel function to matrix
#Calculate dendrogram 
breaks <- c(seq(-2,2, length.out = 101))

# viab.mat.lims <- deckel(viab.mat,
#                             lower = dplyr::first(breaks),
#                             upper = dplyr::last(breaks))

```

<!-- Arrange annotations for heatmap -->
```{r}

#Sort heatmap annotations
#Select annotations from patMeta_cl
Heatmap_Annotation <- dplyr::select(patMeta_Cluster, 
                                    PatientID, 
                                    IGHV.status, 
                                    Methylation_Cluster,
                                    trisomy12,
                                    TP53,
                                    del17p,
                                    ATM,
                                    treatment, 
                                    gender, 
                                    Cluster) %>%
  
  #Adjust names/levels for legend
  mutate(treatment=case_when(treatment==0~"No", 
                             treatment==1~"Yes")) %>%

  mutate(gender=case_when(gender=="f"~"Female", 
                          gender=="m"~"Male")) %>%
  
  mutate(IGHV.status=case_when(IGHV.status=="U"~"Unmutated", 
                               IGHV.status=="M"~"Mutated"))%>%
  
  mutate(Methylation_Cluster=case_when(Methylation_Cluster=="LP"~"Low", 
                               Methylation_Cluster=="IP"~"Intermediate", 
                               Methylation_Cluster=="HP"~"High"))


#Rename columns for legend
colnames(Heatmap_Annotation) <- c("PatientID", 
                                  "IGHV status",
                                  "Methylation Cluster",
                                  "trisomy 12", 
                                  "TP53", 
                                  "deletion 17p",
                                  "ATM",
                                  "Pretreated", 
                                  "Sex", 
                                  "Cluster")

#make Pat IDs the rownames
Heatmap_Annotation <- as.data.frame(Heatmap_Annotation)
rownames(Heatmap_Annotation) <- unlist(Heatmap_Annotation$PatientID)

#Tidy Up Heatmap Annotation table
Heatmap_Annotation$Cluster <- as.factor(Heatmap_Annotation$Cluster)

Heatmap_Annotation <- dplyr::select(Heatmap_Annotation,-"PatientID")

########### Set Heatmap Aesthetics ###############
#Generate red-blue divergent palette
breaks <- breaks %>% `names<-`(
     colorRampPalette(c(palblues, "white",palreds))(101))

#Specify colors of annotations 
 ann_colors = 
   list(
    "IGHV status" = c("Unmutated"=IGHV[1],
                      "Mutated"=IGHV[2]),
    "Methylation Cluster"= c("Low"=Methylation_cluster[1],
                             "Intermediate"=Methylation_cluster[2],
                             "High"=Methylation_cluster[3]),
    "Sex" = c("Female"=Sex[1],
              "Male"=Sex[2]),
    "Pretreated" = c("No"=Mutant[1],
                  "Yes"=Mutant[2]),
    "trisomy 12" = c("1"=Mutant[2],
                     "0"=Mutant[1]),
    "ATM" = c("1"=Mutant[2],
              "0"=Mutant[1]),
    "TP53" = c("1"=Mutant[2],
               "0"=Mutant[1]),
    "deletion 17p" = c("1"=Mutant[2],
               "0"=Mutant[1]),
    "Cluster" = c("1"=colors[1],
                  "2"=colors[2],
                  "3"=colors[3],
                  "4"=colors[4]))
```



```{r S7, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width=14.5, fig.height=11, warning = FALSE}
Cluster_for_Heatmap<-patMeta_Cluster %>% select(PatientID, Cluster) %>%
  column_to_rownames("PatientID") %>%
  as.data.frame()


#Plot heatmap
viab.mat.tmp<-viab.mat.unscaled %>% 
  t() %>% 
  merge(Cluster_for_Heatmap,. , by=0)%>% 
  column_to_rownames("Row.names") %>% 
  filter(Cluster==1) %>% 
  select(-Cluster) %>% 
  t() 

viab.mat.tmp.ordered<-viab.mat.tmp[ c( "Fludarabine" ,"Nutlin-3a" ,"Luminespib"  ,"BAY-11-7085" ,"Pyridone-6" ,"Ralimetinib" ,"I-BET 762"  ,"Everolimus" ,"Selumetinib" ,"Idelalisib" ,      "Ibrutinib" ,"PRT062607"), ]
  
# row.names(viab.mat.tmp)

C1_drug_heatmap<-pheatmap(viab.mat.tmp.ordered, 
         annRow = Cluster_for_Heatmap,
         type="Cluster",
         doTranspose=TRUE,
         # scale = "row",
         # clustering_method = "complete", 
         cluster_rows = FALSE,
         show_colnames = FALSE,
         annotation_col = Heatmap_Annotation,
         annotation_colors = ann_colors,
         cellheight = 22,
         cellwidth = 4,
         breaks = breaks,
         color= names(breaks),
         border_color=NA,
         fontsize=10,
         fontsize_row=16, 
         legend=FALSE,
         annotation_legend = FALSE, 
         show_rownames=FALSE,
         annotation_names_col=FALSE, 
         silent = TRUE
         )
  
#Plot heatmap
viab.mat.tmp<-viab.mat.unscaled %>% 
  t() %>% 
  merge(Cluster_for_Heatmap,. , by=0)%>% 
  column_to_rownames("Row.names") %>% 
  filter(Cluster==2) %>% 
  select(-Cluster) %>% 
  t() 

viab.mat.tmp.ordered<-viab.mat.tmp[ c( "Fludarabine" ,"Nutlin-3a" ,"Luminespib"  ,"BAY-11-7085" ,"Pyridone-6" ,"Ralimetinib" ,"I-BET 762"  ,"Everolimus" ,"Selumetinib" ,"Idelalisib" ,      "Ibrutinib" ,"PRT062607"), ]
  
# row.names(viab.mat.tmp)

C2_drug_heatmap<-pheatmap(viab.mat.tmp.ordered, 
         annRow = Cluster_for_Heatmap,
         type="Cluster",
         doTranspose=TRUE,
         # scale = "row",
         # clustering_method = "complete", 
         cluster_rows = FALSE,
         show_colnames = FALSE,
         annotation_col = Heatmap_Annotation,
         annotation_colors = ann_colors,
         cellheight = 22,
         cellwidth = 4,
         breaks = breaks,
         color= names(breaks),
         border_color=NA,
         fontsize=10,
         fontsize_row=16, 
         legend=FALSE,
         annotation_legend = FALSE, 
         show_rownames=FALSE,
         annotation_names_col=FALSE, 
         silent = TRUE
         )
  
#Plot heatmap
viab.mat.tmp<-viab.mat.unscaled %>% 
  t() %>% 
  merge(Cluster_for_Heatmap,. , by=0)%>% 
  column_to_rownames("Row.names") %>% 
  filter(Cluster==3) %>% 
  select(-Cluster) %>% 
  t() 

viab.mat.tmp.ordered<-viab.mat.tmp[ c( "Fludarabine" ,"Nutlin-3a" ,"Luminespib"  ,"BAY-11-7085" ,"Pyridone-6" ,"Ralimetinib" ,"I-BET 762"  ,"Everolimus" ,"Selumetinib" ,"Idelalisib" ,      "Ibrutinib" ,"PRT062607"), ]
  
# row.names(viab.mat.tmp)

C3_drug_heatmap<-pheatmap(viab.mat.tmp.ordered, 
         annRow = Cluster_for_Heatmap,
         type="Cluster",
         doTranspose=TRUE,
         # scale = "row",
         # clustering_method = "complete", 
         cluster_rows = FALSE,
         show_colnames = FALSE,
         annotation_col = Heatmap_Annotation,
         annotation_colors = ann_colors,
         cellheight = 22,
         cellwidth = 4,
         breaks = breaks,
         color= names(breaks),
         border_color=NA,
         fontsize=10,
         fontsize_row=16, 
         legend=FALSE,
         annotation_legend = FALSE, 
         show_rownames=FALSE,
         annotation_names_col=FALSE, 
         silent = TRUE
         )
  

#Plot heatmap
viab.mat.tmp<-viab.mat.unscaled %>% 
  t() %>% 
  merge(Cluster_for_Heatmap,. , by=0)%>% 
  column_to_rownames("Row.names") %>% 
  filter(Cluster==4) %>% 
  select(-Cluster) %>% 
  t() 

viab.mat.tmp.ordered<-viab.mat.tmp[ c( "Fludarabine" ,"Nutlin-3a" ,"Luminespib"  ,"BAY-11-7085" ,"Pyridone-6" ,"Ralimetinib" ,"I-BET 762"  ,"Everolimus" ,"Selumetinib" ,"Idelalisib" ,      "Ibrutinib" ,"PRT062607"), ]
  
# row.names(viab.mat.tmp)

C4_drug_heatmap<-pheatmap(viab.mat.tmp.ordered, 
         annRow = Cluster_for_Heatmap,
         type="Cluster",
         doTranspose=TRUE,
         # scale = "row",
         # clustering_method = "complete", 
         cluster_rows = FALSE,
         show_colnames = FALSE,
         annotation_col = Heatmap_Annotation,
         annotation_colors = ann_colors,
         cellheight = 22,
         cellwidth = 4,
         breaks = breaks,
         color= names(breaks),
         border_color=NA,
         fontsize=10,
         fontsize_row=16, 
         silent = TRUE
         )
  


wrap_elements(arrangeGrob(C1_drug_heatmap[[4]]))+
wrap_elements(arrangeGrob(C2_drug_heatmap[[4]]))+
wrap_elements(arrangeGrob(C3_drug_heatmap[[4]]))+
wrap_elements(arrangeGrob(C4_drug_heatmap[[4]]))+
  plot_layout(nrow=1, widths = c(0.93,0.13,0.52,1.03))
```

## Appendix Figure 7. Drug response between clusters.
Heatmap showing the log normalised viability after drug treatment. Rows represent drug treatments and columns represent primary CLL samples (n=192), annotated for their genetic background, sex and pretreatment status above. Red values indicate increased viability upon treatment, blue indicates decreased viability.

```{r S8, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width=11, fig.height=12}

Drug_order<-c("BAY-11-7085","Fludarabine","Everolimus","Nutlin-3a","Idelalisib","Ralimetinib","Ibrutinib","I-BET 762","PRT062607","Pyridone-6","Selumetinib","Luminespib")
  
stat.test <- df %>%
  left_join(Heatmap_clusters, by = "PatientID") %>%
  dplyr::filter(Drug!="DMSO",
                Drug_Concentration=="High",
                Cytlabel=="No Cytokine",
                Cluster%in%c(1,2,3,4)) %>%
  dplyr::mutate(Cluster=factor(Cluster)
                ,Drug=factor(Drug, levels = Drug_order)
                )  %>%
  group_by(Drug) %>%
  rstatix::t_test(Log ~ Cluster, p.adjust.method="BH",paired = FALSE,
  var.equal = FALSE,  alternative = "two.sided", comparisons=list(c(1,2),c(3,4)))%>%
  rstatix::add_y_position(step.increase=0.001)

df %>%
  left_join(Heatmap_clusters, by = "PatientID") %>%
  dplyr::filter(Drug!="DMSO",
                Drug_Concentration=="High",
                Cytlabel=="No Cytokine",
                Cluster%in%c(1,2,3,4)) %>%
  dplyr::mutate(Cluster=factor(Cluster)
                ,Drug=factor(Drug, levels = Drug_order)
                ) %>%
  

  ggplot(aes(x=Cluster, y= Log))+
  stat_pvalue_manual(stat.test, label = "p.adj", tip.length = 0.01)+
  geom_hline(yintercept = 0)+
  geom_boxplot()+
  geom_beeswarm(aes(color=Cluster))+
  scale_color_manual(values=colors[1:4])+
  t2+ theme(strip.background =element_blank())+
  guides(color="none")+
  facet_wrap(.~Drug, ncol=3, scales="free_y")+
  ylab("Log(Viability)")+
  coord_cartesian(clip="off")


```

## Appendix Figure 8. Drug response by clusters.
Natural logarithm of the relative viability after treatment with drugs, faceted by cluster. BH-transformed p-values from non-paired, two-sided Student's t-test are shown (C1: n=77; C2:n=27; C3: n=38; C4: n=50).



```{r S9, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height= 14, fig.width = 12, warning=F, message=F, cache=FALSE}

cytSeq <- dds_smp

#Filter out IGHV genes
cytSeq <- cytSeq[!grepl("IG_", rowData(cytSeq)$biotype),]

#split into 3,4
cytSeq.34 <- cytSeq[,colData(cytSeq)[,"Cluster"] %in% c(3,4)]

#remove patients where IGHV is unknown
cytSeq.34 <- cytSeq.34[,colData(cytSeq.34)[,"IGHV.status"] %in% c("U","M")]

#set order of factors
cytSeq.34$Cluster <- factor(cytSeq.34$Cluster,
                            levels = c("3","4"))

cytSeq.34$IGHV.status <- factor(cytSeq.34$IGHV.status,
                                levels = c("U","M"))

design(cytSeq.34) <- ~ IGHV.status + Cluster

cytSeq.34 <- DESeq(cytSeq.34)

res.ds <-
  results(cytSeq.34, contrast = c("Cluster", 3, 4), tidy = TRUE) %>%
  dplyr::rename(Symbol = "row") %>%
  arrange(pvalue)

#get ensembl ids to Entrez dataframe
ens2entrez <-
  rowData(cytSeq.34)[c("entrezgene", "symbol")] %>%
  as.data.frame() %>%
  rownames_to_column("ENSEMBL") %>%
  tibble::as_tibble()



#Join
res.ds <- left_join(res.ds, ens2entrez, by=c("Symbol"="ENSEMBL"))


d <-
  dplyr::select(res.ds, entrezgene, stat) %>%
  na.omit() %>%
  dplyr::group_by(entrezgene) %>%
  dplyr::summarize(stat=mean(stat), .groups = "keep")

## feature 1: numeric vector
geneList <- d$stat

## feature 2: named vector
names(geneList) <- as.character(d$entrezgene)

## feature 3: decreasing order
geneList <- sort(geneList, decreasing = TRUE)


hm2gene <-
  msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gs_name, entrez_gene)
# head(hm2gene)

#run GSEA
set.seed(1996)
gsea.res <- GSEA(geneList, TERM2GENE = hm2gene, by = "fgsea", seed = TRUE, verbose = FALSE)

#make readable with gene names
gsea.res <- setReadable(gsea.res, org.Hs.eg.db, keyType = "ENTREZID")

## TNFa
TNFaEnrichment <-
  enrichmentPlot(filter(hm2gene, gs_name =="HALLMARK_TNFA_SIGNALING_VIA_NFKB")$entrez_gene,geneList) +
  labs(title="TNFA signalling via NFKB")


#myc targets
mycEnrichment <-
  enrichmentPlot(filter(hm2gene, gs_name =="HALLMARK_MYC_TARGETS_V1")$entrez_gene, geneList) +
  labs(title="MYC targets V1")


#oxidative phosphorylation
oxphosEnrichment <-
  enrichmentPlot(filter(hm2gene, gs_name =="HALLMARK_OXIDATIVE_PHOSPHORYLATION")$entrez_gene, geneList) +
  labs(title="Oxidative Phosphorylation")


#p53 pathway
p53Enrichment <-
  enrichmentPlot(filter(hm2gene, gs_name =="HALLMARK_P53_PATHWAY")$entrez_gene, geneList) +
  labs(title="P53 Pathway")


sgncuoff <- 1.4

volPlot <-
  res.ds %>%
  filter(!is.na(padj), !is.na(symbol), symbol!="") %>%

ggplot( aes(x=log2FoldChange, y=-log10(padj))) +
      annotate("rect", xmin = 0.0, xmax = Inf, ymin = -Inf, ymax = Inf, fill = palreds[1], alpha = 0.2)+
      annotate("rect", xmin = 0.0, xmax = -Inf, ymin = -Inf, ymax = Inf,alpha = 0.2, fill = palblues[1])+
      geom_point(aes(alpha=0.4, colour = ifelse(-log10(padj)>sgncuoff,"black","grey"))) +
      theme(legend.position = "none") +
      geom_label_repel(aes(label=ifelse(-log10(padj)>sgncuoff&abs(log2FoldChange)>2,as.character(symbol),''),
                           colour=ifelse(log2FoldChange>0, palreds[2], palblues[4])), fill="#FDFDFD",
                       size=3.5, max.overlaps=30)+
      scale_colour_manual(values=c(palblues[2], palreds[4], "black", "grey"))+
      t2 +
      labs(title="Cluster 3 versus 4", x = "Log2FC", y = "-log10(adj p-val)", color = "Significance")+
      guides(color="none", alpha="none")


design1<-"
  1111
  2233
  4455
  "

tp <- theme(plot.tag=element_text(size = 30))

  volPlot + tp +
  wrap_elements(TNFaEnrichment) + tp +
  wrap_elements(mycEnrichment) + tp +
  wrap_elements(oxphosEnrichment) + tp +
  wrap_elements(p53Enrichment) + tp +
  plot_layout(design=design1, heights = c(2,1,1))+
  plot_annotation(tag_levels = "A")

# cytSeq[,colData(cytSeq)[,"Cluster"] %in% c(3)]  
# cytSeq[,colData(cytSeq)[,"Cluster"] %in% c(4)]  
  
```

```{r, fig.width=4, fig.height=4}
volPlot+
  theme(title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        panel.grid.major= element_blank(),
        panel.grid.minor = element_blank(), 
        axis.ticks.x = element_blank(),
        axis.ticks.y = element_blank(),
        )+
  ggtitle("")
```


## Appendix Figure 9. RNA-Sequencing of matched samples indicates differential gene expression between  Cluster 3 and Cluster 4.
Volcano plot of differentially expressed genes between Cluster 3 and Cluster 4 (C3: n=9; C4: n=12) **(A)**. X axis indicates log2 fold change values, calculated using the DESeq2 package, y axis gives corresponding -log10(adjusted p value). P values adjusted using BH method. Genes are labeled where p value < 0.05. Enrichment plots of selected pathways **(B-D)**. Gene set enrichment analysis (GSEA) was performed with the Hallmark gene sets from the GSEA Molecular Signatures Database. Wald statistic was used to rank the genes.  The green curve corresponds to the Enrichment Score curve, which is the running sum of the weighted enrichment score obtained from GSEA software. 

\newpage

```{r S10, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height= 7, fig.width = 12, warning=F, message=F, cache=FALSE}

cytSeq <- dds_smp

#Filter out IGHV genes
cytSeq <- cytSeq[!grepl("IG_", rowData(cytSeq)$biotype),]

#split into 1,2
cytSeq.12 <- cytSeq[,colData(cytSeq)[,"Cluster"] %in% c(1,2)]

#remove patients where IGHV is unknown
cytSeq.12 <- cytSeq.12[,colData(cytSeq.12)[,"IGHV.status"] %in% c("U","M")]

#set order of factors
cytSeq.12$Cluster <- factor(cytSeq.12$Cluster,
                            levels = c("1","2"))

cytSeq.12$IGHV.status <- factor(cytSeq.12$IGHV.status,
                                levels = c("U","M"))

design(cytSeq.12) <- ~ IGHV.status + Cluster

cytSeq.12 <- DESeq(cytSeq.12)

res.ds <-
  results(cytSeq.12, contrast = c("Cluster", 1, 2), tidy = TRUE) %>%
  dplyr::rename(Symbol = "row") %>%
  arrange(pvalue)

#get ensembl ids to Entrez dataframe
ens2entrez <-
  rowData(cytSeq.12)[c("entrezgene", "symbol")] %>%
  as.data.frame() %>%
  rownames_to_column("ENSEMBL") %>%
  tibble::as_tibble()



#Join
res.ds <- left_join(res.ds, ens2entrez, by=c("Symbol"="ENSEMBL"))

sgncuoff <- 1.4

volPlot <-
  res.ds %>%
  filter(!is.na(padj), !is.na(symbol), symbol!="") %>%

ggplot( aes(x=log2FoldChange, y=-log10(padj))) +
      annotate("rect", xmin = 0.0, xmax = Inf, ymin = -Inf, ymax = Inf, fill = palreds[1], alpha = 0.2)+
      annotate("rect", xmin = 0.0, xmax = -Inf, ymin = -Inf, ymax = Inf,alpha = 0.2, fill = palblues[1])+
      geom_point(aes(alpha=0.4, colour = ifelse(-log10(padj)>sgncuoff,"black","grey"))) +
      theme(legend.position = "none") +
      geom_label_repel(aes(label=ifelse(-log10(padj)>sgncuoff&abs(log2FoldChange)>2,as.character(symbol),''),
                           colour=ifelse(log2FoldChange>0, palreds[2], palblues[4])),
                       size=3.5, max.overlaps=30)+
      scale_colour_manual(values=c(palblues[2], palreds[4], "black", "grey"))+
      t2 +
      labs(title="Cluster 1 versus 2", x = "Log2FC", y = "-log10(adj p-val)", color = "Significance")+
      guides(color="none", alpha="none")


volPlot

# 
# cytSeq[,colData(cytSeq)[,"Cluster"] %in% c(1)]
# cytSeq[,colData(cytSeq)[,"Cluster"] %in% c(2)]
```

## Appendix Figure 10. Comparison of gene expression between Cluster 1 and 2.
Volcano plot of differentially expressed genes between Cluster 1 and Cluster 2 (C1: n=17; C2: n=11). X axis indicates log2 fold change values, calculated using the DESeq package, y axis gives corresponding -log10(adjusted p value). P values adjusted using BH method. Genes are labeled where p value < 0.05.

\newpage

```{r S11,fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width=12, fig.height=4, message=FALSE, warning=FALSE}
load("../../data/fig3_data_for_lassoplots.RData")


# Get predictor profiles for stimuli of interest

heatMaps_cyt <- lassoPlot3Appendix(dataResult[c("Resiquimod","sCD40L+IL-4")] ,  # 
                          geneMatrix.complete, #use gene matrix for heatmap 
                          viabMatrix, #use viab matrix for scatter plot
                          freqCut = 0.75, #coefficients should be selected in <75% of bootstrapped model files
                          coefCut = 0.02) #no minimum value for coefficients 


# Assemble legend for heatmaps

#G = gene mutations, I = IGHV, M = Methylation Cluster
legendFor = c("G", "I", "M")

#assign colours for I, G and M
coldef<-list()
coldef["I"] <- list(c(offwhite,"#707372")) #U-CLL and M-CLL
coldef["M"] <- list(c(offwhite, "#707372", "#A8A99E")) #IP, HP, LP
coldef["G"] <- list(c(offwhite, "#373A36")) #WT and Mutated


legends = makelegends(legendFor=c("G","I","M"),coldef)



grid.arrange( grobs = heatMaps_cyt[1], ncol =1)
grid.arrange( grobs = heatMaps_cyt[2], ncol =1)

# str(geneMatrix.complete)
```

## Appendix Figure 11. Predictor profiles to represent gene - stimulus associations. 
Stimuli which are not shown in Figure 3B but have predictors >0.02 assigned by multivariate analysis using Gaussian linear modelling with L1-penalty are shown. Within the plots, the bar plots on left indicate size and sign of coefficients assigned to the named predictors. Positive coefficients indicate higher viability after stimulation if the feature is present. Scatter plots indicate log(viability) values, in order of magnitude, for each individual sample. Heatmaps show status for each of the genetic predictors for corresponding sample in scatter plot (n=128). 


\newpage



```{r S12, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width=10, fig.height=7, warning=F}

#VST
vsd <- vst(dds_smp, blind=TRUE)

#rename transformed counts
assayNames(vsd)<-"transformedcounts"

ColData<-as.data.frame(colData(dds_smp))

tidy_RNA_data<-
  vsd%>%
  tidy()%>%
  filter(gene%in%cytReceptors$ENSEMBLID)%>%
  dplyr::rename(PatientID=sample)%>%
  left_join(cyt_and_receptors, by=c("gene"="ENSEMBLID"))

RNA_Screen_data <-
  df %>%
  filter(Drug=="DMSO", Cytokine%in%c("Resiquimod","IL-4" ,"TGF-b1","IL-1b",
                                     "Interferon gamma","SDF-1a","sCD40L",
                                     "sCD40L+IL-4","soluble anti-IgM","CpG ODN",
                                     "IL-6","IL-10","IL-21","anti-IgM Beads",
                                     "HS-5 CM","IL-15","BAFF","IL-2")) %>%
  filter(PatientID %in% unique(tidy_RNA_data$PatientID)) %>%
  select(PatientID, Cytokine, Normalized) %>%

  left_join(tidy_RNA_data, by=c("PatientID", "Cytokine"))

Cytokine_Receptor_combinations <-
  RNA_Screen_data %>%
  group_by(Cytokine, Receptor)%>%
  summarize(.groups = "keep") %>%
  filter(!is.na(Receptor))

Cytokine_Receptor_combinations$p.value<-as.numeric(NA)
Cytokine_Receptor_combinations$pearson_R<-as.numeric(NA)

for(i in 1:nrow(Cytokine_Receptor_combinations)) {

  x=Cytokine_Receptor_combinations[i,]

  Correlation_data_matrix<-RNA_Screen_data%>%
    mutate(Cytokine_Receptor=paste0(Cytokine,":", Receptor))%>%
    filter(Cytokine_Receptor==paste0(x[1],":",x[2]))%>%
    select(Normalized, value)%>%
    as.matrix()

  correlation_res<-cor.test(Correlation_data_matrix[,1],Correlation_data_matrix[,2], method = "pearson")

  Cytokine_Receptor_combinations[which(Cytokine_Receptor_combinations$Cytokine==unlist(x[1])&Cytokine_Receptor_combinations$Receptor==unlist(x[2])),]$p.value=correlation_res$p.value
  Cytokine_Receptor_combinations[which(Cytokine_Receptor_combinations$Cytokine==unlist(x[1])&Cytokine_Receptor_combinations$Receptor==unlist(x[2])),]$pearson_R<-correlation_res$estimate

}

Cytokine_Receptor_combinations <- Cytokine_Receptor_combinations%>%
  mutate(adj.p.value=p.adjust(p.value, method="fdr"))


colnames(Cytokine_labels) <- c("Cytokine", "Cytlabel")
Cytokine_Receptor_combinations <- left_join(Cytokine_Receptor_combinations, Cytokine_labels, by = "Cytokine")


Cytokine_Receptor_combinations$Receptor[grepl("IL-10 R alpha",Cytokine_Receptor_combinations$Receptor)]<-"IL10 R \u03B1"

Cytokine_Receptor_combinations$Receptor[grepl("IL-10 R beta",Cytokine_Receptor_combinations$Receptor)]<-"IL10 R \u03B2"

Cytokine_Receptor_combinations$Receptor[grepl("IL-15 R alpha",Cytokine_Receptor_combinations$Receptor)]<-"IL15 R \u03B1"

Cytokine_Receptor_combinations$Receptor[grepl("IL-2 R beta",Cytokine_Receptor_combinations$Receptor)]<-"IL2 R \u03B2"

Cytokine_Receptor_combinations$Receptor[grepl("IL-2 R gamma",Cytokine_Receptor_combinations$Receptor)]<-"IL2 R \u03B3"


Cytokine_Receptor_combinations$Receptor[grepl("IL-1 R1",Cytokine_Receptor_combinations$Receptor)]<-"IL1 R1"

Cytokine_Receptor_combinations$Receptor[grepl("IL-2 R alpha",Cytokine_Receptor_combinations$Receptor)]<-"IL2 R \u03B1"

Cytokine_Receptor_combinations$Receptor[grepl("IL-13 R A1",Cytokine_Receptor_combinations$Receptor)]<-"IL13 R A1"

Cytokine_Receptor_combinations$Receptor[grepl("IFN-gamma R2",Cytokine_Receptor_combinations$Receptor)]<-"IFN \u03B3 R2"

Cytokine_Receptor_combinations$Receptor[grepl("IL-6 R",Cytokine_Receptor_combinations$Receptor)]<-"IL6 R"

Cytokine_Receptor_combinations$Receptor[grepl("IFN gamma R1",Cytokine_Receptor_combinations$Receptor)]<-"IFN \u03B3 R1"

Cytokine_Receptor_combinations$Receptor[grepl("IL-4 R alpha",Cytokine_Receptor_combinations$Receptor)]<-"IL4 R \u03B1"



CR_volcanoplot <-
  ggplot(Cytokine_Receptor_combinations,
         aes(x=pearson_R,
             y=-log10(adj.p.value)))+

  geom_point(data=filter(Cytokine_Receptor_combinations, pearson_R>(-0.3)&pearson_R<0.3), color="grey")+
  geom_point(data=filter(Cytokine_Receptor_combinations, pearson_R<(-0.3)|pearson_R>0.3), color="red")+
   geom_label_repel(data=filter(Cytokine_Receptor_combinations, pearson_R<(-0.3)|pearson_R>0.3), nudge_x = 0.07, color="red", aes(label=paste0(Cytlabel, " | ", Receptor)), size=5)+
  geom_label_repel(data=filter(Cytokine_Receptor_combinations, pearson_R<(-0.1)|pearson_R>0.1, !(pearson_R<(-0.3)|pearson_R>0.3)), nudge_x = 0.07, color="grey", aes(label=paste0(Cytlabel, " | ", Receptor)), size=5)+
  coord_cartesian(xlim=c(-0.45,0.45))+
  geom_vline(xintercept = -0.3, linetype=3)+
  geom_vline(xintercept = 0.3, linetype=3)+
  theme_minimal()+
  scale_x_continuous(breaks = c(-0.3,0,0.3))+
  xlab("Pearson Correlation Coefficient")+
  ylab("-log10(BH adjusted p-value)")+
  t2

CR_volcanoplot



```

## Appendix Figure 12. Correlation of stimuli response and receptor expression.
The effects of the microenvironmental stimuli on viability were largely independent of the expression of the corresponding stimuli receptors. Volcano plot depicts Pearson correlation coefficients against corresponding BH-adjusted p values, for the correlation between natural logarithm of the relative viability and vst RNA counts of corresponding stimuli receptor (n=49). RNA counts taken from RNA-Sequencing of untreated matched CLL patient sample. Only viability  after treatment with Resiquimod correlated with receptor expression (R>0.3).

\newpage

```{r S13,fig.path=plotDir, dev=c("png", "cairo_pdf"),  fig.height=7, fig.width=12}
tri12.labs <- c("0" = "Non-\ntrisomy 12", "1" = "trisomy 12")
ighv.labs <- c("U" = "IGHV-U", "M" =  "IGHV-M")

#filter for Resiquimod-only treatment, only show patients who are annoyated for IGHV and Trisomy12
plotTab <- left_join(df, patMeta, by = "PatientID") %>% 
  dplyr::filter(drugC %in% c("DMSO:None", "Ibrutinib:High"),
                Cytokine=="Resiquimod",
                !is.na(trisomy12),
                !is.na(IGHV.status))


 
  plotTab %>%
         
  ggplot(aes(x = DCK,
             y=Log,
             color=(trisomy12)))+
  geom_hline(yintercept = 0)+
  geom_boxplot()+
  geom_beeswarm(cex=1.5) +
  guides(color="none", shape="none")+
  facet_grid(~IGHV.status + trisomy12, labeller = labeller(trisomy12 = tri12.labs, IGHV.status = ighv.labs))+
 
  
  stat_compare_means(method = "t.test",
                     label.x.npc = "center", 
                     comparisons = list( c(1,2)),
                     size=5)+
  xlab("") +
  ylab("Log(Viability)") +
  ggtitle("Resiquimod + Ibrutinib") +
  scale_color_manual(values=c(palblues[3], palreds[8])) + 
  t1+
  theme(axis.text.x = element_text(angle = 45, vjust =1))+ 
  coord_cartesian(ylim = c(-2.1, 2.5), clip="off")+
  scale_x_discrete(labels=c("DMSO:Resiquimod"="Resiquimod",
                            "Ibrutinib:Resiquimod"="Ibrutinib + Resiquimod"))
  


rm(plotTab)


```

## Appendix Figure 13. BCR inhibition by ibrutinib counters the protective effect of TLR stimulation in all genetic subgroups of IGHV and trisomy 12.
Natural logarithm of the relative viability after treatment with Resiquimod with and without ibrutinib (n=146). Faceted by IGHV mutation status and trisomy 12.

\newpage

```{r S14, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 8, fig.width = 10}

plotTab <- diffTF_small %>%  filter(!is.na(pvalueAdj))

#make naming consistent with text
plotTab$TF <- gsub("SPI1", "PU1", plotTab$TF)

atac.plot =
  ggplot(data = plotTab,
         aes(x = weighted_meanDifference, y = -log10(pvalueAdj+0.000001))) +

  geom_point(aes(alpha=0.4, colour = ifelse(-log10(pvalueAdj) > 2,"black","grey")),size=3) +
  
  geom_label_repel(aes(label=ifelse(-log10(pvalueAdj) > 4&abs(weighted_meanDifference)>0.15,
                                    as.character(TF),''),
                       colour=ifelse(weighted_meanDifference>0,
                                     palreds[1],
                                     palblues[1])),
                   size=5, max.overlaps=20)+
  scale_colour_manual(values=c(palblues[3], palreds[8], "black", "grey")) +
  t2 +
  ggtitle("ATAC seq: Tri12 v WT") +
  xlab("Change in inferred TF activity") +
  guides(color="none", alpha="none") +
  scale_y_continuous(expression("-log(adjusted  "* italic(p)*"-value)"))+
  coord_cartesian(ylim = c(0,6.5))


atac.plot

rm(plotTab)

#check number of upregulated TFs
# nrow(filter(diffTF_small, weighted_meanDifference>0, pvalueAdj <0.05))
```

## Appendix Figure 14. ATACseq comparing trisomy 12 and non-trisomy 12 CLL samples.
Volcano plot shows change in inferred TF activity (x axis) against adjusted p-values (y axis) for  trisomy 12 (n = 2) versus non-trisomy 12 samples (n = 2). Inferred TF activity calculated using the diffTF package, and measured as weighted mean difference. p-values are obtained through diffTF in analytic mode and transformed by the Benjamini-Hochberg procedure. TFs are labeled where absolute weighted mean difference > 0.15, and transformed p-value < 0.01.


\newpage

```{r S15, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 6, fig.width = 10, eval=FALSE}

plotTab <- diffTF_Beekman %>%  filter(!is.na(pvalueAdj))

#make naming consistent with text
plotTab$TF <- gsub("SPI1", "PU1", plotTab$TF)

atac.plot =
  ggplot(data = plotTab,
         aes(x = weighted_meanDifference, y = -log10(pvalueAdj+0.000001))) +

  geom_point(aes(alpha=0.4, colour = ifelse(-log10(pvalueAdj) > 2,"black","grey")),size=3) +
  geom_label_repel(aes(label=ifelse(-log10(pvalueAdj) > 4&abs(weighted_meanDifference)>0.15,
                                    as.character(TF),''),
                       colour=ifelse(weighted_meanDifference>0,
                                     palreds[1],
                                     palblues[1])),
                   size=7, max.overlaps=20)+
  scale_colour_manual(values=c(palblues[3], palreds[8], "black", "grey")) +
  t2 +
  ggtitle("ATAC seq: Tri12 v WT") +
  xlab("Change in inferred TF activity") +
  guides(color="none", alpha="none") +
  scale_y_continuous(expression("-log(adjusted  "* italic(p)*"-value)"))

atac.plot

rm(plotTab)

#check number of upregulated TFs
# nrow(filter(diffTF_small, weighted_meanDifference>0, pvalueAdj <0.05))
```

## Appendix Figure 15. ATACseq comparing trisomy 12 and non-trisomy 12 CLL samples.
Volcano plot shows change in inferred TF activity (x axis) against transformed p-values (y axis) for  trisomy 12 (n = XXX) versus non-trisomy 12 samples (n = XXX). Inferred TF activity calculated using the diffTF package, and measured as weighted mean difference. p-values are obtained through diffTF in analytic mode and transformed by the Benjamini-Hochberg procedure. TFs are labeled where absolute weighted mean difference > 0.15, and transformed p-value < 0.01. Data from Beekman et al. 2018. 

\newpage


```{r S16, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 15, fig.width = 12}

plotTab <- diffTF_wochr12 %>%  filter(!is.na(pvalueAdj))

#make naming consistent with text
plotTab$TF <- gsub("SPI1", "PU1", plotTab$TF)

atac.plot =
  ggplot(data = plotTab,
         aes(x = weighted_meanDifference, y = -log10(pvalueAdj + 0.000001))) +

  geom_point(aes(alpha=0.4, colour = ifelse(-log10(pvalueAdj) > 2,"black","grey")),size=3) +
  geom_label_repel(aes(label=ifelse(-log10(pvalueAdj) > 4&abs(weighted_meanDifference)>0.15,
                                    as.character(TF),''),
                       colour=ifelse(weighted_meanDifference>0,
                                     palreds[1],
                                     palblues[1])),
                   size=7, max.overlaps=20)+
  scale_colour_manual(values=c(palblues[3], palreds[8], "black", "grey")) +
  t2 +
  ggtitle("ATAC seq: Tri12 v WT") +
  xlab("Change in inferred TF activity") +
  guides(color="none", alpha="none") +
  scale_y_continuous(expression("-log(adjusted  "* italic(p)*"-value)"))+
  coord_cartesian(ylim = c(0,6.5))




rm(plotTab)

plotTab <- diffTF_wochr12_big %>%  filter(!is.na(pvalueAdj))

#make naming consistent with text
plotTab$TF <- gsub("SPI1", "PU1", plotTab$TF)

atac.plot.big =
  ggplot(data = plotTab,
         aes(x = weighted_meanDifference, y = -log10(pvalueAdj + 0.000001))) +

  geom_point(aes(alpha=0.4, colour = ifelse(-log10(pvalueAdj) > 1,"black","grey")),size=3) +
  geom_label_repel(aes(label=ifelse(-log10(pvalueAdj) > 1&abs(weighted_meanDifference)>0.05,
                                    as.character(TF),''),
                       colour=ifelse(weighted_meanDifference>0,
                                     palreds[1],
                                     palblues[1])),
                   size=7, max.overlaps=25)+
  scale_colour_manual(values=c(palblues[3], palreds[8], "black", "grey")) +
  t2 +
  ggtitle("ATAC seq: Tri12 v WT") +
  xlab("Change in inferred TF activity") +
  guides(color="none", alpha="none") +
  scale_y_continuous(expression("-log(adjusted  "* italic(p)*"-value)"))+
  coord_cartesian(ylim = c(0,1.7))

rm(plotTab)

tp <- theme(plot.tag=element_text(size = 30, face="plain"))

(wrap_elements( atac.plot )+ tp)/
(wrap_elements( atac.plot.big )+ tp) +
plot_annotation(tag_levels = "A", theme = theme(title=element_text(size = 20)))

```

## Appendix Figure 16. ATACseq comparing trisomy 12 and non-trisomy 12 CLL samples, independent of third copy of chromosome 12
Volcano plot shows change in inferred TF activity (x axis) against transformed p-values (y axis) for  trisomy 12 versus non-trisomy 12 samples, for (A) the 2 versus 2 comparison and (B) the 9 versus 42 comparison based on the data from Rendeiro et al. 2016^1^. Inferred TF activity calculated using the diffTF package, and measured as weighted mean difference. Input peak files supplied to diffTF did not include peaks on chromosome 12. p-values are obtained through diffTF in (A) analytic mode and (B) permutation mode, adjusted by the Benjamini-Hochberg procedure. TFs are labelled in (A) where absolute weighted mean difference > 0.15, and transformed p-value < 0.01 and in (B) where absolute weighted mean difference > 0.05, and transformed p-value < 0.1.


\newpage

```{r Spib_binding_sites}

#filter SPIB binding sites  for those that are sig between two conditions, and show a sig Log2FC
TFBS <- BIGCLL52_SPIB %>% dplyr::filter(abs(l2FC) >1, pval_adj <0.1)

#prepare to make this into a genomic ranges object for plotting
TFBS <- TFBS %>% 
        #remove chr string
        mutate_at(vars(TFBSID), list(~as.character(gsub("chr.*:", "", .)))) %>% 
        #separate into start and end
        separate(TFBSID, c("start", "end"))

#Convert to a genomic ranges object
peak <- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

```

```{r S17, fig.path=plotDir, dev=c("cairo_pdf"), fig.height = 15, fig.width = 10, warning = FALSE}

#get a dataframe of all SPIB binding sites that show sig Log2FC
peak_df <- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj)

#make factors / numeric
chr.sorted <- unique(sortChrName(peak_df$chr))
peak_df$chr <- factor(peak_df$chr, levels = chr.sorted)
peak_df$start <- as.numeric(peak_df$start)
peak_df$end <- as.numeric(peak_df$end)


chr_length<-filter(human.hg19.genome, V1%in%levels(peak_df$chr)) %>% 
         mutate(V1=factor(V1, levels = rev(levels(peak_df$chr)))) %>% 
          mutate(chromosome=gsub("chr","Chromosome ",V1)) %>% 
  rename(chr_length=V2) %>% 
  select(-V1)

# S18 <- 
  
peak_df%>% 
  mutate(chromosome=gsub("chr","Chromosome ",chr)) %>% 
  left_join(chr_length, by="chromosome") %>%
  mutate(chromosome=factor(chromosome, levels = c(paste0("Chromosome ", 1:22)))) %>% 

ggplot(aes(start/10^6, l2FC))+
  geom_segment(aes(y=0, yend=0, x=0, xend=chr_length/10^6))+
  geom_rect(aes(xmin = start/10^6, ymin = 0, xmax = end/10^6, ymax = l2FC,
                color = ifelse(l2FC<0, palreds[5], palblues[1]), 
                fill= ifelse(l2FC<0, palreds[5], palblues[1]))) +
  scale_fill_manual(values = c(palreds[5], palblues[1]), guide = "none") +
  scale_color_manual(values = c(palreds[5], palblues[1]), guide = "none") +
  facet_grid(chromosome ~ ., scales = "fixed", space = "fixed")+ 
  xlab("Chromosome Size (Mbp)") + 
  ylab("log 2 fold change of TF accessibility") + 
  ggtitle("Change in inferred TF activity for Spi-B binding sites by Chromosome")+
  t2 + 
  theme(strip.text.y = element_text(angle = 360, hjust = 0),
        strip.background.y = element_blank(),
        axis.text.y = element_text(size=10),
        axis.ticks.length.y = unit(3, "pt"),
        panel.grid.major.x=element_blank(),
        panel.spacing = unit(0.1, "lines")) + 
  ylim(-3, 3) 

  

# S18


```

## Appendix Figure 17. Chromosomal locations of Spi-B binding sites that show differential accessibility between trisomy 12 and non-trisomy 12 CLL. 
Log2 fold changes are plotted (y axis) Spi-B binding sites that show an absolute Log2 fold change > 1 and a p value < 0.1, as calculated using the diffTF package for all Spi-B binding sites defined in the HOCOMOCO database. Data shown is the 9 versus 42 comparison based on the data from Rendeiro et al. 2016^1^.


```{r , fig.path=plotDir, dev=c("cairo_pdf"), fig.height = 3, fig.width = 3, warning = FALSE, eval=FALSE}


#get a dataframe of all SPIB binding sites that show sig Log2FC
peak_df <- 
  TFBS %>%
  dplyr::select(TF, chr, start, end, strand, l2FC, pval,  pval_adj)

#make factors / numeric
chr.sorted <- unique(sortChrName(peak_df$chr))
peak_df$chr <- factor(peak_df$chr, levels = chr.sorted)
peak_df$start <- as.numeric(peak_df$start)
peak_df$end <- as.numeric(peak_df$end)

peak_df<-peak_df %>% 
  mutate(chr_num=as.numeric(gsub("chr","",chr)))

chr_length<-filter(human.hg19.genome, V1%in%levels(peak_df$chr)) %>% 
         mutate(V1=factor(V1, levels = rev(levels(peak_df$chr)))) %>% 
          mutate(chr_num=as.numeric(gsub("chr","",V1)))

 ggplot(peak_df)+
  geom_segment(data= chr_length, aes(y=chr_num, yend=chr_num, x=0, xend=V2/10^6))+
  geom_segment(aes(y=chr_num, yend=chr_num-(l2FC/4),  x = start/10^6, xend= end/10^6,
                color = ifelse(l2FC<0, palreds[5], palblues[1]))) +
  scale_fill_manual(values = c(palreds[5], palblues[1]), guide = "none") +
  scale_color_manual(values = c(palreds[5], palblues[1]), guide = "none") +
   scale_y_reverse(breaks=c(1:22))+
  t2+
   theme(panel.grid.major.x=element_blank(),
         axis.title.y = element_blank(),
         axis.title.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y= element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x= element_blank(),
        axis.line.x = element_blank(),
        axis.line.y = element_blank())



```

\newpage

```{r }
#First check whether script for Figure 6 has been run, and if  fig6_data_for_supplement.RData exists.
#fig6_data_for_supplement.RData object contains Lasso_Plots_Fig6, a list of TableGrobs and coeff.long.mat_Fig6, a dataframe (840*3) of drug - stimulus - gene combinations and their coefficients

#generate coefficient matrix and predictor profiles if prework has not been done
if(file.exists("../../data/fig6_data_for_supplement.RData")){
load("../../data/fig6_data_for_supplement.RData")

}else{
#####ARRANGE DATA
#only high concentrations of drugs
df.supp11 <-
  dplyr::filter(df, Drug_Concentration %in% c("None","High")) %>%
  dplyr::select(PatientID, Drug, Cytokine, DCK, Log)

#rename columns
df.supp11 <-
  plyr::rename(df.supp11,
               c("Drug"="treatment_drug",
                 "Cytokine"="treatment_cytokine",
                 "Log"="Viability"))

#lists for drug and cytokine treatments
cyt.list <- unique(df.supp11$treatment_cytokine) %>% setdiff("No Cytokine")
drug.list <- unique(df.supp11$treatment_drug) %>% setdiff("DMSO")




#####GENERATE FEATURE MATRIX
#select features from patient meta file
geneMatrix <- dplyr::select(patMeta, -c(gender:treatment, Methylation_Cluster)) %>%


  #adjust IGHV levels to 1 and 0
  mutate(IGHV.status = ifelse(is.na(IGHV.status), NA,
                              ifelse(IGHV.status == "M", 1, 0))) %>%

  #change factors to characters and then to numeric
  mutate_if(is.factor, as.character) %>%
  mutate_at(vars(-PatientID), as.numeric) %>%

  #convert to matrix format
  data.frame() %>%
  column_to_rownames("PatientID") %>%
  as.matrix()


#Remove genes with higher than 20% missing values
geneMatrix <- geneMatrix[,colSums(is.na(geneMatrix))/nrow(geneMatrix) <= 0.2]

#Filter for patients with complete data
geneMatrix.complete <- geneMatrix[complete.cases(geneMatrix),]
nrow(geneMatrix.complete)

#Combine KRAS, NRAS and BRAF mutations into a single column
##Add Ras_raf column
Ras_Raf <- matrix(NA, nrow = nrow(geneMatrix.complete), ncol = 1)
colnames(Ras_Raf) <- "KRAS,\nNRAS,\nBRAF"
geneMatrix.complete <- cbind(geneMatrix.complete, Ras_Raf)

#Add a 1 where any of KRAS, NRAS or BRAF are mutated
geneMatrix.complete[,"KRAS,\nNRAS,\nBRAF"] <- ifelse(geneMatrix.complete[,"KRAS"]==1,1,
		                                        ifelse(geneMatrix.complete[,"BRAF"]==1,1,
	                	                          ifelse(geneMatrix.complete[,"NRAS"]==1, 1, 0)))

#Remove individual KRAS, NRAS and BRAF columns
geneMatrix.complete <- geneMatrix.complete[, !colnames(geneMatrix.complete) %in%  c("KRAS", "NRAS", "BRAF")]




# GENERATE RESPONSE MATRIX

#create a list of drug - cytokine combinations
combos <- expand.grid(drug.list, cyt.list) %>% mutate(combination = paste(Var1, Var2, sep = ":")) %>% select(combination)


#create object to store linear model fit
fit <- vector(mode = 'list', length = length(combos$combination))
names(fit) <- combos$combination

#create an object to store coefficients from fit
coefficients <- vector(mode = 'list', length = length(combos$combination))
names(coefficients) <- combos$combination

#define drugs and cytokines to run the model for
for(x in drug.list){
  for(y in cyt.list){

    #get data for given drug and stimulis, and matching pateints to feature matrix
    modelTab <- filter(df.supp11,
                       PatientID %in% rownames(geneMatrix.complete),
                       treatment_drug%in% c(x, "DMSO"),
                       treatment_cytokine %in% c(y, "No Cytokine") )

    #define the base level per treatment_type as the "no"-treatment
    modelTab$treatment_drug <- as.factor(modelTab$treatment_drug)
    modelTab$treatment_cytokine <- as.factor(modelTab$treatment_cytokine)
    modelTab$PatientID <- as.factor(modelTab$PatientID)

    modelTab$treatment_cytokine %<>% relevel("No Cytokine")
    modelTab$treatment_drug %<>% relevel("DMSO")
    modelTab$PatientID %<>% relevel("Pat_001")

    # fit linear model, with interaction
    nam <- paste(x,y, sep = ":")
    fit.lm <- lm(Viability ~ treatment_drug * treatment_cytokine * PatientID, modelTab)
    fit[[nam]] <- fit.lm

    #extract coefficients and p values and put into a dataframe
    coeffdf <- summary(fit.lm)$coefficients[,1] %>% as.data.frame()

    #process coeffdf
    setDT(coeffdf, keep.rownames = TRUE)[]
    colnames(coeffdf) <- c("treatment", "beta")

    #filter out non-drug:cytokine:patient coefficients
    coeffdf <- filter(coeffdf,
                      grepl('treatment_drug.*treatment_cytokine.*PatientID*', treatment))

      #remove treatment
      coeffdf <- coeffdf %>%
        #remove treatment_drug string
        mutate_at(vars(treatment), list(~as.character(gsub("treatment_drug", "", .)))) %>%
        #remove treatment_cytokine string
        mutate_at(vars(treatment), list(~as.character(gsub("treatment_cytokine", "", .)))) %>%
        #remove PatientID string
        mutate_at(vars(treatment), list(~as.character(gsub("PatientID", "", .))))

      #split up into sperate drug, cytokine and patient columns, by colons
      coeffdf <- data.frame(coeffdf, do.call(rbind, strsplit(coeffdf$treatment, split = ":", fixed = TRUE)))

      #select columns of interest and rename
      coeffdf <- coeffdf[, c("beta", "X1", "X2", "X3")]
      colnames(coeffdf) <- c("beta","Drug", "Cytokine", "PatientID")

      #store in coefficients object
      coefficients[[nam]] <- coeffdf

  }
}

#bind together all coefficients for all drug:stimulus combinations
coefficients.all <- rbindlist(coefficients)

#make matrix
betaMatrix <-
  #add column with drug and stimulus names
  dplyr:: mutate(coefficients.all,
                 drugCytokine = paste0(Drug," + ",Cytokine)) %>%
  #remove single treatment columns
  dplyr::select(-Cytokine, -Drug) %>%

  spread(key = PatientID, value = beta) %>%
  data.frame() %>%
  remove_rownames() %>%
  column_to_rownames("drugCytokine")

#make sure the sample order is the same as the geneMatrix
geneMatrix.complete <- geneMatrix.complete[ rownames(geneMatrix.complete) != "Pat_001",]
betaMatrix <- betaMatrix[,rownames(geneMatrix.complete)]

#check there are no NAs
which(is.na(betaMatrix))
which(betaMatrix == 0)


# Run lasso - regularised regression
#define object to hold outputs from runGlm()
dataResult <- list()

#for each drug + stimulus combination:
for (x in rownames(betaMatrix)){

      #prepare input and response matrices
      y <- unlist(betaMatrix[x,])
      X <- geneMatrix.complete

      #fit the model and find optimal value of lamba
      cvglmfit <- runGlm(X, y, method="lasso", repeats=30, folds=3)
      dataResult[[x]] <- cvglmfit

}

#MAKE Lasso_Plots_Fig6 OBJECT, CONTAINING ALL PREDICTOR PROFILES
#run lassoPlot function to generate predictor profiles, with no minimum coefficient value
heatMaps <- lassoPlot(dataResult, geneMatrix.complete, betaMatrix, freqCut = 0.9, coefCut = 0.0)

Lasso_Plots_Fig6 <- heatMaps[!is.na(heatMaps)]



#MAKE coeff.long.mat_Fig6 OBJECT, CONTAINING DATAFRAME TO PLOT HEATMAP OF MODEL COEFFICIENTS

#Set up vector to hold coeff values
barValues <- vector(mode="list", length=length(dataResult))
names(barValues) <- names(dataResult)

#Set cut offs
coefCut <- 0.0 #no minimum value of coefficient
freqCut <- 0.9 #coefficient must be selected in 90% of bootstrapped repeats

lassoOut <- dataResult

#for each drug - stimulus combination
for (seaName in names(lassoOut)) {

    #get the result from DataResult for given drug:cytokine interaction, extract coefficient matrix and find row means
    barValue <- rowMeans(lassoOut[[seaName]]$coefMat)

    #check number of occurrences of each coefficient in bootstrapped repeats
    freqValue <- rowMeans(abs(sign(lassoOut[[seaName]]$coefMat)))

    for(nam in names(barValue)){
      #add NA if coefficient value is below thresholds
      if(abs(barValue[nam]) < coefCut | freqValue[nam] < freqCut) {
          barValue[nam] <- NA }
    }
    #add set of coefficients to list
    barValues[[seaName]] <- barValue
}

#bind coefficients together for all interactions
coeff.mat <- do.call(rbind,barValues)

#remove rows where all values are 0
coeff.mat <- coeff.mat[ rowSums(!is.na(coeff.mat)) > 0,]

#remove columns where all values are 0
coeff.mat <- coeff.mat[ ,colSums(!is.na(coeff.mat)) > 0]

#make any NAs = 0
coeff.mat[is.na(coeff.mat)] <- 0

#cluster coeff.mat
fit <-
  coeff.mat %>%
  dist() %>%
  hclust()

order_comb <- rownames(coeff.mat)[fit$order]

#Put matrix into long format for ggplot
coeff.long.mat <- coeff.mat %>%
  as.data.frame() %>%
  tibble::rownames_to_column("int") %>%
  gather(key = "gene", value = "coeff", -int)

#Check number of combinations affected by all genetic alterations
coeff.long.mat %>%
  dplyr::filter(coeff!=0) %>%
  dplyr::group_by(int) %>%
  dplyr::summarize()

coeff.long.mat %>%
  dplyr::filter(coeff!=0) %>%
  dplyr::group_by(gene) %>%
  dplyr::count() %>%
  dplyr::arrange(desc(n))

#get top 8 coefficients
order_alt <-
  coeff.long.mat %>%
  dplyr::filter(coeff!=0) %>%
  dplyr::group_by(gene) %>%
  dplyr::count(sort=T) %>%
  dplyr::select(gene) %>%
  unlist() %>%
  .[1:8]

#Number of combinations affected by Top 8 genetic alterations
coeff.long.mat %>%
  dplyr::filter(coeff!=0, gene%in%order_alt) %>%
  dplyr::group_by(int) %>%
  dplyr::summarize()

coeff.long.mat_Fig6 <- coeff.long.mat

}
```

```{r s18, fig.path=plotDir, dev=c("cairo_pdf"), fig.height=19.5, fig.width=16}
colnames(Cytokine_labels) <- c("Stimulus", "label")

coeff.long.mat_Fig6 %>%
  tidyr::separate(int, sep=" \\+ ", into=c("Drug", "Stimulus"), remove =FALSE)%>%
  left_join(Cytokine_labels, by = "Stimulus") %>%
  dplyr::filter(coeff!=0) %>%


  ggplot(aes(y=Drug, x=gene))+
  geom_tile(aes(fill=coeff),color = "white")+
  scale_fill_gradientn(colors=c(rep(palblues[1:4],each=2),"white", rep(palreds[5:8], each=2)),  limits=c(-0.8,.8))+
  t1+
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.ticks = element_blank(),
        axis.ticks.x = element_blank(),
        panel.background = element_rect(color = "black", fill=NA),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        legend.key.height = unit(0.5, "cm"),
        legend.key.width = unit(2, "cm"),
        legend.title = element_text(face='bold', hjust = 0, size=fontsize+2),
        legend.position = "bottom",
        legend.key = element_blank(),
        legend.text = element_text(size=fontsize),
        legend.background = element_rect(color = NA),
        strip.text.y.left = element_text(angle = 0, face="bold", size = 22),
        strip.background = element_blank())+
  labs(fill = expression(paste(beta, "-int")))+
  scale_y_discrete(position = "right",limits=rev)+
  scale_x_discrete(labels=c("KRAS,\nNRAS,\nBRAF"="RAS/RAF"))+
   facet_grid(label~., scales = "free_y",  space="free_y", switch = "both")

```

## Appendix Figure 18. Genetic predictors of drug - stimulus interactions.
Based on the drug - stimuli viability data of 137 CLL patient samples with complete genetic annotation we fitted a sample specific linear model. 
Heatmap depicting overview of genetic predictors of drug - stimulus interactions (each row represents the coefficients from fitting a single multivariate model). Stimuli are shown on left, and corresponding drugs on right. Drugs, stimuli and genetic alterations are alphabetically sorted. Coloured fields indicate that the $\beta_{int}$ for given drug and stimulus is modulated by corresponding genetic feature. Positive coefficients are shown in red, indicating $\beta_{int}$ is more positive for given drug and stimulus combination if the feature is present.

\newpage

```{r S19, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width = 10, fig.height =4,  eval = TRUE}

# Plot predictor profile for Ibrutinib + IL4
 grid.arrange(grobs = Lasso_Plots_Fig6["Ibrutinib + IL-4"], ncol = 1)


```

## Appendix Figure 19. Genetic predictors of the interaction between ibrutinib and IL4.
Predictor profile depicting genetic features that modulate the size of $\beta_{int}$ for ibrutinib and IL4. To generate predictor profile, linear model in Eqn. (1) was fitted in a sample - specific manner, to calculate drug-stimulus interaction coefficients ($\beta_{int}$) for each patient sample (n=137). Ranked patient-specific $\beta_{int}$ values are shown in lower scatter plot. Associations between the size of $\beta_{int}$ and genetic features were identified using multivariate regression with L1 (lasso) regularisation, with gene mutations and IGHV status as predictors, and selecting coefficients that were chosen in >90% of bootstrapped model fits. The horizontal bars  on left show the size of fitted coefficients assigned to genetic features. Matrix above scatter plot indicates patient mutation status for the selected genetic features. Matrix fields correspond to points in scatter plot (ie patient data is aligned), to indicate how the size of $\beta_{int}$ varies with selected genetic feature.

\newpage

```{r  S20,fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width=8}
S20 <-
IbrStat6iComb %>%
  mutate(trisomy12=as.character(trisomy12)) %>% 
  
  ggplot( aes(x = interaction(Drug_2, Drug_1, Cytokine), y = Log))+
  geom_boxplot(outlier.shape = NA)+
  geom_beeswarm(aes(color = IGHV.status, shape = trisomy12), cex = 2) +
  geom_hline(yintercept = 0) +
  t2 +
  theme(legend.key = element_blank()) +
  ylab("\n\n\nViability") +
  ggtitle(paste0("")) +
  xlab("") +
  ylab("Log(Viability)") +
  scale_color_manual(values = c("M"=palblues[1], "U"=palreds[8]), na.value="#7a7c80") +
  scale_x_discrete(labels=c("DMSO.DMSO.IL-4" = "IL4",
                            "Ibrutinib.DMSO.IL-4" = "IL4\nIbrutinib",
                            "DMSO.STAT6i.IL-4" = "IL4\nSTAT6i",
                            "Ibrutinib.STAT6i.IL-4" = "IL4\nIbrutinib\nSTAT6i"))+

  labs(shape = "trisomy 12", color = "IGHV status")

S20

# IbrStat6iComb %>% 
#   group_by(PatientID) %>% 
#   summarise()
```

## Appendix Figure 20. STAT6 dependency of IL4 signaling.
Natural logarithm of the relative viability after treatment with IL4 in combination with ibrutinib, the STAT6 inhibitor AS1517499, and both (n=16). The effects observed with IL4 stimulation are dependent on STAT6 activation. Addition of the STAT6 inhibitor AS1517499 could revoke the effect on baseline viability as well as drug induced toxicity. 

\newpage

# References  
1. Rendeiro, A., Schmidl, C., Strefford, J. et al. Chromatin accessibility maps of chronic lymphocytic leukaemia identify subtype-specific epigenome signatures and transcription regulatory networks. Nat Commun 7, 11938 (2016).

2. Lu, J., Cannizzaro, E., Meier-Abt, F. et al. Multi-omics reveals clinically relevant proliferative drive associated with mTOR-MYC-OXPHOS activity in chronic lymphocytic leukemia. Nat Cancer 2, 853â864 (2021).


```{r}
knitr::knit_exit()
```


```{r, fig.height=6, fig.width=8}

df_patMeta %>% 
  left_join(Heatmap_clusters, by = "PatientID") %>% 
  filter(Drug%in%c("DMSO", "BAY-11-7085"), Cytokine%in%c("No Cytokine", "BAFF"), Drug_Concentration%in%c("None", "High"), !(Drug=="DMSO"&Cytokine=="No Cytokine")) %>% 
  mutate(Cluster=as.factor(Cluster),
         DCK=factor(DCK, levels=c("DMSO:BAFF", "BAY-11-7085:No Cytokine", "BAY-11-7085:BAFF" ))
         ) %>% 
  
  ggplot(aes(DCK, Log)) +
  
  geom_hline(yintercept = 0, linetype=2, color="black")+
  
  geom_boxplot(aes(fill=Cluster))+
  # geom_beeswarm(aes(color=Cluster))+

  # lemon::geom_pointline(aes(group=PatientID), size=1,   
                         # colour="#003DA5", alpha=0.2)+
  facet_wrap(~Cluster, scales = "fixed", nrow = 1, labeller = labeller(Cluster = c("1" = "Cluster 1", "2" = "Cluster 2","3" = "Cluster 3","4" = "Cluster 4")))+
    scale_color_manual(values=colors[1:4])+
  t1+ theme(strip.background =element_blank())+
  guides(fill="none")+
  scale_x_discrete(labels=c("DMSO:BAFF"="BAFF",
                            "BAY-11-7085:No Cytokine"="BAY-11-7085",
                            "BAY-11-7085:BAFF"="BAY-11-7085 + BAFF"))+
  ylab("Log(Viability)")+
  xlab("Treatment")+
  coord_cartesian(clip="off")+
  stat_compare_means(comparisons = list(c(1,3)), paired = TRUE)

```



