---
title: 'CLL Cytokine Screen 2021: Figure 4'
author: "Holly Giles and Peter Bruch"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
      code_folding: "hide" 
---

# Figure 4
In this sub-vignette we present the analysis and source code for figure 5.  This sub-vignette can be built along with all other sub-vignettes, by running CLLCytokineScreen2021.Rmd. 


## Set up 
```{r setup4}

set.seed(1996)

```

Load libraries
```{r loadLibraries4, cache = FALSE, message = FALSE, warning = FALSE}

library(plyr)
library(clusterProfiler)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(ChIPseeker)
library(genomation)
library(ggbeeswarm)
library(ggpubr)
library(ggplot2)
library(patchwork)
library(msigdbr)
library(png)
library(cowplot)
library(tidyverse)
library(Hmisc)
library(pheatmap)
library(RColorBrewer)
library(org.Hs.eg.db)

```

Set plot directory
```{r plotDir4, message=FALSE}
plotDir = ifelse(exists(".standalone"), "", "../../inst/figs/")
if(plotDir!="") if(!file.exists(plotDir)) dir.create(plotDir)
```


## Load data
Load raw files
```{r loadData4, message=FALSE}

#df: tibble containing all screening viability data
load( "../../data/df.RData")

#patMeta: tibble containing all patient genetic data
load( "../../data/patMeta.RData")


#diffTF_large: tibble containing diffTF weighted mean difference values  for trisomy 12 versus non-trisomy 12 CLL ATACseq data
load( "../../data/diffTF_large.RData")

#diffTF_ibet: tibble containing diffTF weighted mean difference values  for 4 ibet versus 4 control treated CLL ATACseq samples
load( "../../data/diffTF_ibet.RData")


#ChIPseq data for SPIB and PU1 binding in OCILY3 cell line (downloaded from GEO)

#get info on dataset of interest
chipSeq <- getGEOInfo(genome="hg19", simplify=FALSE) %>% dplyr::filter(series_id=="GSE56857")

#Define gsm id for SPIB and PU1 ChIPseq dataset in OCILY3 cell line (DLBCL)
gsm_spib = "GSM1370276"
gsm_pu1 = "GSM1370275"                                                         

#download bed file if not already there                                          
downloadGSMbedFiles(gsm_spib, destDir="../../inst/extdata/")
downloadGSMbedFiles(gsm_pu1, destDir="../../inst/extdata/")

#read the data
#ChIPseq data downloaded from GEO Accession Number: GSE56857
peak.spib <-  genomation ::readBed("../../inst/extdata/GSM1370276_OCILY3_SPIB_GEM_events.bed.gz", track.line = "auto")
peak.pu1 <-  genomation ::readBed( "../../inst/extdata/GSM1370275_OCILY3_PU1_GEM_events.bed.gz", track.line = "auto")

## dds_smp_full: DeSeqDataset containing RNA seq data for all available samples
load( "../../data/dds_smp_full.RData")

## diffTF conversion tables: 
#HOCOMOCO IDs for trisomy 12 versus non-trisomy 12 diffTF run 
tri12_TFs <- read.table(file= "../../inst/extdata/translationTable_hg19.csv",header = TRUE) %>% as_tibble()

#HOCOMOCO IDs for IBET-762 versus control diffTF run 
ibet_TFs <- read.table(file= "../../inst/extdata/translationTable_hg38.csv",header = TRUE) %>% as_tibble()



```


```{r loadData_fromtsv, eval= FALSE}

#df: tibble containing all screening viability data
df <- read.table(file= "../../inst/extdata/df.txt",header = TRUE) %>% as_tibble()

#patMeta: tibble containing all patient genetic data
patMeta <- read.table(file= "../../inst/extdata/patMeta.txt",header = TRUE) %>% as_tibble()


#diffTF_large: tibble containing diffTF weighted mean difference values for 636 TFs, for trisomy 12 versus non-trisomy 12 CLL ATACseq data
diffTF_large <- read_tsv("../../inst/extdata/BIGCLL52.permutations.summary.tsv.gz")

#diffTF_IBET: tibble containing diffTF weighted mean difference values for 636 TFs, for trisomy 12 versus non-trisomy 12 CLL ATACseq data
diffTF_ibet <- read_tsv("../../inst/extdata/IBETvsCTRL.summary.tsv.gz")

#ChIPseq data for SPIB and PU1 binding in OCILY3 cell line (downloaded from GEO)
#get info on dataset of interest
chipSeq <- getGEOInfo(genome="hg19", simplify=FALSE) %>% dplyr::filter(series_id=="GSE56857")

#Define gsm id for SPIB and PU1 ChIPseq dataset in OCILY3 cell line (DLBCL)
gsm_spib = "GSM1370276"
gsm_pu1 = "GSM1370275"                                                         

#download bed file if not already there                                          
downloadGSMbedFiles(gsm_spib, destDir="../../inst/extdata/")
downloadGSMbedFiles(gsm_pu1, destDir="../../inst/extdata/")

#read the data
peak.spib <-  genomation ::readBed("../../inst/extdata/GSM1370276_OCILY3_SPIB_GEM_events.bed.gz", track.line = "auto")
peak.pu1 <-  genomation ::readBed( "../../inst/extdata/GSM1370275_OCILY3_PU1_GEM_events.bed.gz", track.line = "auto")



#Arrangedata
#df
df$Cytokine <- factor(df$Cytokine, levels= c("No Cytokine",
                                              "Resiquimod", 
                                              "IL-4", 
                                              "TGF-b1",
                                              "IL-1b",
                                              "Interferon gamma",
                                              "SDF-1a",
                                              "sCD40L" ,
                                              "sCD40L+IL-4",
                                              "soluble anti-IgM", 
                                              "CpG ODN",
                                              "IL-6",
                                              "IL-10",
                                              "IL-21",
                                              "HS-5 CM", 
                                              "IL-15",
                                              "BAFF",
                                              "IL-2" ))

#patMeta

patMeta <- patMeta %>%mutate_at(vars(-PatientID), as.factor)

```

Process files
```{r processData4}  

df_complete <- df

#get Cytokine only data
df <- dplyr::filter(df, 
                    Drug == "DMSO", 
                    Cytokine != "No Cytokine")
   
```


## Define Aesthetics
```{r defineAesthetics4}

source("../../R/themes_colors.R")

```


## Plot Figures
### Figure 4A
TLR Response stratified by IGHV and trisomy 12
```{r, Fig4A, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 5.5, fig.width = 6}

#filter for Resiquimod-only treatment, only show patients who are annotated for IGHV and Trisomy12
plotTab <- left_join(df, patMeta, by = "PatientID") %>% 
  dplyr::filter(Cytokine=="Resiquimod",
                !is.na(trisomy12),
                !is.na(IGHV.status))

Fig4A =  
 
  plotTab %>%
         
  ggplot(aes(x=interaction(trisomy12, IGHV.status),
                    y=Log,
                    color=(IGHV.status)))+
  geom_hline(yintercept = 0)+
  geom_boxplot()+
  geom_beeswarm(cex=1.5) +
  guides(color="none", shape="none")+
  scale_x_discrete(labels=c("0.M"="IGHV-M\n WT",
                            "0.U"="IGHV-U\n WT",
                            "1.M"="IGHV-M\n trisomy 12",
                            "1.U"="IGHV-U\n trisomy 12"))+
  stat_compare_means(method = "t.test",
                     label.x.npc = "center", 
                     comparisons = list(c(3,4),c(1,3),c(1,2)),
                     size=5)+
  xlab("") +
  ylab("Log(Viability)") +
  ggtitle("Resiquimod") +
  scale_color_manual(values=c(palblues[3], palreds[8])) + 
  coord_cartesian(ylim = c(-1.5, 3.1), clip="off") +
  t1+
  theme(axis.text.x = element_text(angle = 45, vjust =1))


rm(plotTab)

Fig4A


```

### Figure 4B
Ibrutinib Response stratified by IGHV and trisomy 12
```{r  Fig4B, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 5.5, fig.width = 6}

#filter for Resiquimod-only treatment, only show patients who are annoyated for IGHV and Trisomy12
plotTab <- left_join(df_complete, patMeta, by = "PatientID") %>% 
  dplyr::filter(Cytokine=="No Cytokine", Drug=="Ibrutinib", Drug_Concentration=="High",
                !is.na(trisomy12),
                !is.na(IGHV.status))

Fig4B =
 
  plotTab %>%
         
  ggplot(aes(x=interaction(trisomy12, IGHV.status),
                    y=Log,
                    color=(IGHV.status)))+
  geom_hline(yintercept = 0)+
  geom_boxplot()+
  geom_beeswarm(cex=1.5) +
  guides(color="none", shape="none")+
  scale_x_discrete(labels=c("0.M"="IGHV-M\n WT",
                            "0.U"="IGHV-U\n WT",
                            "1.M"="IGHV-M\n trisomy 12",
                            "1.U"="IGHV-U\n trisomy 12"))+
  stat_compare_means(method = "t.test",
                     label.x.npc = "center", 
                     comparisons = list(c(3,4),c(1,3),c(1,2)),
                     size=5)+
  xlab("") +
  ylab("Log(Viability)") +
  ggtitle("Ibrutinib") +
  scale_color_manual(values=c(palblues[3], palreds[8])) + 
  # coord_cartesian(ylim = c(-1.5, 3.1), clip="off") +
  t1+
  theme(axis.text.x = element_text(angle = 45, vjust =1))


rm(plotTab)

Fig4B


```

### Figure 4C
```{r,  message=FALSE}


dds_smp <- dds_smp_full
#filter for trisomy 12 aamples
dds_smp <- dds_smp[,dds_smp$trisomy12 %in% c("0","1")]

#remove genes without gene symbol annotations
dds_smp <- dds_smp[!is.na(rowData(dds_smp)$symbol),]
dds_smp <- dds_smp[rowData(dds_smp)$symbol != "",]

#only keep genes that have counts higher than 10 in any sample
keep <- apply(counts(dds_smp), 1, function(x) any(x >= 10)) 
dds_smp <- dds_smp[keep,]

#Remove transcripts that  do not show variance across samples
dds_smp <- estimateSizeFactors(dds_smp)
sds <- rowSds(counts(dds_smp, normalized = TRUE))
sh <- genefilter::shorth(sds)
dds_smp <- dds_smp[sds >= sh,]

#variance stabilisation
dds_smp.norm <- varianceStabilizingTransformation(dds_smp, blind=TRUE)

#how many genes left
# dim(dds_smp)

#Set design
design(dds_smp) <- ~ trisomy12

dds_smp <- DESeq(dds_smp)


DEres <- results(dds_smp)

```



```{r,  Fig4C, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 6, fig.width = 9, message=FALSE}

hm2gene <- 
  msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)

res.tidy <- 
  results(dds_smp, tidy = TRUE) %>%
  dplyr::rename(Symbol = "row") %>% 
  dplyr::arrange(pvalue) 

#get ensembl ids to Entrez dataframe
ens2entrez <- 
  rowData(dds_smp)[c("entrezgene", "symbol")] %>% 
  as.data.frame() %>% 
  rownames_to_column("ENSEMBL") %>%  
  tibble::as_tibble()
#Join 
res.tidy <- left_join(res.tidy, ens2entrez, by=c("Symbol"="ENSEMBL"))

#Get ranks dataframes to feed to fgsea

d <- 
  dplyr::select(res.tidy, entrezgene, stat) %>%
  na.omit() %>% 
  dplyr::group_by(entrezgene) %>% 
  dplyr::summarize(stat=mean(stat)) 

## feature 1: numeric vector
geneList <- d$stat

## feature 2: named vector
names(geneList) <- as.character(d$entrezgene)

## feature 3: decreasing order
geneList <- sort(geneList, decreasing = TRUE)

#run GSEA
gsea.res <- GSEA(geneList, TERM2GENE = hm2gene, by = "fgsea", seed = TRUE)

#make readable with gene names
gsea.res <- setReadable(gsea.res, org.Hs.eg.db, keyType = "ENTREZID")


#get dataframe of results
gsea.df <- fortify(gsea.res, 
              showCategory = 15, #how many levels to show
              split=NULL)
#order by NES
gsea.df <- dplyr::mutate(gsea.df, "NES" = eval(parse(text="NES")))
idx <- order(gsea.df[["NES"]], decreasing = TRUE)
gsea.df$Description <- factor(gsea.df$Description, levels=rev(unique(gsea.df$Description[idx])))

gsea.df$pos <- 0


Fig4C <-
  ggplot(gsea.df, 
         aes_string(x="NES", 
                    y="Description", 
                    fill="p.adjust")) +
    geom_bar(stat = "identity") +
    scale_fill_continuous(low=palreds[7],
                         high=palreds[2], 
                         name = "Adjusted p value",
                         guide=guide_colorbar(reverse=TRUE),
                         limits=c(0,0.05)) +
    
  ylab("Pathway\n\n") + 
  xlab("Normalised Enrichment Score") +
  ggtitle("GSEA for Trisomy 12 versus WT CLL") +  
  t2 + 
  scale_size(range=c(3, 8)) + 
  geom_text(aes(label=Description, x =ifelse(NES>0,0.1,-2)), hjust = 0,  size = 5, color ="white") +
    guides(fill="none")+
  theme(legend.position="right",
        # legend.background = element_blank(),
        # legend.box.background = element_rect(size = 0.5),
        legend.title = element_text(face='bold',
                                    hjust = 1, size=11),
        legend.key = element_blank(),
        legend.text = element_text(size=12),
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank()) 
  
Fig4C
```

### Figure 4D
List of genes for selected testing in Figure 4D:
```{r,  Fig4D,  fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 4, fig.width = 9, message=FALSE}

kegg2gene <-
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>%
  dplyr::select(gs_name, entrez_gene) %>%
  filter(gs_name%in%c("KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY",
                      "KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                      "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION",
                      "KEGG_JAK_STAT_SIGNALING_PATHWAY",
                      "KEGG_MTOR_SIGNALING_PATHWAY",
                      "KEGG_NOTCH_SIGNALING_PATHWAY",
                      "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                      "KEGG_TGF_BETA_SIGNALING_PATHWAY",
                      "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY",
                      "KEGG_VEGF_SIGNALING_PATHWAY",
                      "KEGG_WNT_SIGNALING_PATHWAY"
                      ))


unique(kegg2gene$gs_name)

res.tidy <- 
  results(dds_smp, tidy = TRUE) %>%
  dplyr::rename(Symbol = "row") %>% 
  dplyr::arrange(pvalue) 

#get ensembl ids to Entrez dataframe
ens2entrez <- 
  rowData(dds_smp)[c("entrezgene", "symbol")] %>% 
  as.data.frame() %>% 
  rownames_to_column("ENSEMBL") %>%  
  tibble::as_tibble()
#Join 
res.tidy <- left_join(res.tidy, ens2entrez, by=c("Symbol"="ENSEMBL"))

#Get ranks dataframes to feed to fgsea

d <- 
  dplyr::select(res.tidy, entrezgene, stat) %>%
  na.omit() %>% 
  dplyr::group_by(entrezgene) %>% 
  dplyr::summarize(stat=mean(stat)) 

## feature 1: numeric vector
geneList <- d$stat

## feature 2: named vector
names(geneList) <- as.character(d$entrezgene)

## feature 3: decreasing order
geneList <- sort(geneList, decreasing = TRUE)

#run GSEA
gsea.res <- GSEA(geneList, TERM2GENE = kegg2gene, by = "fgsea", seed = TRUE)

#make readable with gene names
gsea.res <- setReadable(gsea.res, org.Hs.eg.db, keyType = "ENTREZID")


#get dataframe of results
gsea.df <- fortify(gsea.res, 
              showCategory = 15, #how many levels to show
              split=NULL)
#order by NES
gsea.df <- dplyr::mutate(gsea.df, "NES" = eval(parse(text="NES")))
idx <- order(gsea.df[["NES"]], decreasing = TRUE)
gsea.df$Description <- factor(gsea.df$Description, levels=rev(unique(gsea.df$Description[idx])))

gsea.df$pos <- 0


Fig4D <-
   ggplot(gsea.df, 
         aes_string(x="NES", 
                    y="Description", 
                    fill="p.adjust")) +
    geom_bar(stat = "identity") +
    scale_fill_continuous(low=palreds[7],
                         high=palreds[2], 
                         name = "Adjusted p value",
                         guide=guide_colorbar(reverse=TRUE),
                         limits=c(0,0.05)) +
    
  ylab("Pathway\n\n") + 
  xlab("Normalised Enrichment Score") +
  t2 + 
  scale_size(range=c(3, 8)) + 
  geom_text(aes(label=Description, x =ifelse(NES>0,0.1,-2)), hjust = 0,  size = 5, color = "white") +
  theme(legend.position="bottom",
        legend.title = element_text(face='bold',
                                    hjust = 1, size=11),
        legend.key = element_blank(),
        legend.text = element_text(size=12),
        axis.text.y = element_blank(), 
        axis.ticks.y = element_blank(),
        legend.key.width = unit(3, "cm")) 
  
Fig4D

```

### Figure 4E
Trisomy 12 versus WT ATACseq
Comparison of TF binding site accessibility shows that SPIB and other TFs show higher accessibility of their binding sites in trisomy 12 in 2 independent datasets.
```{r, Fig4E, fig.path = plotDir, dev = c("png", "cairo_pdf"), fig.height = 6, fig.width=15}

#Get all TFs that show differential activity
plotTab.diffTF <- dplyr::filter(diffTF_large, pvalueAdj < 0.05)

#change to name of TF quoted in text
plotTab.diffTF$TF <- gsub("SPI1", "PU1", plotTab.diffTF$TF)

#order by weighted mean difference
idx <- order(plotTab.diffTF[["weighted_meanDifference"]], decreasing = TRUE)
plotTab.diffTF$TF <- factor(plotTab.diffTF$TF,levels=rev(unique(plotTab.diffTF$TF[idx])))


#plot figure
Fig4E <-
  ggplot(plotTab.diffTF,
         aes(x=weighted_meanDifference,
             y=TF,
             fill= ifelse(weighted_meanDifference>0,
                          palreds[5], palblues[1]))) +
  geom_bar(stat = "identity", width = 0.75) +
  xlim(c(-0.2, 0.175)) +

  #colours and theme
  scale_fill_manual(values = c(palblues[1], palreds[8]), guide = "none") +
  t1 +
  theme(axis.text.x = element_text(angle = 45, vjust =1))+

  #Labels
  ylab("") +
  xlab("Change in inferred TF activity") +
  ggtitle("Diffential TF binding site accessibility in trisomy 12 CLL") +

  #Add annotations
  #arrow 1
  annotate("segment", x = -0.15, xend = -0.15, y = 8.4, yend = 0.75,
           colour = "#5e5e5e", size=3, alpha=1, arrow=arrow()) +
  #arrow 2
  annotate("segment", x = -0.15, xend = -0.15, y = 8.6, yend = 17,
           colour = "#5e5e5e", size=3, alpha=1, arrow=arrow()) +

   #2 arrow labels
  annotate("text", x = c(-0.185,-0.185), y = c(4.8,12.5),
          label = c("Lower in trisomy 12 CLL",
                    "Higher in trisomy 12 CLL") ,
          color="black",
          size=5 , fontface="bold") +

  #flip plot
  coord_flip()




Fig4E



```

### Figure 4F
Here, for Spi-B  ChIP peaks in an OCILY3 DLBCL cell line, we use the `clusterProfiler` package to annotate the nearest gene for each significant ChIPpeak (q<0.05). We filter the resulting gene list for those with a peak within ±1kb of the TSS and tested for over-representation of selected KEGG and Reactome pathways.     


```{R, Fig4F,  fig.path = plotDir, dev = c("png", "cairo_pdf"), fig.width=15, fig.height=2, message=FALSE }
# Get significant peaks
#filter ChipSeq data for peaks with significant score -log10(Qscore)>1.3
#if q < 0.05, then Qlog >1.3
spib.filtered <- peak.spib[(elementMetadata(peak.spib)[,1] > 1.3)]


# Annotate ChIP peaks
# Annotate peaks with meta data
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

peakAnno.spib <- annotatePeak(spib.filtered, 
                              tssRegion = c(-3000, 3000),
                              TxDb = txdb, 
                              annoDb = "org.Hs.eg.db",
                              verbose=FALSE)

#Filter for genes within ±1kb
peakAnno.spib.1kb <- as.data.frame(peakAnno.spib) %>% dplyr::filter(abs(distanceToTSS) <1000)


#get  KEGG pathways from msigdbr database
kegg <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG") %>% 
  #select pathway name and entrez id 
  dplyr::select(gs_name, entrez_gene)


#subset for pathways that we want to test
kegg <-
  dplyr::filter(kegg, 
         #All KEGG immune signaling pathways
         gs_name %in% c("KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_CHEMOKINE_SIGNALING_PATHWAY",
                        "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION", 
                        "KEGG_JAK_STAT_SIGNALING_PATHWAY" ,
                        "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY"  , 
                        "KEGG_NOTCH_SIGNALING_PATHWAY", 
                        "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY", 
                        "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY",
                        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY",
                        #Control genesets
                        "KEGG_OXIDATIVE_PHOSPHORYLATION",                  
                        "KEGG_DNA_REPLICATION"
         ))

colnames(kegg) <- c("term", "gene")

#tidy up terms - remove KEGG, underscore and put in lower cases
kegg$term <- gsub("KEGG_", "",kegg$term )
kegg$term <- gsub("_", " ",kegg$term )
kegg$term <- gsub("B CELL RECEPTOR SIGNALING PATHWAY", 
                  "BCR Signaling Pathway",kegg$term )
kegg$term <- gsub("CHEMOKINE SIGNALING PATHWAY", 
                  "Chemokine Signaling Pathway",kegg$term )
kegg$term <- gsub("CYTOKINE CYTOKINE RECEPTOR INTERACTION", 
                  "Cytokine Cytokine Receptor Interaction",kegg$term )
kegg$term <- gsub("JAK STAT SIGNALING PATHWAY", 
                  "JAK STAT signaling pathway",kegg$term )
kegg$term <- gsub("NOD LIKE RECEPTOR SIGNALING PATHWAY", 
                  "NLR Signaling Pathway",kegg$term )
kegg$term <- gsub("NOTCH SIGNALING PATHWAY", 
                  "Notch Signaling Pathway",kegg$term )
kegg$term <- gsub("RIG I LIKE RECEPTOR SIGNALING PATHWAY", 
                  "RLR Signaling Pathway",kegg$term )
kegg$term <- gsub("TOLL LIKE RECEPTOR SIGNALING PATHWAY", 
                  "TLR Signaling Pathway",kegg$term )

kegg$term <- gsub("T CELL RECEPTOR SIGNALING PATHWAY", 
                  "TCR Signaling Pathway",kegg$term )
kegg$term <- gsub("OXIDATIVE PHOSPHORYLATION", 
                  "Oxidative Phosphorylation",kegg$term )
kegg$term <- gsub("P53 SIGNALING PATHWAY", 
                  "P53 Signaling Pathway",kegg$term )
kegg$term <- gsub("DNA REPLICATION", 
                  "DNA Replication",kegg$term )

kegg$genesetDatabase <- "KEGG"


#get reactome pathways
reactome <- 
  msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME") %>% 
  #select pathway name and entrez id
  dplyr::select(gs_name, entrez_gene)

#subset for pathways that we want to test
reactome <-
  dplyr::filter(reactome, 
         gs_name %in% c(#selected immune pathways
                        "REACTOME_SIGNALING_BY_TGF_BETA_RECEPTOR_COMPLEX",
                        "REACTOME_TNF_SIGNALING",
                        "REACTOME_INTERLEUKIN_1_SIGNALING",
                        "REACTOME_INTERLEUKIN_10_SIGNALING",
                        "REACTOME_INTERLEUKIN_15_SIGNALING",
                        "REACTOME_INTERLEUKIN_2_SIGNALING",
                        "REACTOME_INTERLEUKIN_4_AND_INTERLEUKIN_13_SIGNALING",
                        "REACTOME_INTERLEUKIN_6_SIGNALING" ,
                        "REACTOME_OTHER_INTERLEUKIN_SIGNALING",
                        "REACTOME_SIGNALING_BY_INTERLEUKINS",
                        "REACTOME_INTERFERON_ALPHA_BETA_SIGNALING" ,
                        "REACTOME_INTERFERON_GAMMA_SIGNALING", 
                        #Control genesets
                        "REACTOME_CELL_CYCLE",                  
                        "REACTOME_REGULATION_OF_TP53_ACTIVITY" 
                        ))

colnames(reactome) <- c("term", "gene")

#tidy up terms
reactome$term <- gsub("REACTOME_", "",reactome$term )
reactome$term <- gsub("_", " ",reactome$term )


#tidy up and put in lower case
reactome$term <- gsub("CELL CYCLE", 
                      "Cell Cycle",reactome$term )
reactome$term <- gsub("INTERLEUKIN 1 SIGNALING", 
                      "Interleukin 1 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 10 SIGNALING", 
                      "Interleukin 10 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 15 SIGNALING", 
                      "Interleukin 15 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 2 SIGNALING", 
                      "Interleukin 2 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 4 AND INTERLEUKIN 13 SIGNALING", 
                      "Interleukin 4 and Interleukin 13 Signaling",reactome$term )
reactome$term <- gsub("INTERLEUKIN 6 SIGNALING",
                      "Interleukin 6 Signaling",reactome$term )
reactome$term <- gsub("OTHER INTERLEUKIN SIGNALING", 
                      "Other Interleukin Signaling",reactome$term )
reactome$term <- gsub("SIGNALING BY INTERLEUKINS", 
                      "Signaling by Interleukins",reactome$term )
reactome$term <- gsub("REGULATION OF TP53 ACTIVITY", 
                      "Regulation of TP53 Pathway",reactome$term )
reactome$term <- gsub("SIGNALING BY TGF BETA RECEPTOR COMPLEX",
                      "Signaling by TGFbeta \nReceptor Complex", reactome$term )
reactome$term <- gsub("TNF SIGNALING", 
                      "TNF Signaling",reactome$term )
reactome$term <- gsub("INTERFERON ALPHA BETA SIGNALING", 
                      "Interferon Alpha Beta Signaling",reactome$term )

reactome$term <- gsub("INTERFERON GAMMA SIGNALING", 
                      "Interferon Gamma Signaling", reactome$term )

#annotate which database these pathways come from 
reactome$genesetDatabase <- "Reactome"



#generate TERM2GENE dataframe to use in overrepresetation test
#join kegg and reactome gene sets together - this defines the list of pathways we are interested in, and select term and gene columns
term2gene <- rbind(kegg, reactome) %>% dplyr::select(term, gene)

#term2genesetdatabase - keep details of database source for each pathway 
term2genesetdatabase <- rbind(kegg, reactome) %>% dplyr::select(term, genesetDatabase)
colnames(term2genesetdatabase) <- c("term", "name")

##SPIB
#run over-representation test, and extract all results and p values:
#NB Bg ratio = number of genes in set / number of unique genes in term2gene, 
#NB Gene ratio = number of SPIB targets in gene set being tested / total number of SPIB targets that overlap with genes in term to gene

spibRes <- enricher(unique(as.data.frame(peakAnno.spib.1kb)$geneId), 
                    #Use pathways in term2gene
                    TERM2GENE=term2gene, 
                    #assign description for each term, as its database source
                    TERM2NAME = term2genesetdatabase, 
                    #no cutoffs on reported results, we want to see all results and p values 
                    minGSSize = 0,
                    maxGSSize = 1000,
                    pvalueCutoff = 1, 
                    qvalueCutoff = 1)


#get dataframe of results
spibRes.df <- fortify(spibRes, 
              showCategory = length(unique(term2gene$term)), #show all pathways tested
              split=NULL)

resTab.spib <-  
  mutate(spibRes.df, genesetSize = BgRatio*length(unique(term2gene$gene))) %>%  #multiply BgRatio by total genes tested to get number of genes in set
  dplyr::select(ID, Description, pvalue, Count, genesetSize)

colnames(resTab.spib) <- c("ID", "genesetDatabase", "pvalue_SPIB","count_SPIB", "genesetSize")


resTab <- resTab.spib

#where count and p value is NA, put a 0 / 1 (ie there were no SPIB target genes in the set)
resTab$count_SPIB[is.na(resTab$count_SPIB)] <- 0

resTab$pvalue_SPIB[is.na(resTab$pvalue_SPIB)] <- 1

#select columns of interest to show in figure
resTab <- dplyr::select(resTab, ID, genesetDatabase, genesetSize,  count_SPIB, pvalue_SPIB)

#adjust decimal places of p values
resTab$pvalue_SPIB <- round(resTab$pvalue_SPIB, digits = 3)

#get significant results
resTab <- dplyr::filter(resTab, pvalue_SPIB<0.01)

colnames(resTab) <- c("Pathway",
                      "Geneset\nDatabase" , 
                      "Geneset\nSize",  
                      
                      #show the total number of SPIB targets in ChIPseq data 
                      paste("Spi-B targets\n(/" ,length(unique(as.data.frame(peakAnno.spib.1kb)$geneId)),")", sep= ""), 
                      "SPIB\np-value")

Fig4F <- 
ggtexttable(resTab, rows = NULL, 
           theme = ttheme(
             colnames.style = colnames_style(color = "white", fill = palreds[7], size = 16),
             tbody.style = tbody_style(color = "black",size = 18)
           )
)


Fig4F

```

### Figure 4G
IBET-762 response stratified by trisomy 12
```{r, Fig4G, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height = 5, fig.width = 5}

Fig4G <- 
  left_join(df_complete, patMeta, by = "PatientID") %>% 
  dplyr::filter(Drug=="I-BET 762", Drug_Concentration == "High", Cytokine == "No Cytokine") %>%
      filter(!is.na(trisomy12)) %>%
         
  ggplot(aes(x = trisomy12, y = Log, color = (trisomy12)))+
  geom_hline(yintercept = 0)+
  geom_boxplot()+
  geom_beeswarm(cex=1.5) +
  guides(color="none", shape="none")+
  stat_compare_means(method = "t.test",
                     label.x.npc = "center", 
                     size=5,
                     comparisons = list(c(1,2)))+
  ylab("Log(Viability)") +
  ggtitle("IBET-762") +
  scale_color_manual(values=c(colors[1], colors[2])) + 
  coord_cartesian(ylim = c(-1.25, 0.6), clip="off") +

  t2+
  scale_x_discrete(labels=c("0"="WT",
                            "1"="trisomy 12"))+
  xlab("")

Fig4G

```


### Figure 4H
Comparison of TF binding site accessibility in trisomy 12 vs non-trisomy 12 samples and in IBET vs control treated samples  

```{r, Fig4H, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.width = 12, fig.height=10}

colnames(ibet_TFs) <- c("Symbol", "ENSEMBL", "TF")

#add ensembl IDs to diffTF_ibet
diffTF.ibet <- left_join(diffTF.ibet, ibet_TFs[2:3], by = "TF")
colnames(diffTF.ibet)[1] <- "TF_original"

#add TF names from tri12 run to diffTF_ibet
colnames(tri12_TFs)[3] <- "TF"
diffTF.ibet <- left_join(diffTF.ibet, tri12_TFs[2:3], by = "ENSEMBL")

#change to name of TF quoted in text
diffTF.ibet$TF <- gsub("SPI1", "PU1", diffTF.ibet$TF)

#filter for tri 12 signature TFs
plotTab.diffTF.ibet <- diffTF.ibet %>% filter(TF %in%  plotTab.diffTF$TF)

#order by weighted mean difference
idx <- order(plotTab.diffTF.ibet[["weighted_meanDifference"]], decreasing = TRUE)
plotTab.diffTF.ibet$TF <- factor(plotTab.diffTF.ibet$TF,levels=rev(unique(plotTab.diffTF.ibet$TF[idx])))


#make plotting table 
ibet_plot <- plotTab.diffTF.ibet %>% select(TF, weighted_meanDifference)
colnames(ibet_plot) <- c("TF", "ibet")
untreated <- plotTab.diffTF %>% select(TF, weighted_meanDifference)
colnames(untreated) <- c("TF", "tri12")


plotTab <- left_join(untreated, ibet_plot, by = "TF")
plotTab %<>% pivot_longer(c("ibet", "tri12"))
colnames(plotTab) <- c("TF", "condition", "WMD")

#plot figure
Fig4H <-
  ggplot(plotTab,
         aes(x=WMD,
             y=TF,
             fill=condition)) +
  geom_bar(position = "dodge", stat = "identity", width = 0.75) +
  xlim(c(-0.2, 0.175)) +
  
  #colours and theme
  scale_fill_manual(values = c(colors[1], colors[3]), labels = c("tri12"="Trisomy 12 versus WT", "ibet"="IBET-762 versus Control")) +
  t1 + 
  theme(axis.text.x = element_text(angle = 45, vjust =1))+
   theme(legend.position="bottom",
        legend.title = element_text(face='bold', hjust = 1, size=14),
        legend.key = element_blank(),
        legend.text = element_text(size=15)
        )+
  
  #Labels
  ylab("") + 
  xlab("Change in inferred TF activity") +
  ggtitle("Diffential TF binding site accessibility in trisomy 12 CLL and after IBET-762 treatment") + 
  
  #Add annotations
  #arrow 1
  annotate("segment", x = -0.15, xend = -0.15, y = 8.4, yend = 0.75, 
           colour = "#5e5e5e", size=3, alpha=1, arrow=arrow()) + 
  #arrow 2
  annotate("segment", x = -0.15, xend = -0.15, y = 8.6, yend = 17, 
           colour = "#5e5e5e", size=3, alpha=1, arrow=arrow()) + 
 
   #2 arrow labels
  annotate("text", x = c(-0.185,-0.185), y = c(4.8,12.5), 
          label = c("Lower in trisomy 12 CLL / \nHigher after IBET-762 treatment", 
                    "Higher in trisomy 12 CLL / \nLower after IBET-762 treatment") , 
          color="black", 
          size=5 , fontface="bold") + theme(legend.title = element_text(face = "bold"))+

  #flip plot
  coord_flip() +
  theme(legend.position="bottom")+
  guides(fill=guide_legend(title="diffTF comparison"))
  


Fig4H

rm(plotTab)

```

## Assemble Figure 4
```{r Fig4, fig.path=plotDir, dev=c("png", "cairo_pdf"), fig.height=18, fig.width=19, eval = TRUE}

design1<-"
  ABC
  ABD
  EED
  EEF
  GHH
"

tp <- theme(plot.tag=element_text(size = 30, face="plain"))

Fig4 <-


 wrap_elements( Fig4A )+ tp +
 wrap_elements( Fig4B )+ tp +
 wrap_elements( Fig4C )+ tp + 
 wrap_elements( Fig4D )+ tp + 
 wrap_elements( Fig4E )+ tp +  
 wrap_elements( Fig4F )+ tp +
 wrap_elements( Fig4G )+ tp+
 wrap_elements( Fig4H )+ tp+
    
  plot_annotation(tag_levels = "A", title="Figure 4", theme = theme(title=element_text(size = 20)))+
  plot_layout(design = design1, heights = c(0.64,.26,.18,.4,1,1), width=c(1.5,1.5,3))

Fig4


```

## Appendix
```{r appendix4}
Sys.info()
sessionInfo()
```